<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.353">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Mary Piper, Meeta Mistry, Radhika Khetani, Jihe Liu">
<meta name="dcterms.date" content="2022-11-30">

<title>Introduction to single-cell RNA-seq - Single-cell RNA-seq: Normalization, identification of most variable genes</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>


<link rel="stylesheet" href="../css/styles.css">
</head>

<body class="nav-sidebar docked nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">Introduction to single-cell RNA-seq</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../lessons/schedule.html" rel="" target="">
 <span class="menu-text">Schedule</span></a>
  </li>  
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="https://bioinformatics.sph.harvard.edu/" rel="" target="">
 <span class="menu-text">HBC</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="https://github.com/hbctraining/intro_r_rmd" rel="" target="">
 <span class="menu-text">GitHub</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="mailto:hbctraining@hsph.harvard.edu" rel="" target="">
 <span class="menu-text">Contact us</span></a>
  </li>  
</ul>
            <div class="quarto-navbar-tools">
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../lessons/06_SC_SCT_normalization.html">Day 2:</a></li><li class="breadcrumb-item"><a href="../lessons/06_SC_SCT_normalization.html">Single-cell RNA-seq: Normalization, identification of most variable genes</a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation docked overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="false">
 <span class="menu-text">Pre-reading:</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lessons/01_intro_to_scRNA-seq.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Introduction to single-cell RNA-seq</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lessons/02_SC_generation_of_count_matrix.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Generation of count matrix</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="false">
 <span class="menu-text">Day 1:</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lessons/03_SC_quality_control-setup.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Single-cell RNA-seq: Quality Control Analysis</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="false">
 <span class="menu-text">Day 1 Self-learning:</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lessons/04_cellranger_QC.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Single-cell RNA-seq: Quality Control of Cellranger Output</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lessons/04_SC_quality_control.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Single-cell RNA-seq: Quality Control Analysis</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lessons/05_theory_of_PCA.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Single-cell RNA-seq: Theory of PCA</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true">
 <span class="menu-text">Day 2:</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lessons/06_SC_SCT_normalization.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">Single-cell RNA-seq: Normalization, identification of most variable genes</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lessons/06a_integration_cca_theory.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Single-cell RNA-seq: Integration</span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#learning-objectives" id="toc-learning-objectives" class="nav-link active" data-scroll-target="#learning-objectives">Learning Objectives:</a></li>
  <li><a href="#single-cell-rna-seq-normalization-and-regressing-out-unwanted-variation" id="toc-single-cell-rna-seq-normalization-and-regressing-out-unwanted-variation" class="nav-link" data-scroll-target="#single-cell-rna-seq-normalization-and-regressing-out-unwanted-variation">Single-cell RNA-seq: Normalization and regressing out unwanted variation</a>
  <ul class="collapse">
  <li><a href="#normalization" id="toc-normalization" class="nav-link" data-scroll-target="#normalization">Normalization</a>
  <ul class="collapse">
  <li><a href="#methods-for-scrna-seq-normalization" id="toc-methods-for-scrna-seq-normalization" class="nav-link" data-scroll-target="#methods-for-scrna-seq-normalization">Methods for scRNA-seq normalization</a></li>
  </ul></li>
  <li><a href="#explore-sources-of-unwanted-variation" id="toc-explore-sources-of-unwanted-variation" class="nav-link" data-scroll-target="#explore-sources-of-unwanted-variation">Explore sources of unwanted variation</a>
  <ul class="collapse">
  <li><a href="#set-up" id="toc-set-up" class="nav-link" data-scroll-target="#set-up">Set-up</a></li>
  <li><a href="#evaluating-effects-of-cell-cycle" id="toc-evaluating-effects-of-cell-cycle" class="nav-link" data-scroll-target="#evaluating-effects-of-cell-cycle">Evaluating effects of cell cycle</a></li>
  <li><a href="#pca" id="toc-pca" class="nav-link" data-scroll-target="#pca">PCA</a></li>
  <li><a href="#using-pca-to-evaluate-the-effects-of-cell-cycle" id="toc-using-pca-to-evaluate-the-effects-of-cell-cycle" class="nav-link" data-scroll-target="#using-pca-to-evaluate-the-effects-of-cell-cycle">Using PCA to evaluate the effects of cell cycle</a></li>
  </ul></li>
  <li><a href="#normalization-and-regressing-out-sources-of-unwanted-variation-using-sctransform" id="toc-normalization-and-regressing-out-sources-of-unwanted-variation-using-sctransform" class="nav-link" data-scroll-target="#normalization-and-regressing-out-sources-of-unwanted-variation-using-sctransform">Normalization and regressing out sources of unwanted variation using SCTransform</a></li>
  <li><a href="#iterating-over-samples-in-a-dataset" id="toc-iterating-over-samples-in-a-dataset" class="nav-link" data-scroll-target="#iterating-over-samples-in-a-dataset">Iterating over samples in a dataset</a>
  <ul class="collapse">
  <li><a href="#save-the-object" id="toc-save-the-object" class="nav-link" data-scroll-target="#save-the-object">Save the object!</a></li>
  </ul></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title">Single-cell RNA-seq: Normalization, identification of most variable genes</h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source"><i class="bi"></i> Code</button></div></div>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Mary Piper, Meeta Mistry, Radhika Khetani, Jihe Liu </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">November 30, 2022</p>
    </div>
  </div>
  
    
  </div>
  

</header>

<p>Approximate time: 90 minutes</p>
<section id="learning-objectives" class="level2">
<h2 class="anchored" data-anchor-id="learning-objectives">Learning Objectives:</h2>
<ul>
<li>Discuss why normalizing counts is necessary for accurate comparison between cells</li>
<li>Describe different normalization approaches</li>
<li>Evaluate the effects from any unwanted sources of variation and correct for them</li>
</ul>
</section>
<section id="single-cell-rna-seq-normalization-and-regressing-out-unwanted-variation" class="level1">
<h1>Single-cell RNA-seq: Normalization and regressing out unwanted variation</h1>
<p>Now that we have our high quality cells, we can explore our data and see if we are able to identify any sources of unwanted variation. Depending on what we observe, we will utilize that information when performing variance stabilization using SCTransform but also to regress out the effects of any covariates that have an effect on our data.</p>
<p align="center">
<img src="../img/sc_workflow_2022.jpg" width="630">
</p>
<hr>
<p><em><strong>Goals:</strong></em></p>
<ul>
<li><em>To accurately <strong>normalize the gene expression values</strong> to account for differences in sequencing depth and overdispersed count values.</em></li>
<li><em>To <strong>identify the most variant genes</strong> likely to be indicative of the different cell types present.</em></li>
</ul>
<p><em><strong>Challenges:</strong></em></p>
<ul>
<li><em><strong>Checking and removing unwanted variation</strong> so that we do not have cells clustering by artifacts downstream</em></li>
</ul>
<p><em><strong>Recommendations:</strong></em></p>
<ul>
<li><em>Have a good idea of your expectations for the <strong>cell types to be present</strong> prior to performing the clustering. Know whether you expect cell types of low complexity or higher mitochondrial content AND whether the cells are differentiating</em></li>
<li><em><strong>Regress out</strong> number of UMIs (default using sctransform), mitochondrial content, and cell cycle, if needed and appropriate for experiment, so not to drive clustering downstream</em></li>
</ul>
<hr>
<section id="normalization" class="level2">
<h2 class="anchored" data-anchor-id="normalization">Normalization</h2>
<p>An essential first step in the majority of mRNA expression analyses is normalization, whereby systematic variations are adjusted for to <strong>make expression counts comparable across genes and/or samples</strong>. The counts of mapped reads for each gene is proportional to the expression of RNA (“interesting”) in addition to many other factors (“uninteresting”). Normalization is the process of adjusting raw count values to account for the “uninteresting” factors.</p>
<p>The main factors often considered during normalization are:</p>
<ul>
<li><strong>Sequencing depth:</strong> Accounting for sequencing depth is necessary for comparison of gene expression between cells. In the example below, each gene appears to have doubled in expression in cell 2, however this is a consequence of cell 2 having twice the sequencing depth.</li>
</ul>
<p align="center">
<img src="../img/sequencing_depth.png" width="400">
</p>
<p>Each cell in scRNA-seq will have a differing number of reads associated with it. So to accurately compare expression between cells, it is necessary to normalize for sequencing depth.</p>
<ul>
<li><strong>Gene length:</strong> Accounting for gene length is necessary for comparing expression between different genes within the same cell. The number of reads mapped to a longer gene can appear to have equal count/expression as a shorter gene that is more highly expressed.</li>
</ul>
<p align="center">
<img src="../img/length_of_gene.png" width="400">
</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>If using a 3’ or 5’ droplet-based method, the length of the gene will not affect the analysis because only the 5’ or 3’ end of the transcript is sequenced.* However, if using full-length sequencing, the transcript length should be accounted for.</p>
</div>
</div>
<section id="methods-for-scrna-seq-normalization" class="level3">
<h3 class="anchored" data-anchor-id="methods-for-scrna-seq-normalization">Methods for scRNA-seq normalization</h3>
<p>Various methods have been developed specifically for scRNA-seq normalization. Some <strong>simpler methods resemble what we have seen with bulk RNA-seq</strong>; the application of <strong>global scale factors</strong> adjusting for a count-depth relationship that is assumed common across all genes. However, if those assumptions are not true then this basic normalization can lead to over-correction for lowly and moderately expressed genes and, in some cases, under-normalization of highly expressed genes (<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5473255/">Bacher R et al, 2017</a>). <strong>More complex methods will apply correction on a per-gene basis.</strong> In this lesson we will explore both approaches.</p>
<p>Regardless of which method is used for normalization, it can be helpful to <strong>think of it as a two-step process</strong> (even though it is often described as a single step in most papers). The first is a scaling step and the second is a transformation.</p>
<p><strong>1. Scaling</strong></p>
<p>The first step in normalization is to <strong>multiply each UMI count by a cell specific factor to get all cells to have the same UMI counts</strong>. Why would we want to do this? Different cells have different amounts of mRNA; this could be due to differences between cell types or variation within the same cell type depending on how well the chemistry worked in one drop versus another. In either case, we are not interested in comparing these absolute counts between cells. Instead we are interested in comparing concentrations, and scaling helps achieve this.</p>
<p><strong>2. Transformation</strong></p>
<p>The next step is a transformation, and it is at this step where we can distinguish the simpler versus complex methods as mentioned above.</p>
<p><strong>Simple transformations</strong> are those which apply the same function to each individual measurement. Common examples include a <strong>log transform</strong> (which is applied in the original Seurat workflow), or a square root transform (less commonly used).</p>
</section>
</section>
<section id="explore-sources-of-unwanted-variation" class="level2">
<h2 class="anchored" data-anchor-id="explore-sources-of-unwanted-variation">Explore sources of unwanted variation</h2>
<p>The most common biological data correction (or source of “uninteresting” variation) in single cell RNA-seq is the effects of the cell cycle on the transcriptome. We need to explore the data and see if we observe any effects in our data.</p>
<section id="set-up" class="level3">
<h3 class="anchored" data-anchor-id="set-up">Set-up</h3>
<p>Let’s start by creating a new script for the normalization and integration steps. Create a new script (File -&gt; New File -&gt; R script), and save it as <code>SCT_integration_analysis.R</code>.</p>
<p>For the remainder of the workflow we will be mainly using functions available in the Seurat package. Therefore, we need to load the Seurat library in addition to the tidyverse library and a few others listed below.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Single-cell RNA-seq - normalization</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Load libraries</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(RCurl)</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(cowplot)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Before we make any comparisons across cells, we will <strong>apply a simple normalization.</strong> This is solely for the purpose of exploring the sources of variation in our data.</p>
<p>The input for this analysis is a <code>seurat</code> object. We will use the one that we created in the QC lesson called <code>filtered_seurat</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Normalize the counts</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>seurat_phase <span class="ot">&lt;-</span> <span class="fu">NormalizeData</span>(filtered_seurat)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Next, we take this normalized data and check to see if data correction methods are necessary.</p>
</section>
<section id="evaluating-effects-of-cell-cycle" class="level3">
<h3 class="anchored" data-anchor-id="evaluating-effects-of-cell-cycle">Evaluating effects of cell cycle</h3>
<p>To assign each cell a score based on its expression of G2/M and S phase markers, we can use the Seuart function <code>CellCycleScoring()</code>. This function calculates cell cycle phase scores based on canonical markers that required as input.</p>
<p>We have provided a list of human cell cycle markers for you in the <code>data</code> folder as an Rdata file called <code>cycle.rda</code>. However, if you are not working with human data we have <a href="cell_cycle_scoring.md">additional materials</a> detailing how to acquire cell cycle markers for other organisms of interest.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load cell cycle markers</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="st">"../data/cycle.rda"</span>)</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Score cells for cell cycle</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>seurat_phase <span class="ot">&lt;-</span> <span class="fu">CellCycleScoring</span>(seurat_phase, </span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">g2m.features =</span> g2m_genes, </span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">s.features =</span> s_genes)</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a><span class="co"># View cell cycle scores and phases assigned to cells</span></span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(seurat_phase<span class="sc">@</span>meta.data)                                </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                      orig.ident nCount_RNA nFeature_RNA seq_folder nUMI nGene
ctrl_AAACATACAATGCC-1       ctrl       2344          874       ctrl 2344   874
ctrl_AAACATACATTTCC-1       ctrl       3124          895       ctrl 3125   896
ctrl_AAACATACCAGAAA-1       ctrl       2578          725       ctrl 2578   725
ctrl_AAACATACCAGCTA-1       ctrl       3260          978       ctrl 3261   979
ctrl_AAACATACCATGCA-1       ctrl        746          362       ctrl  746   362
ctrl_AAACATACCTCGCT-1       ctrl       3518          865       ctrl 3519   866
                      log10GenesPerUMI  mitoRatio                 cells sample
ctrl_AAACATACAATGCC-1        0.8728630 0.01962457 ctrl_AAACATACAATGCC-1   ctrl
ctrl_AAACATACATTTCC-1        0.8447596 0.01792000 ctrl_AAACATACATTTCC-1   ctrl
ctrl_AAACATACCAGAAA-1        0.8384933 0.01551590 ctrl_AAACATACCAGAAA-1   ctrl
ctrl_AAACATACCAGCTA-1        0.8512622 0.01379945 ctrl_AAACATACCAGCTA-1   ctrl
ctrl_AAACATACCATGCA-1        0.8906861 0.02144772 ctrl_AAACATACCATGCA-1   ctrl
ctrl_AAACATACCTCGCT-1        0.8283053 0.01392441 ctrl_AAACATACCTCGCT-1   ctrl
                          S.Score   G2M.Score Phase
ctrl_AAACATACAATGCC-1  0.04330502  0.05422631   G2M
ctrl_AAACATACATTTCC-1  0.02661900  0.05159679   G2M
ctrl_AAACATACCAGAAA-1 -0.04670650 -0.04841661    G1
ctrl_AAACATACCAGCTA-1 -0.05832833  0.05045960   G2M
ctrl_AAACATACCATGCA-1  0.03929605 -0.02995512     S
ctrl_AAACATACCTCGCT-1  0.03201279  0.01685778     S</code></pre>
</div>
</div>
<p>After scoring the cells for cell cycle, we would like to <strong>determine whether cell cycle is a major source of variation in our dataset using PCA</strong>.</p>
</section>
<section id="pca" class="level3">
<h3 class="anchored" data-anchor-id="pca">PCA</h3>
<p>Principal Component Analysis (PCA) is a technique used to emphasize variation as well as similarity, and to bring out strong patterns in a dataset; it is one of the methods used for <em>“dimensionality reduction”</em>.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>For a more <strong>detailed explanation on PCA</strong>, please <a href="../lessons/05_theory_of_PCA.html">look over this lesson</a> (adapted from StatQuests/Josh Starmer’s YouTube video). We also strongly encourage you to explore the video <a href="https://www.youtube.com/watch?v=_UVHneBUBW0">StatQuest’s video</a> for a more thorough understanding.</p>
</div>
</div>
<p>Let’s say you are working with a single-cell RNA-seq dataset with <em>12,000 cells</em> and you have quantified the expression of <em>20,000 genes</em>. The schematic below demonstrates how you would go from a cell x gene matrix to principal component (PC) scores for each inividual cell.</p>
<p align="center">
<img src="../img/PCA_scrnaseq_1.png" width="900">
</p>
<p>After the PC scores have been calculated, you are looking at a matrix of 12,000 x 12,000 that represents the information about relative gene expression in all the cells. You can select the PC1 and PC2 columns and plot that in a 2D way.</p>
<p align="center">
<img src="../img/PCA_scrnaseq_2.png" width="600">
</p>
<p>You can also use the PC scores from the first 40 PCs for downstream analysis like clustering, marker identification etc., since these represent the majority of the variation in the data. We will be talking a lot more about this later in this workshop.</p>
<p align="center">
<img src="../img/PCA_scrnaseq_3.png" width="600">
</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>For datasets with a larger number of cells, only the PC1 and PC2 scores for each cell are usually plotted, or used for visualization. Since these PCs explain the most variation in the dataset, the expectation is that the cells that are more similar to each other will cluster together with PC1 and PC2.</p>
</div>
</div>
</section>
<section id="using-pca-to-evaluate-the-effects-of-cell-cycle" class="level3">
<h3 class="anchored" data-anchor-id="using-pca-to-evaluate-the-effects-of-cell-cycle">Using PCA to evaluate the effects of cell cycle</h3>
<p>To perform PCA, we need to <strong>first choose the most variable features, then scale the data</strong>. Since highly expressed genes exhibit the highest amount of variation and we don’t want our ‘highly variable genes’ only to reflect high expression, we need to scale the data to scale variation with expression level. The Seurat <code>ScaleData()</code> function will scale the data by:</p>
<ul>
<li>Adjusting the expression of each gene to give a mean expression across cells to be 0</li>
<li>Scaling expression of each gene to give a variance across cells to be 1</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Identify the most variable genes</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>seurat_phase <span class="ot">&lt;-</span> <span class="fu">FindVariableFeatures</span>(seurat_phase, </span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>                     <span class="at">selection.method =</span> <span class="st">"vst"</span>,</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>                     <span class="at">nfeatures =</span> <span class="dv">2000</span>, </span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>                     <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>             </span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Scale the counts</span></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>seurat_phase <span class="ot">&lt;-</span> <span class="fu">ScaleData</span>(seurat_phase)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>For the <code>selection.method</code> and <code>nfeatures</code> arguments the values specified are the default settings. Therefore, you do not necessarily need to include these in your code. We have included it here for transparency and inform you what you are using._</p>
</div>
</div>
<p>Highly variable gene selection is extremely important since many downstream steps are computed only on these genes. Seurat allows us to access the ranked highly variable genes with the <code>VariableFeatures()</code> function. We can additionally visualize the dispersion of all genes using Seurat’s <code>VariableFeaturePlot()</code>, which shows a gene’s average expression across all cells on the x-axis and variance on the y-axis. Ideally we want to use genes that have high variance since this can indicate a change in expression depending on populations of cells. Adding labels using the <code>LabelPoints()</code> helps us understand which genes will be driving shape of our data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Identify the 15 most highly variable genes</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>ranked_variable_genes <span class="ot">&lt;-</span> <span class="fu">VariableFeatures</span>(seurat_phase)</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>top_genes <span class="ot">&lt;-</span> ranked_variable_genes[<span class="dv">1</span><span class="sc">:</span><span class="dv">15</span>]</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot the average expression and variance of these genes</span></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a><span class="co"># With labels to indicate which genes are in the top 15</span></span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">VariableFeaturePlot</span>(seurat_phase)</span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a><span class="fu">LabelPoints</span>(<span class="at">plot =</span> p, <span class="at">points =</span> top_genes, <span class="at">repel =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="06_SC_SCT_normalization_files/figure-html/unnamed-chunk-6-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Now, we can perform the PCA analysis and plot the first two principal components against each other. We also split the figure by cell cycle phase, to evaluate similarities and/or differences. <strong>We do not see large differences due to cell cycle phase. Based on this plot, we would not regress out the variation due to cell cycle.</strong></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Perform PCA</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>seurat_phase <span class="ot">&lt;-</span> <span class="fu">RunPCA</span>(seurat_phase)</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot the PCA colored by cell cycle phase</span></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(seurat_phase,</span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>        <span class="at">reduction =</span> <span class="st">"pca"</span>,</span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>        <span class="at">group.by=</span> <span class="st">"Phase"</span>,</span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>        <span class="at">split.by =</span> <span class="st">"Phase"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="06_SC_SCT_normalization_files/figure-html/unnamed-chunk-7-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<details>
<summary>
<b><i>When should cell cycle phase be regressed out?</i></b>
</summary>
<p><br>Below are two PCA plots taken from the Seurat vignette dealing with <a href="https://satijalab.org/seurat/archive/v3.1/cell_cycle_vignette.html">Cell-Cycle Scoring and Regression</a>.<br></p>
<ul>
<li>
This first plot is similar to what we plotted above, it is a PCA prior to regression to evaluate if the cell cycle is playing a big role in driving PC1 and PC2. Clearly, the cells are separating by cell type in this case, so the vignette suggests regressing out these effects.
</li>
</ul>
<p align="center">
<img src="../img/cell_cycle_not_regressed.png" width="400">
</p>
<ul>
<li>
This second PCA plot is <b>post-regression</b>, and displays how effective the regression was in removing the effect we observed.
</li>
</ul>
<p align="center">
<img src="../img/cell_cycle_regressed.png" width="400">
</p>
</details>
<hr>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise: Evaluating effects of mitochondrial expression
</div>
</div>
<div class="callout-body-container callout-body">
<p>Mitochondrial expression is another factor which can greatly influence clustering. Oftentimes, it is useful to regress out variation due to mitochondrial expression. However, if the differences in mitochondrial gene expression represent a biological phenomenon that may help to distinguish cell clusters, then we advise not regressing this out. In this exercise, we can perform a quick check similar to looking at cell cycle and decide whether or not we want to regress it out.</p>
<ol type="1">
<li>First, turn the mitochondrial ratio variable into a new categorical variable based on quartiles (using the code below):</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Check quartile values</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(seurat_phase<span class="sc">@</span>meta.data<span class="sc">$</span>mitoRatio)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
0.00000 0.01438 0.01993 0.02139 0.02669 0.14464 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Turn mitoRatio into categorical factor vector based on quartile values</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>seurat_phase<span class="sc">@</span>meta.data<span class="sc">$</span>mitoFr <span class="ot">&lt;-</span> <span class="fu">cut</span>(seurat_phase<span class="sc">@</span>meta.data<span class="sc">$</span>mitoRatio, </span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>                   <span class="at">breaks=</span><span class="fu">c</span>(<span class="sc">-</span><span class="cn">Inf</span>, <span class="fl">0.0144</span>, <span class="fl">0.0199</span>, <span class="fl">0.0267</span>, <span class="cn">Inf</span>), </span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>                   <span class="at">labels=</span><span class="fu">c</span>(<span class="st">"Low"</span>,<span class="st">"Medium"</span>,<span class="st">"Medium high"</span>, <span class="st">"High"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ol start="2" type="1">
<li>Next, plot the PCA similar to how we did with cell cycle regression. <em>Hint: use the new <code>mitoFr</code>variable to split cells and color them accordingly.</em></li>
<li>Evaluate the PCA plot generated in #2.
<ol type="1">
<li>Determine whether or not you observe an effect.</li>
<li>Describe what you see.</li>
<li>Would you regress out mitochondrial fraction as a source of unwanted variation?</li>
</ol></li>
</ol>
</div>
</div>
<hr>
</section>
</section>
<section id="normalization-and-regressing-out-sources-of-unwanted-variation-using-sctransform" class="level2">
<h2 class="anchored" data-anchor-id="normalization-and-regressing-out-sources-of-unwanted-variation-using-sctransform">Normalization and regressing out sources of unwanted variation using SCTransform</h2>
<p>In the <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1874-1">Hafemeister and Satija, 2019 paper</a> the authors explored the issues with simple transformations. Specifically they evaluated the standard log normalization approach and found that genes with different abundances are affected differently and that <strong>effective normalization (using the log transform) is only observed with low/medium abundance genes (Figure 1D, below)</strong>. Additionally, <strong>substantial imbalances in variance were observed with the log-normalized data (Figure 1E, below)</strong>. In particular, cells with low total UMI counts exhibited disproportionately higher variance for high-abundance genes, dampening the variance contribution from other gene abundances.&nbsp;</p>
<p align="center">
<img src="../img/SCT_Fig1.png" width="600">
</p>
<p><em><strong>Image credit:</strong> Hafemeister C and Satija R. Normalization and variance stabilization of single-cell RNA-seq data using regularized negative binomial regression, Genom Biology 2019 (https://doi.org/10.1101/576827)</em></p>
<p>The conclusion is, <strong>we cannot treat all genes the same.</strong></p>
<p>The proposed solution was the use of <strong>Pearson residuals for transformation</strong>, as implemented in Seurat’s <code>SCTransform</code> function. With this approach: * Measurements are multiplied by a gene-specific weight * Each gene is weighted based on how much evidence there is that it is non-uniformly expressed across cells * More evidence == more of a weight; Genes that are expressed in only a small fraction of cells will be favored (useful for finding rare cell populations) * Not just a consideration of the expression level is, but also the distribution of expression</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Why don’t we just run SCTransform to normalize?
</div>
</div>
<div class="callout-body-container callout-body">
<p>While the functions <code>NormalizeData</code>, <code>VariableFeatures</code> and <code>ScaleData</code> can be replaced by the function <code>SCTransform</code>, the latter uses a more sophisticated way to perform the normalization and scaling. We suggest using log normalization because it is good to observe the data and any trends using a simple transformation, as methods like SCT can alter the data in a way that is not as intuitive to interpret.</p>
</div>
</div>
<p>Now that we have established which effects are observed in our data, we can use the SCTransform method to regress out these effects. The <strong>SCTransform</strong> method was proposed as a better alternative to the log transform normalization method that we used for exploring sources of unwanted variation. The method not only <strong>normalizes data, but it also performs a variance stabilization and allows for additional covariates to be regressed out</strong>.</p>
<p>As described earlier, all genes cannot be treated the same. As such, the <strong>SCTransform method constructs a generalized linear model (GLM) for each gene</strong> with UMI counts as the response and sequencing depth as the explanatory variable. Information is pooled across genes with similar abundances, to regularize parameter estimates and <strong>obtain residuals which represent effectively normalized data values</strong> which are no longer correlated with sequencing depth.</p>
<p align="center">
<img src="../img/SCT_UMAP.png" width="400">
</p>
<p><em><strong>Image credit:</strong> Hafemeister C and Satija R. Normalization and variance stabilization of single-cell RNA-seq data using regularized negative binomial regression, Genom Biology 2019 (https://doi.org/10.1101/576827)</em></p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Since the UMI counts are part of the GLM, the effects are automatically regressed out. The user can include any additional covariates (<code>vars.to.regress</code>) that may have an effect on expression and will be included in the model.</p>
</div>
</div>
<p>To run the SCTransform we have the code below as an example. <strong>Do not run this code</strong>, as we prefer to run this for each sample separately in the next section below.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="do">## DO NOT RUN CODE ##</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a><span class="co"># SCTranform</span></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>seurat_phase <span class="ot">&lt;-</span> <span class="fu">SCTransform</span>(seurat_phase, <span class="at">vars.to.regress =</span> <span class="fu">c</span>(<span class="st">"mitoRatio"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="iterating-over-samples-in-a-dataset" class="level2">
<h2 class="anchored" data-anchor-id="iterating-over-samples-in-a-dataset">Iterating over samples in a dataset</h2>
<p>Since we have two samples in our dataset (from two conditions), we want to keep them as separate objects and transform them as that is what is required for integration. We will first split the cells in <code>seurat_phase</code> object into “Control” and “Stimulated”:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Split seurat object by condition to perform cell cycle scoring and SCT on all samples</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>split_seurat <span class="ot">&lt;-</span> <span class="fu">SplitObject</span>(seurat_phase, <span class="at">split.by =</span> <span class="st">"sample"</span>)</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>split_seurat</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$ctrl
An object of class Seurat 
14065 features across 14847 samples within 1 assay 
Active assay: RNA (14065 features, 2000 variable features)
 3 layers present: counts, data, scale.data
 1 dimensional reduction calculated: pca

$stim
An object of class Seurat 
14065 features across 14782 samples within 1 assay 
Active assay: RNA (14065 features, 2000 variable features)
 3 layers present: counts, data, scale.data
 1 dimensional reduction calculated: pca</code></pre>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>If you only wanted to integrate on a subset of your samples (e.g.&nbsp;all Ctrl replicates only), you could select which ones you wanted from the <code>split_seurat</code> object as shown below and move forward with those.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>ctrl_reps <span class="ot">&lt;-</span> split_seurat[<span class="fu">c</span>(<span class="st">"ctrl_1"</span>, <span class="st">"ctrl_2"</span>)]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
<p>Now we will <strong>use a ‘for loop’</strong> to run the <code>SCTransform()</code> on each sample, and regress out mitochondrial expression by specifying in the <code>vars.to.regress</code> argument of the <code>SCTransform()</code> function.</p>
<p>Before we run this <code>for loop</code>, we know that the output can generate large R objects/variables in terms of memory. If we have a large dataset, then we might need to <strong>adjust the limit for allowable object sizes within R</strong> (<em>Default is 500 </em> 1024 ^ 2 = 500 Mb*) using the following code:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">future.globals.maxSize =</span> <span class="dv">4000</span> <span class="sc">*</span> <span class="dv">1024</span><span class="sc">^</span><span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now, we run the following loop to <strong>perform the sctransform on all samples</strong>. This may take some time (~10 minutes):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(split_seurat)) {</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>    split_seurat[[i]] <span class="ot">&lt;-</span> <span class="fu">SCTransform</span>(split_seurat[[i]], </span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">vars.to.regress =</span> <span class="fu">c</span>(<span class="st">"mitoRatio"</span>),</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">vst.flavor =</span> <span class="st">"v2"</span>)</span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>    }</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Please note that in the for loop above, we specify that <code>vst.flavor = "v2"</code> to use the updated version of SCT. “v2” was introduced in early 2022, and is now commonly used. This update improves: * Speed and memory consumption * The stability of parameter estimates * Variable feature identification in subsequent steps</p>
<p>For more information, please see the <a href="https://satijalab.org/seurat/articles/sctransform_v2_vignette.html">Seurat vignette’s section on SCTransform, v2 regularization</a>.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>By default, after normalizing, adjusting the variance, and regressing out uninteresting sources of variation, SCTransform will rank the genes by residual variance and output the 3000 most variant genes. If the dataset has larger cell numbers, then it may be beneficial to adjust this parameter higher using the <code>variable.features.n</code> argument._</p>
</div>
</div>
<p>Note, the last line of output specifies <strong>“Set default assay to SCT”</strong>. This specifies that moving forward we would like to use the data after SCT was implemented. We can view the different assays that we have stored in our seurat object.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Check which assays are stored in objects</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>split_seurat<span class="sc">$</span>ctrl<span class="sc">@</span>assays</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$RNA
Assay (v5) data with 14065 features for 14847 cells
Top 10 variable features:
 HBB, HBA2, CCL4L2, HBA1, IGKC, CCL7, PPBP, CCL4, CCL3, CCL8 
Layers:
 counts, data, scale.data 

$SCT
SCTAssay data with 13799 features for 14847 cells, and 1 SCTModel(s) 
Top 10 variable features:
 FTL, CCL2, IGKC, GNLY, IGLC2, CCL3, TIMP1, IGHM, CCL4, PPBP </code></pre>
</div>
</div>
<p>Now we can see that in addition to the raw RNA counts, we now have a SCT component in our <code>assays</code> slot. The most variable features will be the only genes stored inside the SCT assay. As we move through the scRNA-seq analysis, we will choose the most appropriate assay to use for the different steps in the analysis.</p>
<hr>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercises
</div>
</div>
<div class="callout-body-container callout-body">
<ol start="4" type="1">
<li>Are the same assays available for the “stim” samples within the <code>split_seurat</code> object? What is the code you used to check that?</li>
<li>Any observations for the genes or features listed under <em>“First 10 features:”</em> and the <em>“Top 10 variable features:”</em> for “ctrl” versus “stim”?</li>
</ol>
</div>
</div>
<hr>
<section id="save-the-object" class="level3">
<h3 class="anchored" data-anchor-id="save-the-object">Save the object!</h3>
<p>Before finishing up, let’s save this object to the <code>data/</code> folder. It can take a while to get back to this stage especially when working with large datasets, it is best practice to save the object as an easily loadable file locally.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Save the split seurat object</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(split_seurat, <span class="st">"../data/split_seurat.rds"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>To load the <code>.rds</code> file back into your environment you would use the llowing code:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load the split seurat object into the environment</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>split_seurat <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"../data/split_seurat.rds"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>


<!-- -->

</section>
</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb21" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a><span class="an">title:</span><span class="co"> "Single-cell RNA-seq: Normalization, identification of most variable genes"</span></span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a><span class="an">author:</span><span class="co"> "Mary Piper, Meeta Mistry, Radhika Khetani, Jihe Liu"</span></span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a><span class="an">date:</span><span class="co"> November 30, 2022</span></span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a>Approximate time: 90 minutes</span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-11"><a href="#cb21-11" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb21-12"><a href="#cb21-12" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb21-13"><a href="#cb21-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-14"><a href="#cb21-14" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb21-15"><a href="#cb21-15" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb21-16"><a href="#cb21-16" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="st">"../data/seurat_filtered.RData"</span>)</span>
<span id="cb21-17"><a href="#cb21-17" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb21-18"><a href="#cb21-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-19"><a href="#cb21-19" aria-hidden="true" tabindex="-1"></a><span class="fu">## Learning Objectives:</span></span>
<span id="cb21-20"><a href="#cb21-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-21"><a href="#cb21-21" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>Discuss why normalizing counts is necessary for accurate comparison between cells</span>
<span id="cb21-22"><a href="#cb21-22" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>Describe different normalization approaches</span>
<span id="cb21-23"><a href="#cb21-23" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>Evaluate the effects from any unwanted sources of variation and correct for them</span>
<span id="cb21-24"><a href="#cb21-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-25"><a href="#cb21-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-26"><a href="#cb21-26" aria-hidden="true" tabindex="-1"></a><span class="fu"># Single-cell RNA-seq: Normalization and regressing out unwanted variation</span></span>
<span id="cb21-27"><a href="#cb21-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-28"><a href="#cb21-28" aria-hidden="true" tabindex="-1"></a>Now that we have our high quality cells, we can explore our data and see if we are able to identify any sources of unwanted variation. Depending on what we observe, we will utilize that information when performing variance stabilization using SCTransform but also to regress out the effects of any covariates that have an effect on our data.</span>
<span id="cb21-29"><a href="#cb21-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-30"><a href="#cb21-30" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;p</span> <span class="er">align</span><span class="ot">=</span><span class="st">"center"</span><span class="kw">&gt;</span></span>
<span id="cb21-31"><a href="#cb21-31" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;img</span> <span class="er">src</span><span class="ot">=</span><span class="st">"../img/sc_workflow_2022.jpg"</span> <span class="er">width</span><span class="ot">=</span><span class="st">"630"</span><span class="kw">&gt;</span></span>
<span id="cb21-32"><a href="#cb21-32" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;/p&gt;</span></span>
<span id="cb21-33"><a href="#cb21-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-34"><a href="#cb21-34" aria-hidden="true" tabindex="-1"></a>***</span>
<span id="cb21-35"><a href="#cb21-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-36"><a href="#cb21-36" aria-hidden="true" tabindex="-1"></a>_**Goals:**_ </span>
<span id="cb21-37"><a href="#cb21-37" aria-hidden="true" tabindex="-1"></a> </span>
<span id="cb21-38"><a href="#cb21-38" aria-hidden="true" tabindex="-1"></a><span class="ss"> - </span>_To accurately **normalize the gene expression values** to account for differences in sequencing depth and overdispersed count values._</span>
<span id="cb21-39"><a href="#cb21-39" aria-hidden="true" tabindex="-1"></a><span class="ss"> - </span>_To **identify the most variant genes** likely to be indicative of the different cell types present._</span>
<span id="cb21-40"><a href="#cb21-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-41"><a href="#cb21-41" aria-hidden="true" tabindex="-1"></a>_**Challenges:**_</span>
<span id="cb21-42"><a href="#cb21-42" aria-hidden="true" tabindex="-1"></a> </span>
<span id="cb21-43"><a href="#cb21-43" aria-hidden="true" tabindex="-1"></a><span class="ss"> - </span>_**Checking and removing unwanted variation** so that we do not have cells clustering by artifacts downstream_</span>
<span id="cb21-44"><a href="#cb21-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-45"><a href="#cb21-45" aria-hidden="true" tabindex="-1"></a>_**Recommendations:**_</span>
<span id="cb21-46"><a href="#cb21-46" aria-hidden="true" tabindex="-1"></a> </span>
<span id="cb21-47"><a href="#cb21-47" aria-hidden="true" tabindex="-1"></a><span class="ss"> - </span>_Have a good idea of your expectations for the **cell types to be present** prior to performing the clustering. Know whether you expect cell types of low complexity or higher mitochondrial content AND whether the cells are differentiating_</span>
<span id="cb21-48"><a href="#cb21-48" aria-hidden="true" tabindex="-1"></a><span class="ss"> - </span>_**Regress out** number of UMIs (default using sctransform), mitochondrial content, and cell cycle, if needed and appropriate for experiment, so not to drive clustering downstream_</span>
<span id="cb21-49"><a href="#cb21-49" aria-hidden="true" tabindex="-1"></a> </span>
<span id="cb21-50"><a href="#cb21-50" aria-hidden="true" tabindex="-1"></a>***</span>
<span id="cb21-51"><a href="#cb21-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-52"><a href="#cb21-52" aria-hidden="true" tabindex="-1"></a><span class="fu">## Normalization</span></span>
<span id="cb21-53"><a href="#cb21-53" aria-hidden="true" tabindex="-1"></a>An essential first step in the majority of mRNA expression analyses is normalization, whereby systematic variations are adjusted for to **make expression counts comparable across genes and/or samples**. The counts of mapped reads for each gene is proportional to the expression of RNA ("interesting") in addition to many other factors ("uninteresting"). Normalization is the process of adjusting raw count values to account for the "uninteresting" factors. </span>
<span id="cb21-54"><a href="#cb21-54" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-55"><a href="#cb21-55" aria-hidden="true" tabindex="-1"></a>The main factors often considered during normalization are:</span>
<span id="cb21-56"><a href="#cb21-56" aria-hidden="true" tabindex="-1"></a> </span>
<span id="cb21-57"><a href="#cb21-57" aria-hidden="true" tabindex="-1"></a><span class="ss"> - </span>**Sequencing depth:** Accounting for sequencing depth is necessary for comparison of gene expression between cells. In the example below, each gene appears to have doubled in expression in cell 2, however this is a consequence of cell 2 having twice the sequencing depth.</span>
<span id="cb21-58"><a href="#cb21-58" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-59"><a href="#cb21-59" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;p</span> <span class="er">align</span><span class="ot">=</span><span class="st">"center"</span><span class="kw">&gt;</span></span>
<span id="cb21-60"><a href="#cb21-60" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;img</span> <span class="er">src</span><span class="ot">=</span><span class="st">"../img/sequencing_depth.png"</span> <span class="er">width</span><span class="ot">=</span><span class="st">"400"</span><span class="kw">&gt;</span></span>
<span id="cb21-61"><a href="#cb21-61" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;/p&gt;</span></span>
<span id="cb21-62"><a href="#cb21-62" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-63"><a href="#cb21-63" aria-hidden="true" tabindex="-1"></a>Each cell in scRNA-seq will have a differing number of reads associated with it. So to accurately compare expression between cells, it is necessary to normalize for sequencing depth.</span>
<span id="cb21-64"><a href="#cb21-64" aria-hidden="true" tabindex="-1"></a> </span>
<span id="cb21-65"><a href="#cb21-65" aria-hidden="true" tabindex="-1"></a><span class="ss"> - </span>**Gene length:** Accounting for gene length is necessary for comparing expression between different genes within the same cell. The number of reads mapped to a longer gene can appear to have equal count/expression as a shorter gene that is more highly expressed. </span>
<span id="cb21-66"><a href="#cb21-66" aria-hidden="true" tabindex="-1"></a> </span>
<span id="cb21-67"><a href="#cb21-67" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;p</span> <span class="er">align</span><span class="ot">=</span><span class="st">"center"</span><span class="kw">&gt;</span> </span>
<span id="cb21-68"><a href="#cb21-68" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;img</span> <span class="er">src</span><span class="ot">=</span><span class="st">"../img/length_of_gene.png"</span> <span class="er">width</span><span class="ot">=</span><span class="st">"400"</span><span class="kw">&gt;</span></span>
<span id="cb21-69"><a href="#cb21-69" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;/p&gt;</span></span>
<span id="cb21-70"><a href="#cb21-70" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-71"><a href="#cb21-71" aria-hidden="true" tabindex="-1"></a>::: callout-note</span>
<span id="cb21-72"><a href="#cb21-72" aria-hidden="true" tabindex="-1"></a>If using a 3' or 5' droplet-based method, the length of the gene will not affect the analysis because only the 5' or 3' end of the transcript is sequenced.* However, if using full-length sequencing, the transcript length should be accounted for.</span>
<span id="cb21-73"><a href="#cb21-73" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb21-74"><a href="#cb21-74" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-75"><a href="#cb21-75" aria-hidden="true" tabindex="-1"></a><span class="fu">### Methods for scRNA-seq normalization</span></span>
<span id="cb21-76"><a href="#cb21-76" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-77"><a href="#cb21-77" aria-hidden="true" tabindex="-1"></a>Various methods have been developed specifically for scRNA-seq normalization. Some **simpler methods resemble what we have seen with bulk RNA-seq**; the application of **global scale factors** adjusting for a count-depth relationship that is assumed common across all genes. However, if those assumptions are not true then this basic normalization can lead to over-correction for lowly and moderately expressed genes and, in some cases, under-normalization of highly expressed genes ([Bacher R et al, 2017](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5473255/)). **More complex methods will apply correction on a per-gene basis.** In this lesson we will explore both approaches.</span>
<span id="cb21-78"><a href="#cb21-78" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-79"><a href="#cb21-79" aria-hidden="true" tabindex="-1"></a>Regardless of which method is used for normalization, it can be helpful to **think of it as a two-step process** (even though it is often described as a single step in most papers). The first is a scaling step and the second is a transformation.</span>
<span id="cb21-80"><a href="#cb21-80" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-81"><a href="#cb21-81" aria-hidden="true" tabindex="-1"></a>**1. Scaling**</span>
<span id="cb21-82"><a href="#cb21-82" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-83"><a href="#cb21-83" aria-hidden="true" tabindex="-1"></a>The first step in normalization is to **multiply each UMI count by a cell specific factor to get all cells to have the same UMI counts**. Why would we want to do this? Different cells have different amounts of mRNA; this could be due to differences between cell types or variation within the same cell type depending on how well the chemistry worked in one drop versus another. In either case, we are not interested in comparing these absolute counts between cells. Instead we are interested in comparing concentrations, and scaling helps achieve this.</span>
<span id="cb21-84"><a href="#cb21-84" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-85"><a href="#cb21-85" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-86"><a href="#cb21-86" aria-hidden="true" tabindex="-1"></a>**2. Transformation**</span>
<span id="cb21-87"><a href="#cb21-87" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-88"><a href="#cb21-88" aria-hidden="true" tabindex="-1"></a>The next step is a transformation, and it is at this step where we can distinguish the simpler versus complex methods as mentioned above.</span>
<span id="cb21-89"><a href="#cb21-89" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-90"><a href="#cb21-90" aria-hidden="true" tabindex="-1"></a>**Simple transformations** are those which apply the same function to each individual measurement. Common examples include a **log transform** (which is applied in the original Seurat workflow), or a square root transform (less commonly used).</span>
<span id="cb21-91"><a href="#cb21-91" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-92"><a href="#cb21-92" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-93"><a href="#cb21-93" aria-hidden="true" tabindex="-1"></a><span class="fu">##  Explore sources of unwanted variation</span></span>
<span id="cb21-94"><a href="#cb21-94" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-95"><a href="#cb21-95" aria-hidden="true" tabindex="-1"></a>The most common biological data correction (or source of "uninteresting" variation) in single cell RNA-seq is the effects of the cell cycle on the transcriptome. We need to explore the data and see if we observe any effects in our data.   </span>
<span id="cb21-96"><a href="#cb21-96" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-97"><a href="#cb21-97" aria-hidden="true" tabindex="-1"></a><span class="fu">### Set-up</span></span>
<span id="cb21-98"><a href="#cb21-98" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-99"><a href="#cb21-99" aria-hidden="true" tabindex="-1"></a>Let's start by creating a new script for the normalization and integration steps. Create a new script (File -&gt; New File -&gt; R script), and save it as <span class="in">`SCT_integration_analysis.R`</span>.</span>
<span id="cb21-100"><a href="#cb21-100" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-101"><a href="#cb21-101" aria-hidden="true" tabindex="-1"></a>For the remainder of the workflow we will be mainly using functions available in the Seurat package. Therefore, we need to load the Seurat library in addition to the tidyverse library and a few others listed below. </span>
<span id="cb21-102"><a href="#cb21-102" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-105"><a href="#cb21-105" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb21-106"><a href="#cb21-106" aria-hidden="true" tabindex="-1"></a><span class="co"># Single-cell RNA-seq - normalization</span></span>
<span id="cb21-107"><a href="#cb21-107" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-108"><a href="#cb21-108" aria-hidden="true" tabindex="-1"></a><span class="co"># Load libraries</span></span>
<span id="cb21-109"><a href="#cb21-109" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb21-110"><a href="#cb21-110" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb21-111"><a href="#cb21-111" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(RCurl)</span>
<span id="cb21-112"><a href="#cb21-112" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(cowplot)</span>
<span id="cb21-113"><a href="#cb21-113" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb21-114"><a href="#cb21-114" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-115"><a href="#cb21-115" aria-hidden="true" tabindex="-1"></a>Before we make any comparisons across cells, we will **apply a simple normalization.** This is solely for the purpose of exploring the sources of variation in our data.</span>
<span id="cb21-116"><a href="#cb21-116" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-117"><a href="#cb21-117" aria-hidden="true" tabindex="-1"></a>The input for this analysis is a <span class="in">`seurat`</span> object. We will use the one that we created in the QC lesson called <span class="in">`filtered_seurat`</span>.</span>
<span id="cb21-118"><a href="#cb21-118" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-121"><a href="#cb21-121" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb21-122"><a href="#cb21-122" aria-hidden="true" tabindex="-1"></a><span class="co"># Normalize the counts</span></span>
<span id="cb21-123"><a href="#cb21-123" aria-hidden="true" tabindex="-1"></a>seurat_phase <span class="ot">&lt;-</span> <span class="fu">NormalizeData</span>(filtered_seurat)</span>
<span id="cb21-124"><a href="#cb21-124" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb21-125"><a href="#cb21-125" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-126"><a href="#cb21-126" aria-hidden="true" tabindex="-1"></a>Next, we take this normalized data and check to see if data correction methods are necessary. </span>
<span id="cb21-127"><a href="#cb21-127" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-128"><a href="#cb21-128" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-129"><a href="#cb21-129" aria-hidden="true" tabindex="-1"></a><span class="fu">### Evaluating effects of cell cycle </span></span>
<span id="cb21-130"><a href="#cb21-130" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-131"><a href="#cb21-131" aria-hidden="true" tabindex="-1"></a>To assign each cell a score based on its expression of G2/M and S phase markers, we can use the Seuart function <span class="in">`CellCycleScoring()`</span>. This function calculates cell cycle phase scores based on canonical markers that required as input.</span>
<span id="cb21-132"><a href="#cb21-132" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-133"><a href="#cb21-133" aria-hidden="true" tabindex="-1"></a>We have provided a list of human cell cycle markers for you in the <span class="in">`data`</span> folder as an Rdata file called <span class="in">`cycle.rda`</span>. However, if you are not working with human data we have <span class="co">[</span><span class="ot">additional materials</span><span class="co">](cell_cycle_scoring.md)</span> detailing how to acquire cell cycle markers for other organisms of interest.</span>
<span id="cb21-134"><a href="#cb21-134" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-137"><a href="#cb21-137" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb21-138"><a href="#cb21-138" aria-hidden="true" tabindex="-1"></a><span class="co"># Load cell cycle markers</span></span>
<span id="cb21-139"><a href="#cb21-139" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="st">"../data/cycle.rda"</span>)</span>
<span id="cb21-140"><a href="#cb21-140" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-141"><a href="#cb21-141" aria-hidden="true" tabindex="-1"></a><span class="co"># Score cells for cell cycle</span></span>
<span id="cb21-142"><a href="#cb21-142" aria-hidden="true" tabindex="-1"></a>seurat_phase <span class="ot">&lt;-</span> <span class="fu">CellCycleScoring</span>(seurat_phase, </span>
<span id="cb21-143"><a href="#cb21-143" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">g2m.features =</span> g2m_genes, </span>
<span id="cb21-144"><a href="#cb21-144" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">s.features =</span> s_genes)</span>
<span id="cb21-145"><a href="#cb21-145" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-146"><a href="#cb21-146" aria-hidden="true" tabindex="-1"></a><span class="co"># View cell cycle scores and phases assigned to cells</span></span>
<span id="cb21-147"><a href="#cb21-147" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(seurat_phase<span class="sc">@</span>meta.data)                                </span>
<span id="cb21-148"><a href="#cb21-148" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb21-149"><a href="#cb21-149" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-150"><a href="#cb21-150" aria-hidden="true" tabindex="-1"></a>After scoring the cells for cell cycle, we would like to **determine whether cell cycle is a major source of variation in our dataset using PCA**. </span>
<span id="cb21-151"><a href="#cb21-151" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-152"><a href="#cb21-152" aria-hidden="true" tabindex="-1"></a><span class="fu">### PCA</span></span>
<span id="cb21-153"><a href="#cb21-153" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-154"><a href="#cb21-154" aria-hidden="true" tabindex="-1"></a>Principal Component Analysis (PCA) is a technique used to emphasize variation as well as similarity, and to bring out strong patterns in a dataset; it is one of the methods used for *"dimensionality reduction"*. </span>
<span id="cb21-155"><a href="#cb21-155" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-156"><a href="#cb21-156" aria-hidden="true" tabindex="-1"></a>::: callout-note</span>
<span id="cb21-157"><a href="#cb21-157" aria-hidden="true" tabindex="-1"></a>For a more **detailed explanation on PCA**, please <span class="co">[</span><span class="ot">look over this lesson</span><span class="co">](05_theory_of_PCA.qmd)</span> (adapted from StatQuests/Josh Starmer's YouTube video). We also strongly encourage you to explore the video <span class="co">[</span><span class="ot">StatQuest's video</span><span class="co">](https://www.youtube.com/watch?v=_UVHneBUBW0)</span> for a more thorough understanding. </span>
<span id="cb21-158"><a href="#cb21-158" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb21-159"><a href="#cb21-159" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-160"><a href="#cb21-160" aria-hidden="true" tabindex="-1"></a>Let's say you are working with a single-cell RNA-seq dataset with *12,000 cells* and you have quantified the expression of *20,000 genes*. The schematic below demonstrates how you would go from a cell x gene matrix to principal component (PC) scores for each inividual cell.</span>
<span id="cb21-161"><a href="#cb21-161" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-162"><a href="#cb21-162" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;p</span> <span class="er">align</span><span class="ot">=</span><span class="st">"center"</span><span class="kw">&gt;</span></span>
<span id="cb21-163"><a href="#cb21-163" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;img</span> <span class="er">src</span><span class="ot">=</span><span class="st">"../img/PCA_scrnaseq_1.png"</span> <span class="er">width</span><span class="ot">=</span><span class="st">"900"</span><span class="kw">&gt;</span></span>
<span id="cb21-164"><a href="#cb21-164" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;/p&gt;</span></span>
<span id="cb21-165"><a href="#cb21-165" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-166"><a href="#cb21-166" aria-hidden="true" tabindex="-1"></a>After the PC scores have been calculated, you are looking at a matrix of 12,000 x 12,000 that represents the information about relative gene expression in all the cells. You can select the PC1 and PC2 columns and plot that in a 2D way.</span>
<span id="cb21-167"><a href="#cb21-167" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-168"><a href="#cb21-168" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;p</span> <span class="er">align</span><span class="ot">=</span><span class="st">"center"</span><span class="kw">&gt;</span></span>
<span id="cb21-169"><a href="#cb21-169" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;img</span> <span class="er">src</span><span class="ot">=</span><span class="st">"../img/PCA_scrnaseq_2.png"</span> <span class="er">width</span><span class="ot">=</span><span class="st">"600"</span><span class="kw">&gt;</span></span>
<span id="cb21-170"><a href="#cb21-170" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;/p&gt;</span></span>
<span id="cb21-171"><a href="#cb21-171" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-172"><a href="#cb21-172" aria-hidden="true" tabindex="-1"></a>You can also use the PC scores from the first 40 PCs for downstream analysis like clustering, marker identification etc., since these represent the majority of the variation in the data. We will be talking a lot more about this later in this workshop.</span>
<span id="cb21-173"><a href="#cb21-173" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-174"><a href="#cb21-174" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;p</span> <span class="er">align</span><span class="ot">=</span><span class="st">"center"</span><span class="kw">&gt;</span></span>
<span id="cb21-175"><a href="#cb21-175" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;img</span> <span class="er">src</span><span class="ot">=</span><span class="st">"../img/PCA_scrnaseq_3.png"</span> <span class="er">width</span><span class="ot">=</span><span class="st">"600"</span><span class="kw">&gt;</span></span>
<span id="cb21-176"><a href="#cb21-176" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;/p&gt;</span></span>
<span id="cb21-177"><a href="#cb21-177" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-178"><a href="#cb21-178" aria-hidden="true" tabindex="-1"></a>::: callout-note</span>
<span id="cb21-179"><a href="#cb21-179" aria-hidden="true" tabindex="-1"></a>For datasets with a larger number of cells, only the PC1 and PC2 scores for each cell are usually plotted, or used for visualization. Since these PCs explain the most variation in the dataset, the expectation is that the cells that are more similar to each other will cluster together with PC1 and PC2.</span>
<span id="cb21-180"><a href="#cb21-180" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb21-181"><a href="#cb21-181" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-182"><a href="#cb21-182" aria-hidden="true" tabindex="-1"></a><span class="fu">### Using PCA to evaluate the effects of cell cycle</span></span>
<span id="cb21-183"><a href="#cb21-183" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-184"><a href="#cb21-184" aria-hidden="true" tabindex="-1"></a>To perform PCA, we need to **first choose the most variable features, then scale the data**. Since highly expressed genes exhibit the highest amount of variation and we don't want our 'highly variable genes' only to reflect high expression, we need to scale the data to scale variation with expression level. The Seurat <span class="in">`ScaleData()`</span> function will scale the data by:</span>
<span id="cb21-185"><a href="#cb21-185" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-186"><a href="#cb21-186" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>Adjusting the expression of each gene to give a mean expression across cells to be 0</span>
<span id="cb21-187"><a href="#cb21-187" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>Scaling expression of each gene to give a variance across cells to be 1</span>
<span id="cb21-188"><a href="#cb21-188" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-191"><a href="#cb21-191" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb21-192"><a href="#cb21-192" aria-hidden="true" tabindex="-1"></a><span class="co"># Identify the most variable genes</span></span>
<span id="cb21-193"><a href="#cb21-193" aria-hidden="true" tabindex="-1"></a>seurat_phase <span class="ot">&lt;-</span> <span class="fu">FindVariableFeatures</span>(seurat_phase, </span>
<span id="cb21-194"><a href="#cb21-194" aria-hidden="true" tabindex="-1"></a>                     <span class="at">selection.method =</span> <span class="st">"vst"</span>,</span>
<span id="cb21-195"><a href="#cb21-195" aria-hidden="true" tabindex="-1"></a>                     <span class="at">nfeatures =</span> <span class="dv">2000</span>, </span>
<span id="cb21-196"><a href="#cb21-196" aria-hidden="true" tabindex="-1"></a>                     <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb21-197"><a href="#cb21-197" aria-hidden="true" tabindex="-1"></a>             </span>
<span id="cb21-198"><a href="#cb21-198" aria-hidden="true" tabindex="-1"></a><span class="co"># Scale the counts</span></span>
<span id="cb21-199"><a href="#cb21-199" aria-hidden="true" tabindex="-1"></a>seurat_phase <span class="ot">&lt;-</span> <span class="fu">ScaleData</span>(seurat_phase)</span>
<span id="cb21-200"><a href="#cb21-200" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb21-201"><a href="#cb21-201" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-202"><a href="#cb21-202" aria-hidden="true" tabindex="-1"></a>::: callout-note</span>
<span id="cb21-203"><a href="#cb21-203" aria-hidden="true" tabindex="-1"></a>For the <span class="in">`selection.method`</span> and <span class="in">`nfeatures`</span> arguments the values specified are the default settings. Therefore, you do not necessarily need to include these in your code. We have included it here for transparency and inform you what you are using._ </span>
<span id="cb21-204"><a href="#cb21-204" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb21-205"><a href="#cb21-205" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-206"><a href="#cb21-206" aria-hidden="true" tabindex="-1"></a>Highly variable gene selection is extremely important since many downstream steps are computed only on these genes. Seurat allows us to access the ranked highly variable genes with the <span class="in">`VariableFeatures()`</span> function. We can additionally visualize the dispersion of all genes using Seurat's <span class="in">`VariableFeaturePlot()`</span>, which shows a gene's average expression across all cells on the x-axis and variance on the y-axis. Ideally we want to use genes that have high variance since this can indicate a change in expression depending on populations of cells. Adding labels using the <span class="in">`LabelPoints()`</span> helps us understand which genes will be driving shape of our data.</span>
<span id="cb21-207"><a href="#cb21-207" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-210"><a href="#cb21-210" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb21-211"><a href="#cb21-211" aria-hidden="true" tabindex="-1"></a><span class="co"># Identify the 15 most highly variable genes</span></span>
<span id="cb21-212"><a href="#cb21-212" aria-hidden="true" tabindex="-1"></a>ranked_variable_genes <span class="ot">&lt;-</span> <span class="fu">VariableFeatures</span>(seurat_phase)</span>
<span id="cb21-213"><a href="#cb21-213" aria-hidden="true" tabindex="-1"></a>top_genes <span class="ot">&lt;-</span> ranked_variable_genes[<span class="dv">1</span><span class="sc">:</span><span class="dv">15</span>]</span>
<span id="cb21-214"><a href="#cb21-214" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-215"><a href="#cb21-215" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot the average expression and variance of these genes</span></span>
<span id="cb21-216"><a href="#cb21-216" aria-hidden="true" tabindex="-1"></a><span class="co"># With labels to indicate which genes are in the top 15</span></span>
<span id="cb21-217"><a href="#cb21-217" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">VariableFeaturePlot</span>(seurat_phase)</span>
<span id="cb21-218"><a href="#cb21-218" aria-hidden="true" tabindex="-1"></a><span class="fu">LabelPoints</span>(<span class="at">plot =</span> p, <span class="at">points =</span> top_genes, <span class="at">repel =</span> <span class="cn">TRUE</span>)</span>
<span id="cb21-219"><a href="#cb21-219" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb21-220"><a href="#cb21-220" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-221"><a href="#cb21-221" aria-hidden="true" tabindex="-1"></a>Now, we can perform the PCA analysis and plot the first two principal components against each other. We also split the figure by cell cycle phase, to evaluate similarities and/or differences. **We do not see large differences due to cell cycle phase. Based on this plot, we would not regress out the variation due to cell cycle.**</span>
<span id="cb21-222"><a href="#cb21-222" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-225"><a href="#cb21-225" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb21-226"><a href="#cb21-226" aria-hidden="true" tabindex="-1"></a><span class="co"># Perform PCA</span></span>
<span id="cb21-227"><a href="#cb21-227" aria-hidden="true" tabindex="-1"></a>seurat_phase <span class="ot">&lt;-</span> <span class="fu">RunPCA</span>(seurat_phase)</span>
<span id="cb21-228"><a href="#cb21-228" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-229"><a href="#cb21-229" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot the PCA colored by cell cycle phase</span></span>
<span id="cb21-230"><a href="#cb21-230" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(seurat_phase,</span>
<span id="cb21-231"><a href="#cb21-231" aria-hidden="true" tabindex="-1"></a>        <span class="at">reduction =</span> <span class="st">"pca"</span>,</span>
<span id="cb21-232"><a href="#cb21-232" aria-hidden="true" tabindex="-1"></a>        <span class="at">group.by=</span> <span class="st">"Phase"</span>,</span>
<span id="cb21-233"><a href="#cb21-233" aria-hidden="true" tabindex="-1"></a>        <span class="at">split.by =</span> <span class="st">"Phase"</span>)</span>
<span id="cb21-234"><a href="#cb21-234" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb21-235"><a href="#cb21-235" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-236"><a href="#cb21-236" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-237"><a href="#cb21-237" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;details&gt;</span></span>
<span id="cb21-238"><a href="#cb21-238" aria-hidden="true" tabindex="-1"></a>    <span class="kw">&lt;summary&gt;&lt;b&gt;&lt;i&gt;</span>When should cell cycle phase be regressed out?<span class="kw">&lt;/i&gt;&lt;/b&gt;&lt;/summary&gt;</span></span>
<span id="cb21-239"><a href="#cb21-239" aria-hidden="true" tabindex="-1"></a>    <span class="kw">&lt;br&gt;</span>Below are two PCA plots taken from the Seurat vignette dealing with <span class="kw">&lt;a</span> <span class="er">href</span><span class="ot">=</span><span class="st">"https://satijalab.org/seurat/archive/v3.1/cell_cycle_vignette.html"</span><span class="kw">&gt;</span>Cell-Cycle Scoring and Regression<span class="kw">&lt;/a&gt;</span>.<span class="kw">&lt;br&gt;</span></span>
<span id="cb21-240"><a href="#cb21-240" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-241"><a href="#cb21-241" aria-hidden="true" tabindex="-1"></a><span class="in">    &lt;ul&gt;&lt;li&gt;This first plot is similar to what we plotted above, it is a PCA prior to regression to evaluate if the cell cycle is playing a big role in driving PC1 and PC2. Clearly, the cells are separating by cell type in this case, so the vignette suggests regressing out these effects.&lt;/li&gt;&lt;/ul&gt;</span></span>
<span id="cb21-242"><a href="#cb21-242" aria-hidden="true" tabindex="-1"></a><span class="in">    &lt;p align="center"&gt;</span></span>
<span id="cb21-243"><a href="#cb21-243" aria-hidden="true" tabindex="-1"></a><span class="in">    &lt;img src="../img/cell_cycle_not_regressed.png" width="400"&gt;</span></span>
<span id="cb21-244"><a href="#cb21-244" aria-hidden="true" tabindex="-1"></a><span class="in">    &lt;/p&gt;</span></span>
<span id="cb21-245"><a href="#cb21-245" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-246"><a href="#cb21-246" aria-hidden="true" tabindex="-1"></a><span class="in">    &lt;ul&gt;&lt;li&gt;This second PCA plot is &lt;b&gt;post-regression&lt;/b&gt;, and displays how effective the regression was in removing the effect we observed.&lt;/li&gt;&lt;/ul&gt;</span></span>
<span id="cb21-247"><a href="#cb21-247" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-248"><a href="#cb21-248" aria-hidden="true" tabindex="-1"></a><span class="in">    &lt;p align="center"&gt;</span></span>
<span id="cb21-249"><a href="#cb21-249" aria-hidden="true" tabindex="-1"></a><span class="in">    &lt;img src="../img/cell_cycle_regressed.png" width="400"&gt;</span></span>
<span id="cb21-250"><a href="#cb21-250" aria-hidden="true" tabindex="-1"></a><span class="in">    &lt;/p&gt;</span></span>
<span id="cb21-251"><a href="#cb21-251" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;/details&gt;</span></span>
<span id="cb21-252"><a href="#cb21-252" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-253"><a href="#cb21-253" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-254"><a href="#cb21-254" aria-hidden="true" tabindex="-1"></a>***</span>
<span id="cb21-255"><a href="#cb21-255" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-256"><a href="#cb21-256" aria-hidden="true" tabindex="-1"></a>::: callout-tip</span>
<span id="cb21-257"><a href="#cb21-257" aria-hidden="true" tabindex="-1"></a><span class="fu"># Exercise: Evaluating effects of mitochondrial expression</span></span>
<span id="cb21-258"><a href="#cb21-258" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-259"><a href="#cb21-259" aria-hidden="true" tabindex="-1"></a>Mitochondrial expression is another factor which can greatly influence clustering. Oftentimes, it is useful to regress out variation due to mitochondrial expression. However, if the differences in mitochondrial gene expression represent a biological phenomenon that may help to distinguish cell clusters, then we advise not regressing this out. In this exercise, we can perform a quick check similar to looking at cell cycle and decide whether or not we want to regress it out.</span>
<span id="cb21-260"><a href="#cb21-260" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-261"><a href="#cb21-261" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span>First, turn the mitochondrial ratio variable into a new categorical variable based on quartiles (using the code below):</span>
<span id="cb21-262"><a href="#cb21-262" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-265"><a href="#cb21-265" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb21-266"><a href="#cb21-266" aria-hidden="true" tabindex="-1"></a><span class="co"># Check quartile values</span></span>
<span id="cb21-267"><a href="#cb21-267" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(seurat_phase<span class="sc">@</span>meta.data<span class="sc">$</span>mitoRatio)</span>
<span id="cb21-268"><a href="#cb21-268" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-269"><a href="#cb21-269" aria-hidden="true" tabindex="-1"></a><span class="co"># Turn mitoRatio into categorical factor vector based on quartile values</span></span>
<span id="cb21-270"><a href="#cb21-270" aria-hidden="true" tabindex="-1"></a>seurat_phase<span class="sc">@</span>meta.data<span class="sc">$</span>mitoFr <span class="ot">&lt;-</span> <span class="fu">cut</span>(seurat_phase<span class="sc">@</span>meta.data<span class="sc">$</span>mitoRatio, </span>
<span id="cb21-271"><a href="#cb21-271" aria-hidden="true" tabindex="-1"></a>                   <span class="at">breaks=</span><span class="fu">c</span>(<span class="sc">-</span><span class="cn">Inf</span>, <span class="fl">0.0144</span>, <span class="fl">0.0199</span>, <span class="fl">0.0267</span>, <span class="cn">Inf</span>), </span>
<span id="cb21-272"><a href="#cb21-272" aria-hidden="true" tabindex="-1"></a>                   <span class="at">labels=</span><span class="fu">c</span>(<span class="st">"Low"</span>,<span class="st">"Medium"</span>,<span class="st">"Medium high"</span>, <span class="st">"High"</span>))</span>
<span id="cb21-273"><a href="#cb21-273" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb21-274"><a href="#cb21-274" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-275"><a href="#cb21-275" aria-hidden="true" tabindex="-1"></a><span class="ss">2. </span>Next, plot the PCA similar to how we did with cell cycle regression. *Hint: use the new `mitoFr`variable to split cells and color them accordingly.*</span>
<span id="cb21-276"><a href="#cb21-276" aria-hidden="true" tabindex="-1"></a><span class="ss">3. </span>Evaluate the PCA plot generated in #2.</span>
<span id="cb21-277"><a href="#cb21-277" aria-hidden="true" tabindex="-1"></a><span class="ss">    1. </span>Determine whether or not you observe an effect.</span>
<span id="cb21-278"><a href="#cb21-278" aria-hidden="true" tabindex="-1"></a><span class="ss">    2. </span>Describe what you see. </span>
<span id="cb21-279"><a href="#cb21-279" aria-hidden="true" tabindex="-1"></a><span class="ss">    3. </span>Would you regress out mitochondrial fraction as a source of unwanted variation?</span>
<span id="cb21-280"><a href="#cb21-280" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-281"><a href="#cb21-281" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb21-282"><a href="#cb21-282" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-283"><a href="#cb21-283" aria-hidden="true" tabindex="-1"></a>***</span>
<span id="cb21-284"><a href="#cb21-284" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-285"><a href="#cb21-285" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-286"><a href="#cb21-286" aria-hidden="true" tabindex="-1"></a><span class="fu">## Normalization and regressing out sources of unwanted variation using SCTransform</span></span>
<span id="cb21-287"><a href="#cb21-287" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-288"><a href="#cb21-288" aria-hidden="true" tabindex="-1"></a>In the <span class="co">[</span><span class="ot">Hafemeister and Satija, 2019 paper</span><span class="co">](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1874-1)</span> the authors explored the issues with simple transformations. Specifically they evaluated the standard log normalization approach and found that genes with different abundances are affected differently and that **effective normalization (using the log transform) is only observed with low/medium abundance genes (Figure 1D, below)**. Additionally, **substantial imbalances in variance were observed with the log-normalized data (Figure 1E, below)**. In particular, cells with low total UMI counts exhibited disproportionately higher variance for high-abundance genes, dampening the variance contribution from other gene abundances.&nbsp;</span>
<span id="cb21-289"><a href="#cb21-289" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-290"><a href="#cb21-290" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;p</span> <span class="er">align</span><span class="ot">=</span><span class="st">"center"</span><span class="kw">&gt;</span></span>
<span id="cb21-291"><a href="#cb21-291" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;img</span> <span class="er">src</span><span class="ot">=</span><span class="st">"../img/SCT_Fig1.png"</span> <span class="er">width</span><span class="ot">=</span><span class="st">"600"</span><span class="kw">&gt;</span></span>
<span id="cb21-292"><a href="#cb21-292" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;/p&gt;</span></span>
<span id="cb21-293"><a href="#cb21-293" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-294"><a href="#cb21-294" aria-hidden="true" tabindex="-1"></a>_**Image credit:** Hafemeister C and Satija R. Normalization and variance stabilization of single-cell RNA-seq data using regularized negative binomial regression, Genom Biology 2019 (https://doi.org/10.1101/576827)_</span>
<span id="cb21-295"><a href="#cb21-295" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-296"><a href="#cb21-296" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-297"><a href="#cb21-297" aria-hidden="true" tabindex="-1"></a>The conclusion is, **we cannot treat all genes the same.**</span>
<span id="cb21-298"><a href="#cb21-298" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-299"><a href="#cb21-299" aria-hidden="true" tabindex="-1"></a>The proposed solution was the use of **Pearson residuals for transformation**, as implemented in Seurat's <span class="in">`SCTransform`</span> function. With this approach:</span>
<span id="cb21-300"><a href="#cb21-300" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>Measurements are multiplied by a gene-specific weight</span>
<span id="cb21-301"><a href="#cb21-301" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>Each gene is weighted based on how much evidence there is that it is non-uniformly expressed across cells</span>
<span id="cb21-302"><a href="#cb21-302" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>More evidence == more of a weight; Genes that are expressed in only a small fraction of cells will be favored (useful for finding rare cell populations)</span>
<span id="cb21-303"><a href="#cb21-303" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>Not just a consideration of the expression level is, but also the distribution of expression</span>
<span id="cb21-304"><a href="#cb21-304" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-305"><a href="#cb21-305" aria-hidden="true" tabindex="-1"></a>::: callout-note</span>
<span id="cb21-306"><a href="#cb21-306" aria-hidden="true" tabindex="-1"></a><span class="fu"># Why don't we just run SCTransform to normalize?</span></span>
<span id="cb21-307"><a href="#cb21-307" aria-hidden="true" tabindex="-1"></a>While the functions <span class="in">`NormalizeData`</span>, <span class="in">`VariableFeatures`</span> and <span class="in">`ScaleData`</span> can be replaced by the function <span class="in">`SCTransform`</span>, the latter uses a more sophisticated way to perform the normalization and scaling. We suggest using log normalization because it is good to observe the data and any trends using a simple transformation, as methods like SCT can alter the data in a way that is not as intuitive to interpret. </span>
<span id="cb21-308"><a href="#cb21-308" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb21-309"><a href="#cb21-309" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-310"><a href="#cb21-310" aria-hidden="true" tabindex="-1"></a>Now that we have established which effects are observed in our data, we can use the SCTransform method to regress out these effects. The **SCTransform** method was proposed as a better alternative to the log transform normalization method that we used for exploring sources of unwanted variation. The method not only **normalizes data, but it also performs a variance stabilization and allows for additional covariates to be regressed out**.</span>
<span id="cb21-311"><a href="#cb21-311" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-312"><a href="#cb21-312" aria-hidden="true" tabindex="-1"></a>As described earlier, all genes cannot be treated the same. As such, the **SCTransform method constructs a generalized linear model (GLM) for each gene** with UMI counts as the response and sequencing depth as the explanatory variable. Information is pooled across genes with similar abundances, to regularize parameter estimates and **obtain residuals which represent effectively normalized data values** which are no longer correlated with sequencing depth.</span>
<span id="cb21-313"><a href="#cb21-313" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-314"><a href="#cb21-314" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;p</span> <span class="er">align</span><span class="ot">=</span><span class="st">"center"</span><span class="kw">&gt;</span></span>
<span id="cb21-315"><a href="#cb21-315" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;img</span> <span class="er">src</span><span class="ot">=</span><span class="st">"../img/SCT_UMAP.png"</span> <span class="er">width</span><span class="ot">=</span><span class="st">"400"</span><span class="kw">&gt;</span></span>
<span id="cb21-316"><a href="#cb21-316" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;/p&gt;</span></span>
<span id="cb21-317"><a href="#cb21-317" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-318"><a href="#cb21-318" aria-hidden="true" tabindex="-1"></a>_**Image credit:** Hafemeister C and Satija R. Normalization and variance stabilization of single-cell RNA-seq data using regularized negative binomial regression, Genom Biology 2019 (https://doi.org/10.1101/576827)_</span>
<span id="cb21-319"><a href="#cb21-319" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-320"><a href="#cb21-320" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-321"><a href="#cb21-321" aria-hidden="true" tabindex="-1"></a>::: callout-note</span>
<span id="cb21-322"><a href="#cb21-322" aria-hidden="true" tabindex="-1"></a>Since the UMI counts are part of the GLM, the effects are automatically regressed out. The user can include any additional covariates (<span class="in">`vars.to.regress`</span>) that may have an effect on expression and will be included in the model. </span>
<span id="cb21-323"><a href="#cb21-323" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb21-324"><a href="#cb21-324" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-325"><a href="#cb21-325" aria-hidden="true" tabindex="-1"></a>To run the SCTransform we have the code below as an example. **Do not run this code**, as we prefer to run this for each sample separately in the next section below.</span>
<span id="cb21-326"><a href="#cb21-326" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-327"><a href="#cb21-327" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-330"><a href="#cb21-330" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb21-331"><a href="#cb21-331" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb21-332"><a href="#cb21-332" aria-hidden="true" tabindex="-1"></a><span class="do">## DO NOT RUN CODE ##</span></span>
<span id="cb21-333"><a href="#cb21-333" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-334"><a href="#cb21-334" aria-hidden="true" tabindex="-1"></a><span class="co"># SCTranform</span></span>
<span id="cb21-335"><a href="#cb21-335" aria-hidden="true" tabindex="-1"></a>seurat_phase <span class="ot">&lt;-</span> <span class="fu">SCTransform</span>(seurat_phase, <span class="at">vars.to.regress =</span> <span class="fu">c</span>(<span class="st">"mitoRatio"</span>))</span>
<span id="cb21-336"><a href="#cb21-336" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb21-337"><a href="#cb21-337" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-338"><a href="#cb21-338" aria-hidden="true" tabindex="-1"></a><span class="fu">## Iterating over samples in a dataset</span></span>
<span id="cb21-339"><a href="#cb21-339" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-340"><a href="#cb21-340" aria-hidden="true" tabindex="-1"></a>Since we have two samples in our dataset (from two conditions), we want to keep them as separate objects and transform them as that is what is required for integration. We will first split the cells in <span class="in">`seurat_phase`</span> object into "Control" and "Stimulated":</span>
<span id="cb21-341"><a href="#cb21-341" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-344"><a href="#cb21-344" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb21-345"><a href="#cb21-345" aria-hidden="true" tabindex="-1"></a><span class="co"># Split seurat object by condition to perform cell cycle scoring and SCT on all samples</span></span>
<span id="cb21-346"><a href="#cb21-346" aria-hidden="true" tabindex="-1"></a>split_seurat <span class="ot">&lt;-</span> <span class="fu">SplitObject</span>(seurat_phase, <span class="at">split.by =</span> <span class="st">"sample"</span>)</span>
<span id="cb21-347"><a href="#cb21-347" aria-hidden="true" tabindex="-1"></a>split_seurat</span>
<span id="cb21-348"><a href="#cb21-348" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb21-349"><a href="#cb21-349" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-350"><a href="#cb21-350" aria-hidden="true" tabindex="-1"></a>::: callout-note</span>
<span id="cb21-351"><a href="#cb21-351" aria-hidden="true" tabindex="-1"></a>If you only wanted to integrate on a subset of your samples (e.g. all Ctrl replicates only), you could select which ones you wanted from the <span class="in">`split_seurat`</span> object as shown below and move forward with those.</span>
<span id="cb21-352"><a href="#cb21-352" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-355"><a href="#cb21-355" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb21-356"><a href="#cb21-356" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb21-357"><a href="#cb21-357" aria-hidden="true" tabindex="-1"></a>ctrl_reps <span class="ot">&lt;-</span> split_seurat[<span class="fu">c</span>(<span class="st">"ctrl_1"</span>, <span class="st">"ctrl_2"</span>)]</span>
<span id="cb21-358"><a href="#cb21-358" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb21-359"><a href="#cb21-359" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb21-360"><a href="#cb21-360" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-361"><a href="#cb21-361" aria-hidden="true" tabindex="-1"></a>Now we will **use a 'for loop'** to run the <span class="in">`SCTransform()`</span> on each sample, and regress out mitochondrial expression by specifying in the <span class="in">`vars.to.regress`</span> argument of the <span class="in">`SCTransform()`</span> function.</span>
<span id="cb21-362"><a href="#cb21-362" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-363"><a href="#cb21-363" aria-hidden="true" tabindex="-1"></a>Before we run this <span class="in">`for loop`</span>, we know that the output can generate large R objects/variables in terms of memory. If we have a large dataset, then we might need to **adjust the limit for allowable object sizes within R** (*Default is 500 * 1024 ^ 2 = 500 Mb*) using the following code:</span>
<span id="cb21-364"><a href="#cb21-364" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-367"><a href="#cb21-367" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb21-368"><a href="#cb21-368" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">future.globals.maxSize =</span> <span class="dv">4000</span> <span class="sc">*</span> <span class="dv">1024</span><span class="sc">^</span><span class="dv">2</span>)</span>
<span id="cb21-369"><a href="#cb21-369" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb21-370"><a href="#cb21-370" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-371"><a href="#cb21-371" aria-hidden="true" tabindex="-1"></a>Now, we run the following loop to **perform the sctransform on all samples**. This may take some time (~10 minutes):</span>
<span id="cb21-372"><a href="#cb21-372" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-375"><a href="#cb21-375" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb21-376"><a href="#cb21-376" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(split_seurat)) {</span>
<span id="cb21-377"><a href="#cb21-377" aria-hidden="true" tabindex="-1"></a>    split_seurat[[i]] <span class="ot">&lt;-</span> <span class="fu">SCTransform</span>(split_seurat[[i]], </span>
<span id="cb21-378"><a href="#cb21-378" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">vars.to.regress =</span> <span class="fu">c</span>(<span class="st">"mitoRatio"</span>),</span>
<span id="cb21-379"><a href="#cb21-379" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">vst.flavor =</span> <span class="st">"v2"</span>)</span>
<span id="cb21-380"><a href="#cb21-380" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb21-381"><a href="#cb21-381" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb21-382"><a href="#cb21-382" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-383"><a href="#cb21-383" aria-hidden="true" tabindex="-1"></a>Please note that in the for loop above, we specify that <span class="in">`vst.flavor = "v2"`</span> to use the updated version of SCT. "v2" was introduced in early 2022, and is now commonly used. This update improves:</span>
<span id="cb21-384"><a href="#cb21-384" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>Speed and memory consumption</span>
<span id="cb21-385"><a href="#cb21-385" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>The stability of parameter estimates</span>
<span id="cb21-386"><a href="#cb21-386" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>Variable feature identification in subsequent steps</span>
<span id="cb21-387"><a href="#cb21-387" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-388"><a href="#cb21-388" aria-hidden="true" tabindex="-1"></a>For more information, please see the <span class="co">[</span><span class="ot">Seurat vignette's section on SCTransform, v2 regularization</span><span class="co">](https://satijalab.org/seurat/articles/sctransform_v2_vignette.html)</span>. </span>
<span id="cb21-389"><a href="#cb21-389" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-390"><a href="#cb21-390" aria-hidden="true" tabindex="-1"></a>::: callout-note</span>
<span id="cb21-391"><a href="#cb21-391" aria-hidden="true" tabindex="-1"></a>By default, after normalizing, adjusting the variance, and regressing out uninteresting sources of variation, SCTransform will rank the genes by residual variance and output the 3000 most variant genes. If the dataset has larger cell numbers, then it may be beneficial to adjust this parameter higher using the <span class="in">`variable.features.n`</span> argument._ </span>
<span id="cb21-392"><a href="#cb21-392" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb21-393"><a href="#cb21-393" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-394"><a href="#cb21-394" aria-hidden="true" tabindex="-1"></a>Note, the last line of output specifies **"Set default assay to SCT"**. This specifies that moving forward we would like to use the data after SCT was implemented. We can view the different assays that we have stored in our seurat object.</span>
<span id="cb21-395"><a href="#cb21-395" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-398"><a href="#cb21-398" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb21-399"><a href="#cb21-399" aria-hidden="true" tabindex="-1"></a><span class="co"># Check which assays are stored in objects</span></span>
<span id="cb21-400"><a href="#cb21-400" aria-hidden="true" tabindex="-1"></a>split_seurat<span class="sc">$</span>ctrl<span class="sc">@</span>assays</span>
<span id="cb21-401"><a href="#cb21-401" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb21-402"><a href="#cb21-402" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-403"><a href="#cb21-403" aria-hidden="true" tabindex="-1"></a>Now we can see that in addition to the raw RNA counts, we now have a SCT component in our <span class="in">`assays`</span> slot. The most variable features will be the only genes stored inside the SCT assay. As we move through the scRNA-seq analysis, we will choose the most appropriate assay to use for the different steps in the analysis. </span>
<span id="cb21-404"><a href="#cb21-404" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-405"><a href="#cb21-405" aria-hidden="true" tabindex="-1"></a>***</span>
<span id="cb21-406"><a href="#cb21-406" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-407"><a href="#cb21-407" aria-hidden="true" tabindex="-1"></a>::: callout-tip</span>
<span id="cb21-408"><a href="#cb21-408" aria-hidden="true" tabindex="-1"></a><span class="fu"># Exercises</span></span>
<span id="cb21-409"><a href="#cb21-409" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-410"><a href="#cb21-410" aria-hidden="true" tabindex="-1"></a><span class="ss">4. </span>Are the same assays available for the "stim" samples within the <span class="in">`split_seurat`</span> object? What is the code you used to check that?</span>
<span id="cb21-411"><a href="#cb21-411" aria-hidden="true" tabindex="-1"></a><span class="ss">5. </span>Any observations for the genes or features listed under *"First 10 features:"* and the *"Top 10 variable features:"* for "ctrl" versus "stim"?</span>
<span id="cb21-412"><a href="#cb21-412" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb21-413"><a href="#cb21-413" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-414"><a href="#cb21-414" aria-hidden="true" tabindex="-1"></a>***</span>
<span id="cb21-415"><a href="#cb21-415" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-416"><a href="#cb21-416" aria-hidden="true" tabindex="-1"></a><span class="fu">### Save the object!</span></span>
<span id="cb21-417"><a href="#cb21-417" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-418"><a href="#cb21-418" aria-hidden="true" tabindex="-1"></a>Before finishing up, let's save this object to the <span class="in">`data/`</span> folder. It can take a while to get back to this stage especially when working with large datasets, it is best practice to save the object as an easily loadable file locally.</span>
<span id="cb21-419"><a href="#cb21-419" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-422"><a href="#cb21-422" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb21-423"><a href="#cb21-423" aria-hidden="true" tabindex="-1"></a><span class="co"># Save the split seurat object</span></span>
<span id="cb21-424"><a href="#cb21-424" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(split_seurat, <span class="st">"../data/split_seurat.rds"</span>)</span>
<span id="cb21-425"><a href="#cb21-425" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb21-426"><a href="#cb21-426" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-427"><a href="#cb21-427" aria-hidden="true" tabindex="-1"></a>::: callout-note</span>
<span id="cb21-428"><a href="#cb21-428" aria-hidden="true" tabindex="-1"></a>To load the <span class="in">`.rds`</span> file back into your environment you would use the llowing code:</span>
<span id="cb21-431"><a href="#cb21-431" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb21-432"><a href="#cb21-432" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb21-433"><a href="#cb21-433" aria-hidden="true" tabindex="-1"></a><span class="co"># Load the split seurat object into the environment</span></span>
<span id="cb21-434"><a href="#cb21-434" aria-hidden="true" tabindex="-1"></a>split_seurat <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"../data/split_seurat.rds"</span>)</span>
<span id="cb21-435"><a href="#cb21-435" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb21-436"><a href="#cb21-436" aria-hidden="true" tabindex="-1"></a>:::</span>
</code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center">This lesson has been developed by members of the teaching team at the <a href="http://bioinformatics.sph.harvard.edu/">Harvard Chan Bioinformatics Core (HBC)</a>. <br> These are open access materials distributed under the terms of the <a href="https://creativecommons.org/licenses/by/4.0/">Creative Commons Attribution license</a> (CC BY 4.0), which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are credited. <br></div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>



</body></html>