<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.353">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Mary Piper, Lorena Pantano, Meeta Mistry, Radhika Khetani, Jihe Liu">

<title>Introduction to single-cell RNA-seq - Marker identification</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>


<link rel="stylesheet" href="../css/styles.css">
</head>

<body class="nav-sidebar docked nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">Introduction to single-cell RNA-seq</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../lessons/schedule.html" rel="" target="">
 <span class="menu-text">Schedule</span></a>
  </li>  
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="https://bioinformatics.sph.harvard.edu/" rel="" target="">
 <span class="menu-text">HBC</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="https://github.com/hbctraining/Intro-to-scRNAseq-Quarto" rel="" target="">
 <span class="menu-text">GitHub</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="mailto:hbctraining@hsph.harvard.edu" rel="" target="">
 <span class="menu-text">Contact us</span></a>
  </li>  
</ul>
            <div class="quarto-navbar-tools">
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../lessons/09_merged_SC_marker_identification.html">Day 3:</a></li><li class="breadcrumb-item"><a href="../lessons/09_merged_SC_marker_identification.html">Marker identification</a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation docked overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="false">
 <span class="menu-text">Pre-reading:</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lessons/01_intro_to_scRNA-seq.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Introduction to single-cell RNA-seq</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lessons/02_SC_generation_of_count_matrix.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Generation of count matrix</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="false">
 <span class="menu-text">Day 1:</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lessons/03_SC_quality_control-setup.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Quality Control Setup</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="false">
 <span class="menu-text">Day 1 Self-learning:</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lessons/04_cellranger_QC.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Quality Control of Cellranger Output</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lessons/04_SC_quality_control.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Quality Control Analysis</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lessons/05_theory_of_PCA.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Theory of PCA</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="false">
 <span class="menu-text">Day 2:</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lessons/06_SC_SCT_normalization.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Normalization, identification of most variable genes</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lessons/06a_integration_cca_theory.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Integration</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="false">
 <span class="menu-text">Day 2 Self-learning</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lessons/06b_integration_code_harmony.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Performing Integration</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lessons/07_SC_clustering_cells_SCT.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Clustering Analysis</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lessons/08_SC_clustering_quality_control.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Clustering QC</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lessons/seurat_cheatsheet.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Seurat Cheatsheet</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="true">
 <span class="menu-text">Day 3:</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-6" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lessons/09_merged_SC_marker_identification.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">Marker identification</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lessons/scRNAseq_workflow.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Workflow Overview</span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#learning-objectives" id="toc-learning-objectives" class="nav-link active" data-scroll-target="#learning-objectives">Learning Objectives:</a></li>
  <li><a href="#single-cell-rna-seq-marker-identification" id="toc-single-cell-rna-seq-marker-identification" class="nav-link" data-scroll-target="#single-cell-rna-seq-marker-identification">Single-cell RNA-seq marker identification</a>
  <ul class="collapse">
  <li><a href="#identification-of-all-markers-for-each-cluster" id="toc-identification-of-all-markers-for-each-cluster" class="nav-link" data-scroll-target="#identification-of-all-markers-for-each-cluster">Identification of all markers for each cluster</a></li>
  <li><a href="#identification-of-conserved-markers-in-all-conditions" id="toc-identification-of-conserved-markers-in-all-conditions" class="nav-link" data-scroll-target="#identification-of-conserved-markers-in-all-conditions">Identification of conserved markers in all conditions</a>
  <ul class="collapse">
  <li><a href="#adding-gene-annotations" id="toc-adding-gene-annotations" class="nav-link" data-scroll-target="#adding-gene-annotations">Adding Gene Annotations</a></li>
  <li><a href="#running-on-multiple-samples" id="toc-running-on-multiple-samples" class="nav-link" data-scroll-target="#running-on-multiple-samples">Running on multiple samples</a></li>
  <li><a href="#evaluating-marker-genes" id="toc-evaluating-marker-genes" class="nav-link" data-scroll-target="#evaluating-marker-genes">Evaluating marker genes</a></li>
  <li><a href="#visualizing-marker-genes" id="toc-visualizing-marker-genes" class="nav-link" data-scroll-target="#visualizing-marker-genes">Visualizing marker genes</a></li>
  </ul></li>
  <li><a href="#identifying-gene-markers-for-each-cluster" id="toc-identifying-gene-markers-for-each-cluster" class="nav-link" data-scroll-target="#identifying-gene-markers-for-each-cluster">Identifying gene markers for each cluster</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title">Marker identification</h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source"><i class="bi"></i> Code</button></div></div>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Mary Piper, Lorena Pantano, Meeta Mistry, Radhika Khetani, Jihe Liu </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">Invalid Date</p>
    </div>
  </div>
  
    
  </div>
  

</header>

<p>Approximate time: 75 minutes</p>
<section id="learning-objectives" class="level2">
<h2 class="anchored" data-anchor-id="learning-objectives">Learning Objectives:</h2>
<ul>
<li>Describe how to determine markers of individual clusters</li>
<li>Discuss the iterative processes of clustering and marker identification</li>
</ul>
</section>
<section id="single-cell-rna-seq-marker-identification" class="level1">
<h1>Single-cell RNA-seq marker identification</h1>
<p>Now that we have identified our desired clusters, we can move on to marker identification, which will allow us to verify the identity of certain clusters and help surmise the identity of any unknown clusters.</p>
<p align="center">
<img src="../img/sc_workflow_2022.jpg" width="630">
</p>
<hr>
<p><em><strong>Goals:</strong></em></p>
<ul>
<li><em>To <strong>determine the gene markers</strong> for each of the clusters</em></li>
<li><em>To <strong>identify cell types</strong> of each cluster using markers</em></li>
<li><em>To determine whether there’s a need to <strong>re-cluster based on cell type markers</strong>, perhaps clusters need to be merged or split</em></li>
</ul>
<p><em><strong>Challenges:</strong></em></p>
<ul>
<li><em>Over-interpretation of the results</em></li>
<li><em>Combining different types of marker identification</em></li>
</ul>
<p><em><strong>Recommendations:</strong></em></p>
<ul>
<li><em>Think of the results as hypotheses that need verification. Inflated p-values can lead to over-interpretation of results (essentially each cell is used as a replicate). Top markers are most trustworthy.</em></li>
<li><em>Identify all markers conserved between conditions for each cluster</em></li>
<li><em>Identify markers that are differentially expressed between specific clusters</em></li>
</ul>
<hr>
<p>Our clustering analysis resulted in the following clusters:</p>
<p align="center">
<img src="../img/SC_umap_SCTv2.png" width="800">
</p>
<p><strong>Remember that we had the following questions from the clustering analysis</strong>:</p>
<ol type="1">
<li><em>Do the clusters corresponding to the same cell types have biologically meaningful differences? Are there subpopulations of these cell types?</em></li>
<li><em>Can we acquire higher confidence in these cell type identities by identifying other marker genes for these clusters?</em></li>
</ol>
<p>There are a few different types of marker identification that we can explore using Seurat to get to the answer of these questions. Each with their own benefits and drawbacks:</p>
<ol type="1">
<li><strong>Identification of all markers for each cluster:</strong> this analysis compares each cluster against all others and outputs the genes that are differentially expressed/present.
<ul>
<li><em>Useful for identifying unknown clusters and improving confidence in hypothesized cell types.</em></li>
</ul></li>
<li><strong>Identification of conserved markers for each cluster:</strong> This analysis looks for genes that are differentially expressed/present within each condition first, and then reports those genes that are conserved in the cluster across all conditions. These genes can help to figure out the identity for the cluster.
<ul>
<li><em>Useful with more than one condition to identify cell type markers that are conserved across conditions.</em></li>
</ul></li>
<li><strong>Marker identification between specific clusters:</strong> this analysis explores differentially expressed genes between specific clusters.
<ul>
<li><em>Useful for determining differences in gene expression between clusters that appear to be representing the same celltype (i.e with markers that are similar) from the above analyses.</em></li>
</ul></li>
</ol>
<section id="identification-of-all-markers-for-each-cluster" class="level2">
<h2 class="anchored" data-anchor-id="identification-of-all-markers-for-each-cluster">Identification of all markers for each cluster</h2>
<p>This type of analysis is typically <strong>recommended for when evaluating a single sample group/condition</strong>. With the <code>FindAllMarkers()</code> function we are comparing each cluster against all other clusters to identify potential marker genes. The cells in each cluster are treated as replicates, and essentially a <strong>differential expression analysis is performed</strong> with some statistical test.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>The default is a Wilcoxon Rank Sum test, but there are other options available.</p>
</div>
</div>
<p align="center">
<img src="../img/marker_ident_function1.png" width="350">
</p>
<p>The <code>FindAllMarkers()</code> function has <strong>three important arguments</strong> which provide thresholds for determining whether a gene is a marker:</p>
<ul>
<li><code>logfc.threshold</code>: minimum log2 fold change for average expression of gene in cluster relative to the average expression in all other clusters combined. Default is 0.25.
<ul>
<li><strong>Cons:</strong>
<ul>
<li>could miss those cell markers that are expressed in a small fraction of cells within the cluster of interest, but not in the other clusters, if the average logfc doesn’t meet the threshold</li>
<li>could return a lot of metabolic/ribosomal genes due to slight differences in metabolic output by different cell types, which are not as useful to distinguish cell type identities</li>
</ul></li>
</ul></li>
<li><code>min.diff.pct</code>: minimum percent difference between the percent of cells expressing the gene in the cluster and the percent of cells expressing gene in all other clusters combined.
<ul>
<li><strong>Cons:</strong> could miss those cell markers that are expressed in all cells, but are highly up-regulated in this specific cell type</li>
</ul></li>
<li><code>min.pct</code>: only test genes that are detected in a minimum fraction of cells in either of the two populations. Meant to speed up the function by not testing genes that are very infrequently expressed. Default is 0.1.
<ul>
<li><strong>Cons:</strong> if set to a very high value could incur many false negatives due to the fact that not all genes are detected in all cells (even if it is expressed)</li>
</ul></li>
</ul>
<p>You could use any combination of these arguments depending on how stringent/lenient you want to be. Also, by default this function will return to you genes that exhibit both positive and negative expression changes. Typically, we add an argument <code>only.pos</code> to opt for keeping only the positive changes. The code to find markers for each cluster is shown below. <strong>We will not run this code.</strong></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="do">## DO NOT RUN THIS CODE ##</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Find markers for every cluster compared to all remaining cells, report only the positive ones</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>markers <span class="ot">&lt;-</span> <span class="fu">FindAllMarkers</span>(<span class="at">object =</span> seurat_integrated, </span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>                          <span class="at">only.pos =</span> <span class="cn">TRUE</span>,</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>                          <span class="at">logfc.threshold =</span> <span class="fl">0.25</span>)                     </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>This command can take quite long to run, as it is processing each individual cluster against all other cells.</p>
</div>
</div>
</section>
<section id="identification-of-conserved-markers-in-all-conditions" class="level2">
<h2 class="anchored" data-anchor-id="identification-of-conserved-markers-in-all-conditions">Identification of conserved markers in all conditions</h2>
<p>Since we have samples representing different conditions in our dataset, <strong>our best option is to find conserved markers</strong>. This function internally separates out cells by sample group/condition, and then performs differential gene expression testing for a single specified cluster against all other clusters (or a second cluster, if specified). Gene-level p-values are computed for each condition and then combined across groups using meta-analysis methods from the MetaDE R package.</p>
<p align="center">
<img src="../img/marker_ident_function2.png" width="350">
</p>
<p>Before we start our marker identification we will explicitly set our default assay, we want to use the <strong>normalized data, but not the integrated data</strong>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(seurat_integrated) <span class="ot">&lt;-</span> <span class="st">"RNA"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The default assay should have already been <code>RNA</code>, because we set it up in the previous clustering quality control lesson. But we encourage you to run this line of code above to be absolutely sure in case the active slot was changed somewhere upstream in your analysis.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Why don’t we use SCT normalized data?
</div>
</div>
<div class="callout-body-container callout-body">
<p>Note that the raw and normalized counts are stored in the <code>counts</code> and <code>data</code> slots of <code>RNA</code> assay, respectively. By default, the functions for finding markers will use normalized data if RNA is the DefaultAssay. The number of features in the <code>RNA</code> assay corresponds to all genes in our dataset.</p>
<p>Now if we consider the <code>SCT</code> assay, functions for finding markers would use the <code>scale.data</code> slot which is the pearson residuals that come out of regularized NB regression. Differential expression on these values can be difficult interpret. Additionally, only the variable features are represented in this assay and so we may not have data for some of our marker genes.</p>
</div>
</div>
<p>The function <code>FindConservedMarkers()</code>, has the following structure:</p>
<p><strong><code>FindConservedMarkers()</code> syntax:</strong></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="do">## DO NOT RUN ##</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">FindConservedMarkers</span>(seurat_integrated,</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>                      <span class="at">ident.1 =</span> cluster,</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>                      <span class="at">grouping.var =</span> <span class="st">"sample"</span>,</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>                      <span class="at">only.pos =</span> <span class="cn">TRUE</span>,</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>                      <span class="at">min.diff.pct =</span> <span class="fl">0.25</span>,</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>                      <span class="at">min.pct =</span> <span class="fl">0.25</span>,</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>                      <span class="at">logfc.threshold =</span> <span class="fl">0.25</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>You will recognize some of the arguments we described previously for the <code>FindAllMarkers()</code> function; this is because internally it is using that function to first find markers within each group. Here, we list <strong>some additional arguments</strong> which provide for when using <code>FindConservedMarkers()</code>:</p>
<ul>
<li><code>ident.1</code>: this function only evaluates one cluster at a time; here you would specify the cluster of interest.</li>
<li><code>grouping.var</code>: the variable (column header) in your metadata which specifies the separation of cells into groups</li>
</ul>
<p>For our analysis we will be fairly lenient and <strong>use only the log fold change threshold greater than 0.25</strong>. We will also specify to return only the positive markers for each cluster.</p>
<p>Let’s <strong>test it out on one cluster</strong> to see how it works:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>cluster0_conserved_markers <span class="ot">&lt;-</span> <span class="fu">FindConservedMarkers</span>(seurat_integrated,</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>                                                    <span class="at">ident.1 =</span> <span class="dv">0</span>,</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>                                                    <span class="at">grouping.var =</span> <span class="st">"sample"</span>,</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>                                                    <span class="at">only.pos =</span> <span class="cn">TRUE</span>,</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>                                                    <span class="at">logfc.threshold =</span> <span class="fl">0.25</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><img src="../img/conserved_markers_table.png" width="800"></p>
<p><strong>The output from the <code>FindConservedMarkers()</code> function</strong>, is a matrix containing a ranked list of putative markers listed by gene ID for the cluster we specified, and associated statistics. Note that the same set of statistics are computed for each group (in our case, Ctrl and Stim) and the last two columns correspond to the combined p-value across the two groups. We describe some of these columns below:</p>
<ul>
<li><strong>gene:</strong> gene symbol</li>
<li><strong>condition_p_val:</strong> p-value not adjusted for multiple test correction for condition</li>
<li><strong>condition_avg_logFC:</strong> average log fold change for condition. Positive values indicate that the gene is more highly expressed in the cluster.<br>
</li>
<li><strong>condition_pct.1:</strong> percentage of cells where the gene is detected in the cluster for condition<br>
</li>
<li><strong>condition_pct.2:</strong> percentage of cells where the gene is detected on average in the other clusters for condition</li>
<li><strong>condition_p_val_adj:</strong> adjusted p-value for condition, based on bonferroni correction using all genes in the dataset, used to determine significance</li>
<li><strong>max_pval:</strong> largest p value of p value calculated by each group/condition</li>
<li><strong>minimump_p_val:</strong> combined p value</li>
</ul>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Since each cell is being treated as a replicate this will result in inflated p-values within each group! A gene may have an incredibly low p-value &lt; 1e-50 but that doesn’t translate as a highly reliable marker gene.</p>
</div>
</div>
<p>When looking at the output, <strong>we suggest looking for markers with large differences in expression between <code>pct.1</code> and <code>pct.2</code> and larger fold changes</strong>. For instance if <code>pct.1</code> = 0.90 and <code>pct.2</code> = 0.80, it may not be as exciting of a marker. However, if <code>pct.2</code> = 0.1 instead, the bigger difference would be more convincing. Also, of interest is if the majority of cells expressing the marker is in my cluster of interest. If <code>pct.1</code> is low, such as 0.3, it may not be as interesting. Both of these are also possible parameters to include when running the function, as described above.</p>
<section id="adding-gene-annotations" class="level3">
<h3 class="anchored" data-anchor-id="adding-gene-annotations">Adding Gene Annotations</h3>
<p>It can be helpful to add columns with gene annotation information. In order to do that we will load in an annotation file located in your <code>data</code> folder, using the code provided below:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>annotations <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">"../data/annotation.csv"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>If you are interested in knowing how we obtained this annotation file, take a look at <a href="../lessons/fetching_annotations.html">the linked materials</a>.</p>
</div>
</div>
<p>First, we will turn the row names with gene identifiers into its own columns. Then we will merge this annotation file with our results from the <code>FindConservedMarkers()</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Combine markers with gene descriptions </span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>cluster0_ann_markers <span class="ot">&lt;-</span> cluster0_conserved_markers <span class="sc">%&gt;%</span> </span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>                <span class="fu">rownames_to_column</span>(<span class="at">var=</span><span class="st">"gene"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>                <span class="fu">left_join</span>(<span class="at">y =</span> <span class="fu">unique</span>(annotations[, <span class="fu">c</span>(<span class="st">"gene_name"</span>, <span class="st">"description"</span>)]),</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>                          <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"gene"</span> <span class="ot">=</span> <span class="st">"gene_name"</span>))</span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(cluster0_ann_markers)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>    gene stim_p_val stim_avg_log2FC stim_pct.1 stim_pct.2 stim_p_val_adj
1   CCR7          0       1.2798175      0.927      0.430              0
2   SELL          0       1.4023434      0.832      0.370              0
3   LDHB          0       1.4447683      0.732      0.304              0
4 GIMAP7          0       1.1120136      0.934      0.507              0
5    LTB          0       1.3746583      0.702      0.295              0
6 RPL10A          0       0.7947952      0.966      0.664              0
     ctrl_p_val ctrl_avg_log2FC ctrl_pct.1 ctrl_pct.2 ctrl_p_val_adj
1  0.000000e+00       1.3185064      0.839      0.368   0.000000e+00
2  0.000000e+00       1.8564100      0.610      0.186   0.000000e+00
3 2.056415e-293       1.2074798      0.734      0.356  2.892347e-289
4  0.000000e+00       1.2337458      0.804      0.380   0.000000e+00
5  0.000000e+00       1.3920719      0.770      0.327   0.000000e+00
6  0.000000e+00       0.5195068      0.969      0.809   0.000000e+00
       max_pval minimump_p_val
1  0.000000e+00              0
2  0.000000e+00              0
3 2.056415e-293              0
4  0.000000e+00              0
5  0.000000e+00              0
6  0.000000e+00              0
                                                        description
1 C-C motif chemokine receptor 7 [Source:HGNC Symbol;Acc:HGNC:1608]
2                    selectin L [Source:HGNC Symbol;Acc:HGNC:10720]
3        lactate dehydrogenase B [Source:HGNC Symbol;Acc:HGNC:6541]
4  GTPase, IMAP family member 7 [Source:HGNC Symbol;Acc:HGNC:22404]
5               lymphotoxin beta [Source:HGNC Symbol;Acc:HGNC:6711]
6        ribosomal protein L10a [Source:HGNC Symbol;Acc:HGNC:10299]</code></pre>
</div>
</div>
<hr>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercises
</div>
</div>
<div class="callout-body-container callout-body">
<ol type="1">
<li>In the previous lesson, we identified cluster 10 as FCGR3A+ monocytes by inspecting the expression of known cell markers FCGR3A and MS4A7. Use <code>FindConservedMarkers()</code> function to find conserved markers for cluster 10. What do you observe? Do you see FCGR3A and MS4A7 as highly expressed genes in cluster 10?</li>
</ol>
</div>
</div>
<hr>
</section>
<section id="running-on-multiple-samples" class="level3">
<h3 class="anchored" data-anchor-id="running-on-multiple-samples">Running on multiple samples</h3>
<p>The function <code>FindConservedMarkers()</code> <strong>accepts a single cluster at a time</strong>, and we could run this function as many times as we have clusters. However, this is not very efficient. Instead we will first create a function to find the conserved markers including all the parameters we want to include. We will also <strong>add a few lines of code to modify the output</strong>. Our function will:</p>
<ol type="1">
<li>Run the <code>FindConservedMarkers()</code> function</li>
<li>Transfer row names to a column using <code>rownames_to_column()</code> function</li>
<li>Merge in annotations</li>
<li>Create the column of cluster IDs using the <code>cbind()</code> function</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create function to get conserved markers for any given cluster</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>get_conserved <span class="ot">&lt;-</span> <span class="cf">function</span>(cluster){</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">FindConservedMarkers</span>(seurat_integrated,</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>                       <span class="at">ident.1 =</span> cluster,</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>                       <span class="at">grouping.var =</span> <span class="st">"sample"</span>,</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>                       <span class="at">only.pos =</span> <span class="cn">TRUE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">rownames_to_column</span>(<span class="at">var =</span> <span class="st">"gene"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">left_join</span>(<span class="at">y =</span> <span class="fu">unique</span>(annotations[, <span class="fu">c</span>(<span class="st">"gene_name"</span>, <span class="st">"description"</span>)]),</span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>               <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"gene"</span> <span class="ot">=</span> <span class="st">"gene_name"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cbind</span>(<span class="at">cluster_id =</span> cluster, .)</span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a>  }</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now that we have this function created we can use it as an argument to the appropriate <code>map</code> function. We want the output of the <code>map</code> family of functions to be a <strong>dataframe with each cluster output bound together by rows</strong>, we will use the <code>map_dfr()</code> function.</p>
<p><strong><code>map</code> family syntax:</strong></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="do">## DO NOT RUN ##</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="fu">map_dfr</span>(inputs_to_function, name_of_function)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now, let’s try this function to <strong>find the conserved markers for the clusters that were identified as CD4+ T cells (4,0,6,2)</strong> from our use of known marker genes. Let’s see what genes we identify and of there are overlaps or obvious differences that can help us tease this apart a bit more.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Iterate function across desired clusters</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>conserved_markers <span class="ot">&lt;-</span> <span class="fu">map_dfr</span>(<span class="fu">c</span>(<span class="dv">4</span>,<span class="dv">0</span>,<span class="dv">6</span>,<span class="dv">2</span>), get_conserved)</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(conserved_markers)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  cluster_id       gene    stim_p_val stim_avg_log2FC stim_pct.1 stim_pct.2
1          4       EIF1  0.000000e+00       0.4723411      0.992      0.926
2          4 AC026979.2 7.026043e-238       4.4271681      0.157      0.010
3          4      SRSF2 7.842593e-260       2.1962219      0.543      0.161
4          4       BTG1 9.918923e-225       0.7148641      0.936      0.726
5          4      HSPH1 2.191992e-151       2.4007275      0.333      0.091
6          4      NR4A2 2.914853e-115       3.2800433      0.126      0.016
  stim_p_val_adj    ctrl_p_val ctrl_avg_log2FC ctrl_pct.1 ctrl_pct.2
1   0.000000e+00 3.178029e-271       0.4172717      0.976      0.912
2  9.882129e-234 4.082678e-291       4.6142429      0.177      0.010
3  1.103061e-255 2.348094e-234       2.0261441      0.537      0.181
4  1.395097e-220 3.139357e-250       0.5747494      0.952      0.827
5  3.083036e-147 1.587155e-245       2.9187251      0.359      0.073
6  4.099741e-111 1.877585e-229       3.6920102      0.197      0.019
  ctrl_p_val_adj      max_pval minimump_p_val
1  4.469898e-267 3.178029e-271   0.000000e+00
2  5.742287e-287 7.026043e-238  8.165357e-291
3  3.302594e-230 2.348094e-234  1.568519e-259
4  4.415506e-246 9.918923e-225  6.278715e-250
5  2.232333e-241 2.191992e-151  3.174310e-245
6  2.640824e-225 2.914853e-115  3.755170e-229
                                                                        description
1     eukaryotic translation initiation factor 1 [Source:HGNC Symbol;Acc:HGNC:3249]
2                                                                  novel transcript
3    serine and arginine rich splicing factor 2 [Source:HGNC Symbol;Acc:HGNC:10783]
4                BTG anti-proliferation factor 1 [Source:HGNC Symbol;Acc:HGNC:1130]
5 heat shock protein family H (Hsp110) member 1 [Source:HGNC Symbol;Acc:HGNC:16969]
6  nuclear receptor subfamily 4 group A member 2 [Source:HGNC Symbol;Acc:HGNC:7981]</code></pre>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Finding markers for all clusters
</div>
</div>
<div class="callout-body-container callout-body">
<p>For your data, you may want to run this function on all clusters, in which case you could input <code>0:20</code> instead of <code>c(4,0,6,2)</code>; however, it would take quite a while to run. Also, it is possible that when you run this function on all clusters, in <strong>some cases you will have clusters that do not have enough cells for a particular group</strong> - and your function will fail. For these clusters you will need to use <code>FindAllMarkers()</code>.</p>
</div>
</div>
</section>
<section id="evaluating-marker-genes" class="level3">
<h3 class="anchored" data-anchor-id="evaluating-marker-genes">Evaluating marker genes</h3>
<p>We would like to use these gene lists to see of we can <strong>identify which celltypes these clusters identify with.</strong> Let’s take a look at the top genes for each of the clusters and see if that gives us any hints. We can view the top 10 markers by average fold change across the two groups, for each cluster for a quick perusal:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract top 10 markers per cluster</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>top10 <span class="ot">&lt;-</span> conserved_markers <span class="sc">%&gt;%</span> </span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">avg_fc =</span> (ctrl_avg_log2FC <span class="sc">+</span> stim_avg_log2FC) <span class="sc">/</span><span class="dv">2</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(cluster_id) <span class="sc">%&gt;%</span> </span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">top_n</span>(<span class="at">n =</span> <span class="dv">10</span>, </span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>        <span class="at">wt =</span> avg_fc)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Visualize top 10 markers per cluster</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="fu">View</span>(top10)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p align="center">
<img src="../img/tcell_marker_table_SCTv2.png" width="800">
</p>
<p>When we look at the entire list, we see clusters 0 and 6 have some overlapping genes, like CCR7 and SELL which correspond to <strong>markers of memory T cells</strong>. It is possible that these two clusters are more similar to one another and could be merged together as naive T cells. On the other hand, with cluster 2 we observe CREM as one of our top genes; a <strong>marker gene of activation</strong>. This suggests that perhaps cluster 2 represents activated T cells.</p>
<table class="table">
<thead>
<tr class="header">
<th style="text-align: center;">Cell State</th>
<th style="text-align: center;">Marker</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">Naive T cells</td>
<td style="text-align: center;">CCR7, SELL</td>
</tr>
<tr class="even">
<td style="text-align: center;">Activated T cells</td>
<td style="text-align: center;">CREM, CD69</td>
</tr>
</tbody>
</table>
<p>For cluster 4, we see a lot of heat shock and DNA damage genes appear in the top gene list. Based on these markers, it is likely that these are <strong>stressed or dying cells</strong>. However, if we explore the quality metrics for these cells in more detail (i.e.&nbsp;mitoRatio and nUMI overlayed on the cluster) we don’t really support for this argument. There is a breadth of research supporting the association of heat shock proteins with reactive T cells in the induction of anti‐inflammatory cytokines in chronic inflammation. This is a cluster for which we would need a deeper understanding of immune cells to really tease apart the results and make a final conclusion.</p>
</section>
<section id="visualizing-marker-genes" class="level3">
<h3 class="anchored" data-anchor-id="visualizing-marker-genes">Visualizing marker genes</h3>
<p>To get a better idea of cell type identity for <strong>cluster 4</strong> we can <strong>explore the expression of different identified markers</strong> by cluster using the <code>FeaturePlot()</code> function. We see that only a subset of cluster 4 are highly expressing these genes.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot interesting marker gene expression for cluster 4</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a><span class="fu">FeaturePlot</span>(<span class="at">object =</span> seurat_integrated, </span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>            <span class="at">features =</span> <span class="fu">c</span>(<span class="st">"HSPH1"</span>, <span class="st">"HSPE1"</span>, <span class="st">"DNAJB1"</span>),</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>            <span class="at">order =</span> <span class="cn">TRUE</span>,</span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>            <span class="at">min.cutoff =</span> <span class="st">'q10'</span>, </span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>            <span class="at">label =</span> <span class="cn">TRUE</span>,</span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>            <span class="at">repel =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="09_merged_SC_marker_identification_files/figure-html/unnamed-chunk-13-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<p>We can also explore the range in expression of specific markers by using <strong>violin plots</strong>:</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Violin plots
</div>
</div>
<div class="callout-body-container callout-body">
<p>Violin plots are similar to box plots, except that they also show the probability density of the data at different values, usually smoothed by a kernel density estimator. A violin plot is more informative than a plain box plot. While a box plot only shows summary statistics such as mean/median and interquartile ranges, the violin plot shows the full distribution of the data. The difference is particularly useful when the data distribution is multimodal (more than one peak). In this case a violin plot shows the presence of different peaks, their position and relative amplitude.</p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Vln plot - cluster 4</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="fu">VlnPlot</span>(<span class="at">object =</span> seurat_integrated, </span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>        <span class="at">features =</span> <span class="fu">c</span>(<span class="st">"HSPH1"</span>, <span class="st">"HSPE1"</span>, <span class="st">"DNAJB1"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="09_merged_SC_marker_identification_files/figure-html/unnamed-chunk-14-1.png" class="img-fluid" width="1440"></p>
</div>
</div>
<p>These results and plots can help us determine the identity of these clusters or verify what we hypothesize the identity to be after exploring the canonical markers of expected cell types previously.</p>
</section>
</section>
<section id="identifying-gene-markers-for-each-cluster" class="level2">
<h2 class="anchored" data-anchor-id="identifying-gene-markers-for-each-cluster">Identifying gene markers for each cluster</h2>
<p>Sometimes the list of markers returned don’t sufficiently separate some of the clusters. For instance, we had previously identified clusters 0, 4, 6 and 2 as CD4+ T cells, but when looking at marker gene lists we identfied markers to help us further subset cells. We were lucky and the signal observed from <code>FindAllMarkers()</code> helped us differentiate between naive and activated cells. Another option to identify biologically meaningful differences would be to use the <strong><code>FindMarkers()</code> function to determine the genes that are differentially expressed between two specific clusters</strong>.</p>
<p align="center">
<img src="../img/marker_ident_function3.png" width="350">
</p>
<p>We can try all combinations of comparisons, but we’ll start with cluster 2 versus all other CD4+ T cell clusters:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Determine differentiating markers for CD4+ T cell</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>cd4_tcells <span class="ot">&lt;-</span> <span class="fu">FindMarkers</span>(seurat_integrated,</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>                          <span class="at">ident.1 =</span> <span class="dv">2</span>,</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>                          <span class="at">ident.2 =</span> <span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">4</span>,<span class="dv">6</span>))                  </span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Add gene symbols to the DE table</span></span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a>cd4_tcells <span class="ot">&lt;-</span> cd4_tcells <span class="sc">%&gt;%</span></span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames_to_column</span>(<span class="at">var =</span> <span class="st">"gene"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(<span class="at">y =</span> <span class="fu">unique</span>(annotations[, <span class="fu">c</span>(<span class="st">"gene_name"</span>, <span class="st">"description"</span>)]),</span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a>             <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"gene"</span> <span class="ot">=</span> <span class="st">"gene_name"</span>))</span>
<span id="cb16-11"><a href="#cb16-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-12"><a href="#cb16-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Reorder columns and sort by padj      </span></span>
<span id="cb16-13"><a href="#cb16-13" aria-hidden="true" tabindex="-1"></a>cd4_tcells <span class="ot">&lt;-</span> cd4_tcells[, <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">3</span><span class="sc">:</span><span class="dv">5</span>,<span class="dv">2</span>,<span class="dv">6</span><span class="sc">:</span><span class="dv">7</span>)]</span>
<span id="cb16-14"><a href="#cb16-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-15"><a href="#cb16-15" aria-hidden="true" tabindex="-1"></a>cd4_tcells <span class="ot">&lt;-</span> cd4_tcells <span class="sc">%&gt;%</span></span>
<span id="cb16-16"><a href="#cb16-16" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">arrange</span>(p_val_adj) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co"># View data</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="fu">View</span>(cd4_tcells)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p align="center">
<img src="../img/cd4t_markers_table_SCTv2.png" width="800">
</p>
<p>Of these top genes the <strong>CREM gene</strong> stands out as a marker of activation with a positive fold change. We also see markers of naive or memory cells include the SELL and CCR7 genes with negative fold changes, which is in line with previous results.</p>
<p>As markers for the naive and activated states both showed up in the marker list, it is helpful to visualize expression. Based on these plots it seems as though clusters 0 and 2 are reliably the naive T cells. However, for the activated T cells it is hard to tell. We might say that clusters 4 and 18 are activated T cells, but the CD69 expression is not as apparent as CREM. We will label the naive cells and leave the remaining clusters labeled as CD4+ T cells.</p>
<p>Now taking all of this information, we can surmise the cell types of the different clusters and plot the cells with cell type labels.</p>
<table class="table">
<thead>
<tr class="header">
<th style="text-align: center;">Cluster ID</th>
<th style="text-align: center;">Cell Type</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">0</td>
<td style="text-align: center;">Naive or memory CD4+ T cells</td>
</tr>
<tr class="even">
<td style="text-align: center;">1</td>
<td style="text-align: center;">CD14+ monocytes</td>
</tr>
<tr class="odd">
<td style="text-align: center;">2</td>
<td style="text-align: center;">Activated T cells</td>
</tr>
<tr class="even">
<td style="text-align: center;">3</td>
<td style="text-align: center;">CD14+ monocytes</td>
</tr>
<tr class="odd">
<td style="text-align: center;">4</td>
<td style="text-align: center;">Stressed cells / Unknown</td>
</tr>
<tr class="even">
<td style="text-align: center;">5</td>
<td style="text-align: center;">CD8+ T cells</td>
</tr>
<tr class="odd">
<td style="text-align: center;">6</td>
<td style="text-align: center;">Naive or memory CD4+ T cells</td>
</tr>
<tr class="even">
<td style="text-align: center;">7</td>
<td style="text-align: center;">B cells</td>
</tr>
<tr class="odd">
<td style="text-align: center;">8</td>
<td style="text-align: center;">NK cells</td>
</tr>
<tr class="even">
<td style="text-align: center;">9</td>
<td style="text-align: center;">CD8+ T cells</td>
</tr>
<tr class="odd">
<td style="text-align: center;">10</td>
<td style="text-align: center;">FCGR3A+ monocytes</td>
</tr>
<tr class="even">
<td style="text-align: center;">11</td>
<td style="text-align: center;">B cells</td>
</tr>
<tr class="odd">
<td style="text-align: center;">12</td>
<td style="text-align: center;">NK cells</td>
</tr>
<tr class="even">
<td style="text-align: center;">13</td>
<td style="text-align: center;">B cells</td>
</tr>
<tr class="odd">
<td style="text-align: center;">14</td>
<td style="text-align: center;">Conventional dendritic cells</td>
</tr>
<tr class="even">
<td style="text-align: center;">15</td>
<td style="text-align: center;">Megakaryocytes</td>
</tr>
<tr class="odd">
<td style="text-align: center;">16</td>
<td style="text-align: center;">Plasmacytoid dendritic cells</td>
</tr>
</tbody>
</table>
<p>We can then reassign the identity of the clusters to these cell types:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Rename all identities</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>seurat_integrated <span class="ot">&lt;-</span> <span class="fu">RenameIdents</span>(<span class="at">object =</span> seurat_integrated, </span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>                               <span class="st">"0"</span> <span class="ot">=</span> <span class="st">"Naive or memory CD4+ T cells"</span>,</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>                               <span class="st">"1"</span> <span class="ot">=</span> <span class="st">"CD14+ monocytes"</span>,</span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>                               <span class="st">"2"</span> <span class="ot">=</span> <span class="st">"Activated T cells"</span>,</span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>                               <span class="st">"3"</span> <span class="ot">=</span> <span class="st">"CD14+ monocytes"</span>,</span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a>                               <span class="st">"4"</span> <span class="ot">=</span> <span class="st">"Stressed cells / Unknown"</span>,</span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a>                               <span class="st">"5"</span> <span class="ot">=</span> <span class="st">"CD8+ T cells"</span>,</span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a>                               <span class="st">"6"</span> <span class="ot">=</span> <span class="st">"Naive or memory CD4+ T cells"</span>,</span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a>                               <span class="st">"7"</span> <span class="ot">=</span> <span class="st">"B cells"</span>,</span>
<span id="cb18-11"><a href="#cb18-11" aria-hidden="true" tabindex="-1"></a>                               <span class="st">"8"</span> <span class="ot">=</span> <span class="st">"NK cells"</span>,</span>
<span id="cb18-12"><a href="#cb18-12" aria-hidden="true" tabindex="-1"></a>                               <span class="st">"9"</span> <span class="ot">=</span> <span class="st">"CD8+ T cells"</span>,</span>
<span id="cb18-13"><a href="#cb18-13" aria-hidden="true" tabindex="-1"></a>                               <span class="st">"10"</span> <span class="ot">=</span> <span class="st">"FCGR3A+ monocytes"</span>,</span>
<span id="cb18-14"><a href="#cb18-14" aria-hidden="true" tabindex="-1"></a>                               <span class="st">"11"</span> <span class="ot">=</span> <span class="st">"B cells"</span>,</span>
<span id="cb18-15"><a href="#cb18-15" aria-hidden="true" tabindex="-1"></a>                               <span class="st">"12"</span> <span class="ot">=</span> <span class="st">"NK cells"</span>,</span>
<span id="cb18-16"><a href="#cb18-16" aria-hidden="true" tabindex="-1"></a>                               <span class="st">"13"</span> <span class="ot">=</span> <span class="st">"B cells"</span>,</span>
<span id="cb18-17"><a href="#cb18-17" aria-hidden="true" tabindex="-1"></a>                               <span class="st">"14"</span> <span class="ot">=</span> <span class="st">"Conventional dendritic cells"</span>,</span>
<span id="cb18-18"><a href="#cb18-18" aria-hidden="true" tabindex="-1"></a>                               <span class="st">"15"</span> <span class="ot">=</span> <span class="st">"Megakaryocytes"</span>,</span>
<span id="cb18-19"><a href="#cb18-19" aria-hidden="true" tabindex="-1"></a>                   <span class="st">"16"</span> <span class="ot">=</span> <span class="st">"Plasmacytoid dendritic cells"</span>)</span>
<span id="cb18-20"><a href="#cb18-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-21"><a href="#cb18-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-22"><a href="#cb18-22" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot the UMAP</span></span>
<span id="cb18-23"><a href="#cb18-23" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(<span class="at">object =</span> seurat_integrated, </span>
<span id="cb18-24"><a href="#cb18-24" aria-hidden="true" tabindex="-1"></a>        <span class="at">reduction =</span> <span class="st">"umap"</span>, </span>
<span id="cb18-25"><a href="#cb18-25" aria-hidden="true" tabindex="-1"></a>        <span class="at">label =</span> <span class="cn">TRUE</span>,</span>
<span id="cb18-26"><a href="#cb18-26" aria-hidden="true" tabindex="-1"></a>        <span class="at">label.size =</span> <span class="dv">3</span>,</span>
<span id="cb18-27"><a href="#cb18-27" aria-hidden="true" tabindex="-1"></a>        <span class="at">repel =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="09_merged_SC_marker_identification_files/figure-html/unnamed-chunk-17-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>If we wanted to remove the potentially stressed cells, we could use the <code>subset()</code> function:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove the stressed or dying cells</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>seurat_subset_labeled <span class="ot">&lt;-</span> <span class="fu">subset</span>(seurat_integrated,</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>                               <span class="at">idents =</span> <span class="st">"Stressed cells / Unknown"</span>, <span class="at">invert =</span> <span class="cn">TRUE</span>)</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Re-visualize the clusters</span></span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(<span class="at">object =</span> seurat_subset_labeled, </span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a>        <span class="at">reduction =</span> <span class="st">"umap"</span>, </span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a>        <span class="at">label =</span> <span class="cn">TRUE</span>,</span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a>        <span class="at">label.size =</span> <span class="dv">3</span>,</span>
<span id="cb19-10"><a href="#cb19-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">repel =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="09_merged_SC_marker_identification_files/figure-html/unnamed-chunk-18-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Now we would want to save our final labelled Seurat object and the output of <code>sessionInfo()</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Save final R object</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a><span class="fu">write_rds</span>(seurat_integrated,</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>          <span class="at">file =</span> <span class="st">"../results/seurat_labelled.rds"</span>)</span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Create and save a text file with sessionInfo</span></span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a><span class="fu">sink</span>(<span class="st">"../results/sessionInfo_scrnaseq_Feb2023.txt"</span>)</span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a><span class="fu">sink</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>You can find out more about the <code>sink()</code> function <a href="https://www.rdocumentation.org/packages/base/versions/3.6.2/topics/sink">at this link</a>.</p>
</div>
</div>
<hr>
<p>Now that we have our clusters defined and the markers for each of our clusters, we have a few different questions we can answer:</p>
<ul>
<li>Determine if there is a shift in cell populations between <code>ctrl</code> and <code>stim</code>. Ideally this would be done with replicates to determine if the changes are significant.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Add celltype annotation as a column in meta.data </span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>seurat_subset_labeled<span class="sc">$</span>celltype <span class="ot">&lt;-</span> <span class="fu">Idents</span>(seurat_subset_labeled)</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute number of cells per celltype</span></span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>n_cells <span class="ot">&lt;-</span> <span class="fu">FetchData</span>(seurat_subset_labeled, </span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a>                     <span class="at">vars =</span> <span class="fu">c</span>(<span class="st">"celltype"</span>, <span class="st">"sample"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a>        dplyr<span class="sc">::</span><span class="fu">count</span>(celltype, sample)</span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-9"><a href="#cb21-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Barplot of number of cells per celltype by sample</span></span>
<span id="cb21-10"><a href="#cb21-10" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(n_cells, <span class="fu">aes</span>(<span class="at">x=</span>celltype, <span class="at">y=</span>n, <span class="at">fill=</span>sample)) <span class="sc">+</span></span>
<span id="cb21-11"><a href="#cb21-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_bar</span>(<span class="at">position=</span><span class="fu">position_dodge</span>(), <span class="at">stat=</span><span class="st">"identity"</span>) <span class="sc">+</span></span>
<span id="cb21-12"><a href="#cb21-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb21-13"><a href="#cb21-13" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_text</span>(<span class="fu">aes</span>(<span class="at">label=</span>n), <span class="at">vjust =</span> <span class="sc">-</span>.<span class="dv">2</span>, <span class="at">position=</span><span class="fu">position_dodge</span>(<span class="dv">1</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="09_merged_SC_marker_identification_files/figure-html/unnamed-chunk-20-1.png" class="img-fluid" width="1728"></p>
</div>
</div>
<ul>
<li>Perform differential expression analysis between conditions <code>ctrl</code> and <code>stim</code>
<ul>
<li>Biological replicates are <strong>necessary</strong> to proceed with this analysis, and we have <a href="https://hbctraining.github.io/Pseudobulk-for-scRNAseq/">additional materials</a> to help walk through this analysis.</li>
<li>For a first pass look, we can use the <code>FindMarkers()</code> function we have been using to do a simple wilcox test to see the difference in gene expression between conditions for the B cells</li>
</ul></li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Subset seurat object to just B cells</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>seurat_b_cells <span class="ot">&lt;-</span> <span class="fu">subset</span>(seurat_subset_labeled, <span class="at">subset =</span> (celltype <span class="sc">==</span> <span class="st">"B cells"</span>))</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Run a wilcox test to compare ctrl vs stim</span></span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a><span class="fu">Idents</span>(seurat_b_cells) <span class="ot">&lt;-</span> <span class="st">"sample"</span></span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a>b_markers <span class="ot">&lt;-</span> <span class="fu">FindMarkers</span>(seurat_b_cells,</span>
<span id="cb22-7"><a href="#cb22-7" aria-hidden="true" tabindex="-1"></a>                          <span class="at">ident.1 =</span> <span class="st">"ctrl"</span>,</span>
<span id="cb22-8"><a href="#cb22-8" aria-hidden="true" tabindex="-1"></a>                          <span class="at">ident.2 =</span> <span class="st">"stim"</span>,</span>
<span id="cb22-9"><a href="#cb22-9" aria-hidden="true" tabindex="-1"></a>                          <span class="at">grouping.var =</span> <span class="st">"sample"</span>,</span>
<span id="cb22-10"><a href="#cb22-10" aria-hidden="true" tabindex="-1"></a>                          <span class="at">only.pos =</span> <span class="cn">FALSE</span>,</span>
<span id="cb22-11"><a href="#cb22-11" aria-hidden="true" tabindex="-1"></a>                          <span class="at">logfc.threshold =</span> <span class="fl">0.25</span>)</span>
<span id="cb22-12"><a href="#cb22-12" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(b_markers)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      p_val avg_log2FC pct.1 pct.2 p_val_adj
IFIT3     0  -5.311956 0.036 0.927         0
IFI6      0  -4.340583 0.067 0.931         0
IFIT1     0  -5.611324 0.023 0.845         0
ISG15     0  -3.433947 0.174 0.995         0
MX1       0  -3.682686 0.089 0.857         0
LY6E      0  -3.232020 0.126 0.848         0</code></pre>
</div>
</div>
<ul>
<li>For added visualization, we can used the <code>EnhancedVolcano()</code> function to see how the genes fall on a volcano plot.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(EnhancedVolcano)</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a><span class="fu">EnhancedVolcano</span>(b_markers,</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">row.names</span>(b_markers),</span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">x=</span><span class="st">"avg_log2FC"</span>,</span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">y=</span><span class="st">"p_val_adj"</span>,</span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">title=</span><span class="st">"B Cells"</span>,</span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">subtitle=</span><span class="st">"Stim vs. Ctrl"</span></span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="09_merged_SC_marker_identification_files/figure-html/unnamed-chunk-22-1.png" class="img-fluid" width="960"></p>
</div>
</div>
<ul>
<li>Experimentally validate intriguing markers for our identified cell types.</li>
<li>Explore a subset of the cell types to discover subclusters of cells as described <a href="../lessons/seurat_subclustering.html">here</a></li>
<li>Trajectory analysis, or lineage tracing, could be performed if trying to determine the progression between cell types or cell states. For example, we could explore any of the following using this type of analysis:
<ul>
<li>Differentiation processes</li>
<li>Expression changes over time</li>
<li>Cell state changes in expression</li>
</ul></li>
</ul>


<!-- -->

</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb25" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a><span class="an">title:</span><span class="co"> "Marker identification"</span></span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a><span class="an">author:</span><span class="co"> "Mary Piper, Lorena Pantano, Meeta Mistry, Radhika Khetani, Jihe Liu"</span></span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a><span class="an">date:</span><span class="co"> Wednesday, February 5th, 2020</span></span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a>Approximate time: 75 minutes</span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-11"><a href="#cb25-11" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-12"><a href="#cb25-12" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb25-13"><a href="#cb25-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-14"><a href="#cb25-14" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb25-15"><a href="#cb25-15" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb25-16"><a href="#cb25-16" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb25-17"><a href="#cb25-17" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(cowplot)</span>
<span id="cb25-18"><a href="#cb25-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-19"><a href="#cb25-19" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="fu">bzfile</span>(<span class="st">"../data/additional_data/seurat_integrated.RData.bz2"</span>))</span>
<span id="cb25-20"><a href="#cb25-20" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-21"><a href="#cb25-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-22"><a href="#cb25-22" aria-hidden="true" tabindex="-1"></a><span class="fu">## Learning Objectives:</span></span>
<span id="cb25-23"><a href="#cb25-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-24"><a href="#cb25-24" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>Describe how to determine markers of individual clusters</span>
<span id="cb25-25"><a href="#cb25-25" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>Discuss the iterative processes of clustering and marker identification</span>
<span id="cb25-26"><a href="#cb25-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-27"><a href="#cb25-27" aria-hidden="true" tabindex="-1"></a><span class="fu"># Single-cell RNA-seq marker identification</span></span>
<span id="cb25-28"><a href="#cb25-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-29"><a href="#cb25-29" aria-hidden="true" tabindex="-1"></a>Now that we have identified our desired clusters, we can move on to marker identification, which will allow us to verify the identity of certain clusters and help surmise the identity of any unknown clusters. </span>
<span id="cb25-30"><a href="#cb25-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-31"><a href="#cb25-31" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;p</span> <span class="er">align</span><span class="ot">=</span><span class="st">"center"</span><span class="kw">&gt;</span></span>
<span id="cb25-32"><a href="#cb25-32" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;img</span> <span class="er">src</span><span class="ot">=</span><span class="st">"../img/sc_workflow_2022.jpg"</span> <span class="er">width</span><span class="ot">=</span><span class="st">"630"</span><span class="kw">&gt;</span></span>
<span id="cb25-33"><a href="#cb25-33" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;/p&gt;</span></span>
<span id="cb25-34"><a href="#cb25-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-35"><a href="#cb25-35" aria-hidden="true" tabindex="-1"></a>***</span>
<span id="cb25-36"><a href="#cb25-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-37"><a href="#cb25-37" aria-hidden="true" tabindex="-1"></a>_**Goals:**_ </span>
<span id="cb25-38"><a href="#cb25-38" aria-hidden="true" tabindex="-1"></a> </span>
<span id="cb25-39"><a href="#cb25-39" aria-hidden="true" tabindex="-1"></a><span class="ss"> - </span>_To **determine the gene markers** for each of the clusters_</span>
<span id="cb25-40"><a href="#cb25-40" aria-hidden="true" tabindex="-1"></a><span class="ss"> - </span>_To **identify cell types** of each cluster using markers_</span>
<span id="cb25-41"><a href="#cb25-41" aria-hidden="true" tabindex="-1"></a><span class="ss"> - </span>_To determine whether there's a need to **re-cluster based on cell type markers**, perhaps clusters need to be merged or split_</span>
<span id="cb25-42"><a href="#cb25-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-43"><a href="#cb25-43" aria-hidden="true" tabindex="-1"></a>_**Challenges:**_</span>
<span id="cb25-44"><a href="#cb25-44" aria-hidden="true" tabindex="-1"></a> </span>
<span id="cb25-45"><a href="#cb25-45" aria-hidden="true" tabindex="-1"></a><span class="ss"> - </span>_Over-interpretation of the results_</span>
<span id="cb25-46"><a href="#cb25-46" aria-hidden="true" tabindex="-1"></a><span class="ss"> - </span>_Combining different types of marker identification_</span>
<span id="cb25-47"><a href="#cb25-47" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-48"><a href="#cb25-48" aria-hidden="true" tabindex="-1"></a>_**Recommendations:**_</span>
<span id="cb25-49"><a href="#cb25-49" aria-hidden="true" tabindex="-1"></a> </span>
<span id="cb25-50"><a href="#cb25-50" aria-hidden="true" tabindex="-1"></a><span class="ss"> - </span>_Think of the results as hypotheses that need verification. Inflated p-values can lead to over-interpretation of results (essentially each cell is used as a replicate). Top markers are most trustworthy._</span>
<span id="cb25-51"><a href="#cb25-51" aria-hidden="true" tabindex="-1"></a><span class="ss"> - </span>_Identify all markers conserved between conditions for each cluster_</span>
<span id="cb25-52"><a href="#cb25-52" aria-hidden="true" tabindex="-1"></a><span class="ss"> - </span>_Identify markers that are differentially expressed between specific clusters_</span>
<span id="cb25-53"><a href="#cb25-53" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-54"><a href="#cb25-54" aria-hidden="true" tabindex="-1"></a>***</span>
<span id="cb25-55"><a href="#cb25-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-56"><a href="#cb25-56" aria-hidden="true" tabindex="-1"></a>Our clustering analysis resulted in the following clusters:</span>
<span id="cb25-57"><a href="#cb25-57" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-58"><a href="#cb25-58" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;p</span> <span class="er">align</span><span class="ot">=</span><span class="st">"center"</span><span class="kw">&gt;</span></span>
<span id="cb25-59"><a href="#cb25-59" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;img</span> <span class="er">src</span><span class="ot">=</span><span class="st">"../img/SC_umap_SCTv2.png"</span> <span class="er">width</span><span class="ot">=</span><span class="st">"800"</span><span class="kw">&gt;</span></span>
<span id="cb25-60"><a href="#cb25-60" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;/p&gt;</span></span>
<span id="cb25-61"><a href="#cb25-61" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-62"><a href="#cb25-62" aria-hidden="true" tabindex="-1"></a>**Remember that we had the following questions from the clustering analysis**: </span>
<span id="cb25-63"><a href="#cb25-63" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-64"><a href="#cb25-64" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span>_Do the clusters corresponding to the same cell types have biologically meaningful differences? Are there subpopulations of these cell types?_</span>
<span id="cb25-65"><a href="#cb25-65" aria-hidden="true" tabindex="-1"></a><span class="ss">2. </span>_Can we acquire higher confidence in these cell type identities by identifying other marker genes for these clusters?_</span>
<span id="cb25-66"><a href="#cb25-66" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-67"><a href="#cb25-67" aria-hidden="true" tabindex="-1"></a>There are a few different types of marker identification that we can explore using Seurat to get to the answer of these questions. Each with their own benefits and drawbacks:</span>
<span id="cb25-68"><a href="#cb25-68" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-69"><a href="#cb25-69" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span>**Identification of all markers for each cluster:** this analysis compares each cluster against all others and outputs the genes that are differentially expressed/present. </span>
<span id="cb25-70"><a href="#cb25-70" aria-hidden="true" tabindex="-1"></a><span class="ss">    - </span>*Useful for identifying unknown clusters and improving confidence in hypothesized cell types.*</span>
<span id="cb25-71"><a href="#cb25-71" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-72"><a href="#cb25-72" aria-hidden="true" tabindex="-1"></a><span class="ss">2. </span>**Identification of conserved markers for each cluster:** This analysis looks for genes that are differentially expressed/present within each condition first, and then reports those genes that are conserved in the cluster across all conditions. These genes can help to figure out the identity for the cluster. </span>
<span id="cb25-73"><a href="#cb25-73" aria-hidden="true" tabindex="-1"></a><span class="ss">    - </span>*Useful with more than one condition to identify cell type markers that are conserved across conditions.*     </span>
<span id="cb25-74"><a href="#cb25-74" aria-hidden="true" tabindex="-1"></a> </span>
<span id="cb25-75"><a href="#cb25-75" aria-hidden="true" tabindex="-1"></a><span class="ss">3. </span>**Marker identification between specific clusters:** this analysis explores differentially expressed genes between specific clusters. </span>
<span id="cb25-76"><a href="#cb25-76" aria-hidden="true" tabindex="-1"></a><span class="ss">    - </span>*Useful for determining differences in gene expression between clusters that appear to be representing the same celltype (i.e with markers that are similar) from the above analyses.*</span>
<span id="cb25-77"><a href="#cb25-77" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-78"><a href="#cb25-78" aria-hidden="true" tabindex="-1"></a><span class="fu">## Identification of all markers for each cluster</span></span>
<span id="cb25-79"><a href="#cb25-79" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-80"><a href="#cb25-80" aria-hidden="true" tabindex="-1"></a>This type of analysis is typically **recommended for when evaluating a single sample group/condition**. With the ` FindAllMarkers()` function we are comparing each cluster against all other clusters to identify potential marker genes. The cells in each cluster are treated as replicates, and essentially a **differential expression analysis is performed** with some statistical test. </span>
<span id="cb25-81"><a href="#cb25-81" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-82"><a href="#cb25-82" aria-hidden="true" tabindex="-1"></a>::: callout-note</span>
<span id="cb25-83"><a href="#cb25-83" aria-hidden="true" tabindex="-1"></a>The default is a Wilcoxon Rank Sum test, but there are other options available. </span>
<span id="cb25-84"><a href="#cb25-84" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb25-85"><a href="#cb25-85" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-86"><a href="#cb25-86" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;p</span> <span class="er">align</span><span class="ot">=</span><span class="st">"center"</span><span class="kw">&gt;</span></span>
<span id="cb25-87"><a href="#cb25-87" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;img</span> <span class="er">src</span><span class="ot">=</span><span class="st">"../img/marker_ident_function1.png"</span> <span class="er">width</span><span class="ot">=</span><span class="st">"350"</span><span class="kw">&gt;</span></span>
<span id="cb25-88"><a href="#cb25-88" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;/p&gt;</span></span>
<span id="cb25-89"><a href="#cb25-89" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-90"><a href="#cb25-90" aria-hidden="true" tabindex="-1"></a>The <span class="in">`FindAllMarkers()`</span> function has **three important arguments** which provide thresholds for determining whether a gene is a marker:</span>
<span id="cb25-91"><a href="#cb25-91" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-92"><a href="#cb25-92" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span><span class="in">`logfc.threshold`</span>: minimum log2 fold change for average expression of gene in cluster relative to the average expression in all other clusters combined. Default is 0.25.</span>
<span id="cb25-93"><a href="#cb25-93" aria-hidden="true" tabindex="-1"></a><span class="ss">    - </span>**Cons:** </span>
<span id="cb25-94"><a href="#cb25-94" aria-hidden="true" tabindex="-1"></a><span class="ss">        - </span>could miss those cell markers that are expressed in a small fraction of cells within the cluster of interest, but not in the other clusters, if the average logfc doesn't meet the threshold</span>
<span id="cb25-95"><a href="#cb25-95" aria-hidden="true" tabindex="-1"></a><span class="ss">        - </span>could return a lot of metabolic/ribosomal genes due to slight differences in metabolic output by different cell types, which are not as useful to distinguish cell type identities</span>
<span id="cb25-96"><a href="#cb25-96" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span><span class="in">`min.diff.pct`</span>: minimum percent difference between the percent of cells expressing the gene in the cluster and the percent of cells expressing gene in all other clusters combined.</span>
<span id="cb25-97"><a href="#cb25-97" aria-hidden="true" tabindex="-1"></a><span class="ss">    - </span>**Cons:** could miss those cell markers that are expressed in all cells, but are highly up-regulated in this specific cell type</span>
<span id="cb25-98"><a href="#cb25-98" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span><span class="in">`min.pct`</span>: only test genes that are detected in a minimum fraction of cells in either of the two populations. Meant to speed up the function by not testing genes that are very infrequently expressed. Default is 0.1.</span>
<span id="cb25-99"><a href="#cb25-99" aria-hidden="true" tabindex="-1"></a><span class="ss">    - </span>**Cons:** if set to a very high value could incur many false negatives due to the fact that not all genes are detected in all cells (even if it is expressed) </span>
<span id="cb25-100"><a href="#cb25-100" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb25-101"><a href="#cb25-101" aria-hidden="true" tabindex="-1"></a>You could use any combination of these arguments depending on how stringent/lenient you want to be. Also, by default this function will return to you genes that exhibit both positive and negative expression changes. Typically, we add an argument <span class="in">`only.pos`</span> to opt for keeping only the positive changes. The code to find markers for each cluster is shown below. **We will not run this code.**</span>
<span id="cb25-102"><a href="#cb25-102" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-105"><a href="#cb25-105" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-106"><a href="#cb25-106" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb25-107"><a href="#cb25-107" aria-hidden="true" tabindex="-1"></a><span class="do">## DO NOT RUN THIS CODE ##</span></span>
<span id="cb25-108"><a href="#cb25-108" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-109"><a href="#cb25-109" aria-hidden="true" tabindex="-1"></a><span class="co"># Find markers for every cluster compared to all remaining cells, report only the positive ones</span></span>
<span id="cb25-110"><a href="#cb25-110" aria-hidden="true" tabindex="-1"></a>markers <span class="ot">&lt;-</span> <span class="fu">FindAllMarkers</span>(<span class="at">object =</span> seurat_integrated, </span>
<span id="cb25-111"><a href="#cb25-111" aria-hidden="true" tabindex="-1"></a>                          <span class="at">only.pos =</span> <span class="cn">TRUE</span>,</span>
<span id="cb25-112"><a href="#cb25-112" aria-hidden="true" tabindex="-1"></a>                          <span class="at">logfc.threshold =</span> <span class="fl">0.25</span>)                     </span>
<span id="cb25-113"><a href="#cb25-113" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-114"><a href="#cb25-114" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-115"><a href="#cb25-115" aria-hidden="true" tabindex="-1"></a>::: callout-note</span>
<span id="cb25-116"><a href="#cb25-116" aria-hidden="true" tabindex="-1"></a>This command can take quite long to run, as it is processing each individual cluster against all other cells.</span>
<span id="cb25-117"><a href="#cb25-117" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb25-118"><a href="#cb25-118" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-119"><a href="#cb25-119" aria-hidden="true" tabindex="-1"></a><span class="fu">## Identification of conserved markers in all conditions</span></span>
<span id="cb25-120"><a href="#cb25-120" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-121"><a href="#cb25-121" aria-hidden="true" tabindex="-1"></a>Since we have samples representing different conditions in our dataset, **our best option is to find conserved markers**. This function internally separates out cells by sample group/condition, and then performs differential gene expression testing for a single specified cluster against all other clusters (or a second cluster, if specified). Gene-level p-values are computed for each condition and then combined across groups using meta-analysis methods from the MetaDE R package.</span>
<span id="cb25-122"><a href="#cb25-122" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-123"><a href="#cb25-123" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;p</span> <span class="er">align</span><span class="ot">=</span><span class="st">"center"</span><span class="kw">&gt;</span></span>
<span id="cb25-124"><a href="#cb25-124" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;img</span> <span class="er">src</span><span class="ot">=</span><span class="st">"../img/marker_ident_function2.png"</span> <span class="er">width</span><span class="ot">=</span><span class="st">"350"</span><span class="kw">&gt;</span></span>
<span id="cb25-125"><a href="#cb25-125" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;/p&gt;</span></span>
<span id="cb25-126"><a href="#cb25-126" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-127"><a href="#cb25-127" aria-hidden="true" tabindex="-1"></a>Before we start our marker identification we will explicitly set our default assay, we want to use the **normalized data, but not the integrated data**.</span>
<span id="cb25-128"><a href="#cb25-128" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-131"><a href="#cb25-131" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-132"><a href="#cb25-132" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(seurat_integrated) <span class="ot">&lt;-</span> <span class="st">"RNA"</span></span>
<span id="cb25-133"><a href="#cb25-133" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-134"><a href="#cb25-134" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-135"><a href="#cb25-135" aria-hidden="true" tabindex="-1"></a>The default assay should have already been <span class="in">`RNA`</span>, because we set it up in the previous clustering quality control lesson. But we encourage you to run this line of code above to be absolutely sure in case the active slot was changed somewhere upstream in your analysis. </span>
<span id="cb25-136"><a href="#cb25-136" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-137"><a href="#cb25-137" aria-hidden="true" tabindex="-1"></a>::: callout-note</span>
<span id="cb25-138"><a href="#cb25-138" aria-hidden="true" tabindex="-1"></a><span class="fu"># Why don't we use SCT normalized data?</span></span>
<span id="cb25-139"><a href="#cb25-139" aria-hidden="true" tabindex="-1"></a>Note that the raw and normalized counts are stored in the <span class="in">`counts`</span> and <span class="in">`data`</span> slots of <span class="in">`RNA`</span> assay, respectively. By default, the functions for finding markers will use normalized data if RNA is the DefaultAssay. The number of features in the <span class="in">`RNA`</span> assay corresponds to all genes in our dataset.</span>
<span id="cb25-140"><a href="#cb25-140" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-141"><a href="#cb25-141" aria-hidden="true" tabindex="-1"></a>Now if we consider the <span class="in">`SCT`</span> assay, functions for finding markers would use the <span class="in">`scale.data`</span> slot which is the pearson residuals that come out of regularized NB regression. Differential expression on these values can be difficult interpret. Additionally, only the variable features are represented in this assay and so we may not have data for some of our marker genes.</span>
<span id="cb25-142"><a href="#cb25-142" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb25-143"><a href="#cb25-143" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-144"><a href="#cb25-144" aria-hidden="true" tabindex="-1"></a>The function <span class="in">`FindConservedMarkers()`</span>, has the following structure:</span>
<span id="cb25-145"><a href="#cb25-145" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-146"><a href="#cb25-146" aria-hidden="true" tabindex="-1"></a>**`FindConservedMarkers()` syntax:**</span>
<span id="cb25-147"><a href="#cb25-147" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-150"><a href="#cb25-150" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-151"><a href="#cb25-151" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb25-152"><a href="#cb25-152" aria-hidden="true" tabindex="-1"></a><span class="do">## DO NOT RUN ##</span></span>
<span id="cb25-153"><a href="#cb25-153" aria-hidden="true" tabindex="-1"></a><span class="fu">FindConservedMarkers</span>(seurat_integrated,</span>
<span id="cb25-154"><a href="#cb25-154" aria-hidden="true" tabindex="-1"></a>                      <span class="at">ident.1 =</span> cluster,</span>
<span id="cb25-155"><a href="#cb25-155" aria-hidden="true" tabindex="-1"></a>                      <span class="at">grouping.var =</span> <span class="st">"sample"</span>,</span>
<span id="cb25-156"><a href="#cb25-156" aria-hidden="true" tabindex="-1"></a>                      <span class="at">only.pos =</span> <span class="cn">TRUE</span>,</span>
<span id="cb25-157"><a href="#cb25-157" aria-hidden="true" tabindex="-1"></a>                      <span class="at">min.diff.pct =</span> <span class="fl">0.25</span>,</span>
<span id="cb25-158"><a href="#cb25-158" aria-hidden="true" tabindex="-1"></a>                      <span class="at">min.pct =</span> <span class="fl">0.25</span>,</span>
<span id="cb25-159"><a href="#cb25-159" aria-hidden="true" tabindex="-1"></a>                      <span class="at">logfc.threshold =</span> <span class="fl">0.25</span>)</span>
<span id="cb25-160"><a href="#cb25-160" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-161"><a href="#cb25-161" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-162"><a href="#cb25-162" aria-hidden="true" tabindex="-1"></a>You will recognize some of the arguments we described previously for the <span class="in">`FindAllMarkers()`</span> function; this is because internally it is using that function to first find markers within each group. Here, we list **some additional arguments** which provide for when using <span class="in">`FindConservedMarkers()`</span>:</span>
<span id="cb25-163"><a href="#cb25-163" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-164"><a href="#cb25-164" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span><span class="in">`ident.1`</span>: this function only evaluates one cluster at a time; here you would specify the cluster of interest.</span>
<span id="cb25-165"><a href="#cb25-165" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span><span class="in">`grouping.var`</span>: the variable (column header) in your metadata which specifies the separation of cells into groups</span>
<span id="cb25-166"><a href="#cb25-166" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-167"><a href="#cb25-167" aria-hidden="true" tabindex="-1"></a>For our analysis we will be fairly lenient and **use only the log fold change threshold greater than 0.25**. We will also specify to return only the positive markers for each cluster.</span>
<span id="cb25-168"><a href="#cb25-168" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-169"><a href="#cb25-169" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-170"><a href="#cb25-170" aria-hidden="true" tabindex="-1"></a>Let's **test it out on one cluster** to see how it works:</span>
<span id="cb25-171"><a href="#cb25-171" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-174"><a href="#cb25-174" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-175"><a href="#cb25-175" aria-hidden="true" tabindex="-1"></a>cluster0_conserved_markers <span class="ot">&lt;-</span> <span class="fu">FindConservedMarkers</span>(seurat_integrated,</span>
<span id="cb25-176"><a href="#cb25-176" aria-hidden="true" tabindex="-1"></a>                                                    <span class="at">ident.1 =</span> <span class="dv">0</span>,</span>
<span id="cb25-177"><a href="#cb25-177" aria-hidden="true" tabindex="-1"></a>                                                    <span class="at">grouping.var =</span> <span class="st">"sample"</span>,</span>
<span id="cb25-178"><a href="#cb25-178" aria-hidden="true" tabindex="-1"></a>                                                    <span class="at">only.pos =</span> <span class="cn">TRUE</span>,</span>
<span id="cb25-179"><a href="#cb25-179" aria-hidden="true" tabindex="-1"></a>                                                    <span class="at">logfc.threshold =</span> <span class="fl">0.25</span>)</span>
<span id="cb25-180"><a href="#cb25-180" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-181"><a href="#cb25-181" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-182"><a href="#cb25-182" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;img</span> <span class="er">src</span><span class="ot">=</span><span class="st">"../img/conserved_markers_table.png"</span> <span class="er">width</span><span class="ot">=</span><span class="st">"800"</span><span class="kw">&gt;</span></span>
<span id="cb25-183"><a href="#cb25-183" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-184"><a href="#cb25-184" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-185"><a href="#cb25-185" aria-hidden="true" tabindex="-1"></a>**The output from the `FindConservedMarkers()` function**, is a matrix containing a ranked list of putative markers listed by gene ID for the cluster we specified, and associated statistics. Note that the same set of statistics are computed for each group (in our case, Ctrl and Stim) and the last two columns correspond to the combined p-value across the two groups. We describe some of these columns below:</span>
<span id="cb25-186"><a href="#cb25-186" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-187"><a href="#cb25-187" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>**gene:** gene symbol</span>
<span id="cb25-188"><a href="#cb25-188" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>**condition_p_val:** p-value not adjusted for multiple test correction for condition</span>
<span id="cb25-189"><a href="#cb25-189" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>**condition_avg_logFC:** average log fold change for condition. Positive values indicate that the gene is more highly expressed in the cluster.   </span>
<span id="cb25-190"><a href="#cb25-190" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>**condition_pct.1:** percentage of cells where the gene is detected in the cluster for condition      </span>
<span id="cb25-191"><a href="#cb25-191" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>**condition_pct.2:** percentage of cells where the gene is detected on average in the other clusters for condition</span>
<span id="cb25-192"><a href="#cb25-192" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>**condition_p_val_adj:** adjusted p-value for condition, based on bonferroni correction using all genes in the dataset, used to determine significance</span>
<span id="cb25-193"><a href="#cb25-193" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>**max_pval:** largest p value of p value calculated by each group/condition</span>
<span id="cb25-194"><a href="#cb25-194" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>**minimump_p_val:** combined p value</span>
<span id="cb25-195"><a href="#cb25-195" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-196"><a href="#cb25-196" aria-hidden="true" tabindex="-1"></a>::: callout-note</span>
<span id="cb25-197"><a href="#cb25-197" aria-hidden="true" tabindex="-1"></a>Since each cell is being treated as a replicate this will result in inflated p-values within each group! A gene may have an incredibly low p-value &lt; 1e-50 but that doesn't translate as a highly reliable marker gene. </span>
<span id="cb25-198"><a href="#cb25-198" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb25-199"><a href="#cb25-199" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-200"><a href="#cb25-200" aria-hidden="true" tabindex="-1"></a>When looking at the output, **we suggest looking for markers with large differences in expression between `pct.1` and `pct.2` and larger fold changes**. For instance if <span class="in">`pct.1`</span> = 0.90 and <span class="in">`pct.2`</span> = 0.80, it may not be as exciting of a marker. However, if <span class="in">`pct.2`</span> = 0.1 instead, the bigger difference would be more convincing. Also, of interest is if the majority of cells expressing the marker is in my cluster of interest. If <span class="in">`pct.1`</span> is low, such as 0.3, it may not be as interesting. Both of these are also possible parameters to include when running the function, as described above.</span>
<span id="cb25-201"><a href="#cb25-201" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-202"><a href="#cb25-202" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-203"><a href="#cb25-203" aria-hidden="true" tabindex="-1"></a><span class="fu">### Adding Gene Annotations</span></span>
<span id="cb25-204"><a href="#cb25-204" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-205"><a href="#cb25-205" aria-hidden="true" tabindex="-1"></a>It can be helpful to add columns with gene annotation information. In order to do that we will load in an annotation file located in your <span class="in">`data`</span> folder, using the code provided below: </span>
<span id="cb25-206"><a href="#cb25-206" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-209"><a href="#cb25-209" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-210"><a href="#cb25-210" aria-hidden="true" tabindex="-1"></a>annotations <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">"../data/annotation.csv"</span>)</span>
<span id="cb25-211"><a href="#cb25-211" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-212"><a href="#cb25-212" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-213"><a href="#cb25-213" aria-hidden="true" tabindex="-1"></a>::: callout-note</span>
<span id="cb25-214"><a href="#cb25-214" aria-hidden="true" tabindex="-1"></a>If you are interested in knowing how we obtained this annotation file, take a look at <span class="co">[</span><span class="ot">the linked materials</span><span class="co">](fetching_annotations.qmd)</span>.</span>
<span id="cb25-215"><a href="#cb25-215" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb25-216"><a href="#cb25-216" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-217"><a href="#cb25-217" aria-hidden="true" tabindex="-1"></a>First, we will turn the row names with gene identifiers into its own columns. Then we will merge this annotation file with our results from the <span class="in">`FindConservedMarkers()`</span>:</span>
<span id="cb25-218"><a href="#cb25-218" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-221"><a href="#cb25-221" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-222"><a href="#cb25-222" aria-hidden="true" tabindex="-1"></a><span class="co"># Combine markers with gene descriptions </span></span>
<span id="cb25-223"><a href="#cb25-223" aria-hidden="true" tabindex="-1"></a>cluster0_ann_markers <span class="ot">&lt;-</span> cluster0_conserved_markers <span class="sc">%&gt;%</span> </span>
<span id="cb25-224"><a href="#cb25-224" aria-hidden="true" tabindex="-1"></a>                <span class="fu">rownames_to_column</span>(<span class="at">var=</span><span class="st">"gene"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb25-225"><a href="#cb25-225" aria-hidden="true" tabindex="-1"></a>                <span class="fu">left_join</span>(<span class="at">y =</span> <span class="fu">unique</span>(annotations[, <span class="fu">c</span>(<span class="st">"gene_name"</span>, <span class="st">"description"</span>)]),</span>
<span id="cb25-226"><a href="#cb25-226" aria-hidden="true" tabindex="-1"></a>                          <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"gene"</span> <span class="ot">=</span> <span class="st">"gene_name"</span>))</span>
<span id="cb25-227"><a href="#cb25-227" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-228"><a href="#cb25-228" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(cluster0_ann_markers)</span>
<span id="cb25-229"><a href="#cb25-229" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-230"><a href="#cb25-230" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-231"><a href="#cb25-231" aria-hidden="true" tabindex="-1"></a>***</span>
<span id="cb25-232"><a href="#cb25-232" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-233"><a href="#cb25-233" aria-hidden="true" tabindex="-1"></a>::: callout-tip</span>
<span id="cb25-234"><a href="#cb25-234" aria-hidden="true" tabindex="-1"></a><span class="fu"># Exercises</span></span>
<span id="cb25-235"><a href="#cb25-235" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span>In the previous lesson, we identified cluster 10 as FCGR3A+ monocytes by inspecting the expression of known cell markers FCGR3A and MS4A7. Use <span class="in">`FindConservedMarkers()`</span> function to find conserved markers for cluster 10. What do you observe? Do you see FCGR3A and MS4A7 as highly expressed genes in cluster 10?</span>
<span id="cb25-236"><a href="#cb25-236" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb25-237"><a href="#cb25-237" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-238"><a href="#cb25-238" aria-hidden="true" tabindex="-1"></a>***</span>
<span id="cb25-239"><a href="#cb25-239" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-240"><a href="#cb25-240" aria-hidden="true" tabindex="-1"></a><span class="fu">### Running on multiple samples</span></span>
<span id="cb25-241"><a href="#cb25-241" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-242"><a href="#cb25-242" aria-hidden="true" tabindex="-1"></a>The function <span class="in">`FindConservedMarkers()`</span> **accepts a single cluster at a time**, and we could run this function as many times as we have clusters. However, this is not very efficient. Instead we will first create a function to find the conserved markers including all the parameters we want to include. We will also **add a few lines of code to modify the output**. Our function will:</span>
<span id="cb25-243"><a href="#cb25-243" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-244"><a href="#cb25-244" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span>Run the <span class="in">`FindConservedMarkers()`</span> function</span>
<span id="cb25-245"><a href="#cb25-245" aria-hidden="true" tabindex="-1"></a><span class="ss">2. </span>Transfer row names to a column using <span class="in">`rownames_to_column()`</span> function</span>
<span id="cb25-246"><a href="#cb25-246" aria-hidden="true" tabindex="-1"></a><span class="ss">3. </span>Merge in annotations</span>
<span id="cb25-247"><a href="#cb25-247" aria-hidden="true" tabindex="-1"></a><span class="ss">3. </span>Create the column of cluster IDs using the <span class="in">`cbind()`</span> function</span>
<span id="cb25-248"><a href="#cb25-248" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-249"><a href="#cb25-249" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-252"><a href="#cb25-252" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-253"><a href="#cb25-253" aria-hidden="true" tabindex="-1"></a><span class="co"># Create function to get conserved markers for any given cluster</span></span>
<span id="cb25-254"><a href="#cb25-254" aria-hidden="true" tabindex="-1"></a>get_conserved <span class="ot">&lt;-</span> <span class="cf">function</span>(cluster){</span>
<span id="cb25-255"><a href="#cb25-255" aria-hidden="true" tabindex="-1"></a>  <span class="fu">FindConservedMarkers</span>(seurat_integrated,</span>
<span id="cb25-256"><a href="#cb25-256" aria-hidden="true" tabindex="-1"></a>                       <span class="at">ident.1 =</span> cluster,</span>
<span id="cb25-257"><a href="#cb25-257" aria-hidden="true" tabindex="-1"></a>                       <span class="at">grouping.var =</span> <span class="st">"sample"</span>,</span>
<span id="cb25-258"><a href="#cb25-258" aria-hidden="true" tabindex="-1"></a>                       <span class="at">only.pos =</span> <span class="cn">TRUE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb25-259"><a href="#cb25-259" aria-hidden="true" tabindex="-1"></a>    <span class="fu">rownames_to_column</span>(<span class="at">var =</span> <span class="st">"gene"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb25-260"><a href="#cb25-260" aria-hidden="true" tabindex="-1"></a>    <span class="fu">left_join</span>(<span class="at">y =</span> <span class="fu">unique</span>(annotations[, <span class="fu">c</span>(<span class="st">"gene_name"</span>, <span class="st">"description"</span>)]),</span>
<span id="cb25-261"><a href="#cb25-261" aria-hidden="true" tabindex="-1"></a>               <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"gene"</span> <span class="ot">=</span> <span class="st">"gene_name"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb25-262"><a href="#cb25-262" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cbind</span>(<span class="at">cluster_id =</span> cluster, .)</span>
<span id="cb25-263"><a href="#cb25-263" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb25-264"><a href="#cb25-264" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-265"><a href="#cb25-265" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-266"><a href="#cb25-266" aria-hidden="true" tabindex="-1"></a>Now that we have this function created  we can use it as an argument to the appropriate <span class="in">`map`</span> function. We want the output of the <span class="in">`map`</span> family of functions to be a **dataframe with each cluster output bound together by rows**, we will use the <span class="in">`map_dfr()`</span> function.</span>
<span id="cb25-267"><a href="#cb25-267" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-268"><a href="#cb25-268" aria-hidden="true" tabindex="-1"></a>**`map` family syntax:**</span>
<span id="cb25-269"><a href="#cb25-269" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-272"><a href="#cb25-272" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-273"><a href="#cb25-273" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb25-274"><a href="#cb25-274" aria-hidden="true" tabindex="-1"></a><span class="do">## DO NOT RUN ##</span></span>
<span id="cb25-275"><a href="#cb25-275" aria-hidden="true" tabindex="-1"></a><span class="fu">map_dfr</span>(inputs_to_function, name_of_function)</span>
<span id="cb25-276"><a href="#cb25-276" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-277"><a href="#cb25-277" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-278"><a href="#cb25-278" aria-hidden="true" tabindex="-1"></a>Now, let's try this function to **find the conserved markers for the clusters that were identified as CD4+ T cells (4,0,6,2)** from our use of known marker genes. Let's see what genes we identify and of there are overlaps or obvious differences that can help us tease this apart a bit more.</span>
<span id="cb25-279"><a href="#cb25-279" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-282"><a href="#cb25-282" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-283"><a href="#cb25-283" aria-hidden="true" tabindex="-1"></a><span class="co"># Iterate function across desired clusters</span></span>
<span id="cb25-284"><a href="#cb25-284" aria-hidden="true" tabindex="-1"></a>conserved_markers <span class="ot">&lt;-</span> <span class="fu">map_dfr</span>(<span class="fu">c</span>(<span class="dv">4</span>,<span class="dv">0</span>,<span class="dv">6</span>,<span class="dv">2</span>), get_conserved)</span>
<span id="cb25-285"><a href="#cb25-285" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(conserved_markers)</span>
<span id="cb25-286"><a href="#cb25-286" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-287"><a href="#cb25-287" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-288"><a href="#cb25-288" aria-hidden="true" tabindex="-1"></a>::: callout-note</span>
<span id="cb25-289"><a href="#cb25-289" aria-hidden="true" tabindex="-1"></a><span class="fu"># Finding markers for all clusters</span></span>
<span id="cb25-290"><a href="#cb25-290" aria-hidden="true" tabindex="-1"></a>For your data, you may want to run this function on all clusters, in which case you could input <span class="in">`0:20`</span> instead of <span class="in">`c(4,0,6,2)`</span>; however, it would take quite a while to run. Also, it is possible that when you run this function on all clusters, in **some cases you will have clusters that do not have enough cells for a particular group** - and  your function will fail. For these clusters you will need to use <span class="in">`FindAllMarkers()`</span>.</span>
<span id="cb25-291"><a href="#cb25-291" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb25-292"><a href="#cb25-292" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-293"><a href="#cb25-293" aria-hidden="true" tabindex="-1"></a><span class="fu">### Evaluating marker genes</span></span>
<span id="cb25-294"><a href="#cb25-294" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-295"><a href="#cb25-295" aria-hidden="true" tabindex="-1"></a>We would like to use these gene lists to see of we can **identify which celltypes these clusters identify with.** Let's take a look at the top genes for each of the clusters and see if that gives us any hints. We can view the top 10 markers by average fold change across the two groups, for each cluster for a quick perusal:</span>
<span id="cb25-296"><a href="#cb25-296" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-299"><a href="#cb25-299" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-300"><a href="#cb25-300" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract top 10 markers per cluster</span></span>
<span id="cb25-301"><a href="#cb25-301" aria-hidden="true" tabindex="-1"></a>top10 <span class="ot">&lt;-</span> conserved_markers <span class="sc">%&gt;%</span> </span>
<span id="cb25-302"><a href="#cb25-302" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">avg_fc =</span> (ctrl_avg_log2FC <span class="sc">+</span> stim_avg_log2FC) <span class="sc">/</span><span class="dv">2</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb25-303"><a href="#cb25-303" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(cluster_id) <span class="sc">%&gt;%</span> </span>
<span id="cb25-304"><a href="#cb25-304" aria-hidden="true" tabindex="-1"></a>  <span class="fu">top_n</span>(<span class="at">n =</span> <span class="dv">10</span>, </span>
<span id="cb25-305"><a href="#cb25-305" aria-hidden="true" tabindex="-1"></a>        <span class="at">wt =</span> avg_fc)</span>
<span id="cb25-306"><a href="#cb25-306" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-307"><a href="#cb25-307" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-308"><a href="#cb25-308" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-311"><a href="#cb25-311" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-312"><a href="#cb25-312" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb25-313"><a href="#cb25-313" aria-hidden="true" tabindex="-1"></a><span class="co"># Visualize top 10 markers per cluster</span></span>
<span id="cb25-314"><a href="#cb25-314" aria-hidden="true" tabindex="-1"></a><span class="fu">View</span>(top10)</span>
<span id="cb25-315"><a href="#cb25-315" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-316"><a href="#cb25-316" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-317"><a href="#cb25-317" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;p</span> <span class="er">align</span><span class="ot">=</span><span class="st">"center"</span><span class="kw">&gt;</span></span>
<span id="cb25-318"><a href="#cb25-318" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;img</span> <span class="er">src</span><span class="ot">=</span><span class="st">"../img/tcell_marker_table_SCTv2.png"</span> <span class="er">width</span><span class="ot">=</span><span class="st">"800"</span><span class="kw">&gt;</span></span>
<span id="cb25-319"><a href="#cb25-319" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;/p&gt;</span></span>
<span id="cb25-320"><a href="#cb25-320" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-321"><a href="#cb25-321" aria-hidden="true" tabindex="-1"></a>When we look at the entire list, we see clusters 0 and 6 have some overlapping genes, like CCR7 and SELL which correspond to **markers of memory T cells**. It is possible that these two clusters are more similar to one another and could be merged together as naive T cells. On the other hand, with cluster 2 we observe CREM as one of our top genes; a **marker gene of activation**. This suggests that perhaps cluster 2 represents activated T cells.</span>
<span id="cb25-322"><a href="#cb25-322" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-323"><a href="#cb25-323" aria-hidden="true" tabindex="-1"></a>| Cell State | Marker |</span>
<span id="cb25-324"><a href="#cb25-324" aria-hidden="true" tabindex="-1"></a>|:---:|:---:|</span>
<span id="cb25-325"><a href="#cb25-325" aria-hidden="true" tabindex="-1"></a>| Naive T cells | CCR7, SELL | </span>
<span id="cb25-326"><a href="#cb25-326" aria-hidden="true" tabindex="-1"></a>| Activated T cells | CREM, CD69 |</span>
<span id="cb25-327"><a href="#cb25-327" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-328"><a href="#cb25-328" aria-hidden="true" tabindex="-1"></a>For cluster 4, we see a lot of heat shock and DNA damage genes appear in the top gene list. Based on these markers, it is likely that these are **stressed or dying cells**. However, if we explore the quality metrics for these cells in more detail (i.e. mitoRatio and nUMI overlayed on the cluster) we don't really support for this argument. There is a breadth of research supporting the association of heat shock proteins with reactive T cells in the induction of anti‐inflammatory cytokines in chronic inflammation. This is a cluster for which we would need a deeper understanding of immune cells to really tease apart the results and make a final conclusion.</span>
<span id="cb25-329"><a href="#cb25-329" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-330"><a href="#cb25-330" aria-hidden="true" tabindex="-1"></a><span class="fu">### Visualizing marker genes</span></span>
<span id="cb25-331"><a href="#cb25-331" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-332"><a href="#cb25-332" aria-hidden="true" tabindex="-1"></a>To get a better idea of cell type identity for **cluster 4** we can **explore the expression of different identified markers** by cluster using the <span class="in">`FeaturePlot()`</span> function. We see that only a subset of cluster 4 are highly expressing these genes.</span>
<span id="cb25-333"><a href="#cb25-333" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-334"><a href="#cb25-334" aria-hidden="true" tabindex="-1"></a><span class="in">```{r fig.width=8}</span></span>
<span id="cb25-335"><a href="#cb25-335" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot interesting marker gene expression for cluster 4</span></span>
<span id="cb25-336"><a href="#cb25-336" aria-hidden="true" tabindex="-1"></a><span class="fu">FeaturePlot</span>(<span class="at">object =</span> seurat_integrated, </span>
<span id="cb25-337"><a href="#cb25-337" aria-hidden="true" tabindex="-1"></a>            <span class="at">features =</span> <span class="fu">c</span>(<span class="st">"HSPH1"</span>, <span class="st">"HSPE1"</span>, <span class="st">"DNAJB1"</span>),</span>
<span id="cb25-338"><a href="#cb25-338" aria-hidden="true" tabindex="-1"></a>            <span class="at">order =</span> <span class="cn">TRUE</span>,</span>
<span id="cb25-339"><a href="#cb25-339" aria-hidden="true" tabindex="-1"></a>            <span class="at">min.cutoff =</span> <span class="st">'q10'</span>, </span>
<span id="cb25-340"><a href="#cb25-340" aria-hidden="true" tabindex="-1"></a>            <span class="at">label =</span> <span class="cn">TRUE</span>,</span>
<span id="cb25-341"><a href="#cb25-341" aria-hidden="true" tabindex="-1"></a>            <span class="at">repel =</span> <span class="cn">TRUE</span>)</span>
<span id="cb25-342"><a href="#cb25-342" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-343"><a href="#cb25-343" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-344"><a href="#cb25-344" aria-hidden="true" tabindex="-1"></a>We can also explore the range in expression of specific markers by using **violin plots**:</span>
<span id="cb25-345"><a href="#cb25-345" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-346"><a href="#cb25-346" aria-hidden="true" tabindex="-1"></a>::: callout-note</span>
<span id="cb25-347"><a href="#cb25-347" aria-hidden="true" tabindex="-1"></a><span class="fu"># Violin plots</span></span>
<span id="cb25-348"><a href="#cb25-348" aria-hidden="true" tabindex="-1"></a>Violin plots are similar to box plots, except that they also show the probability density of the data at different values, usually smoothed by a kernel density estimator. A violin plot is more informative than a plain box plot. While a box plot only shows summary statistics such as mean/median and interquartile ranges, the violin plot shows the full distribution of the data. The difference is particularly useful when the data distribution is multimodal (more than one peak). In this case a violin plot shows the presence of different peaks, their position and relative amplitude.</span>
<span id="cb25-349"><a href="#cb25-349" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb25-350"><a href="#cb25-350" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-351"><a href="#cb25-351" aria-hidden="true" tabindex="-1"></a><span class="in">```{r fig.width=15}</span></span>
<span id="cb25-352"><a href="#cb25-352" aria-hidden="true" tabindex="-1"></a><span class="co"># Vln plot - cluster 4</span></span>
<span id="cb25-353"><a href="#cb25-353" aria-hidden="true" tabindex="-1"></a><span class="fu">VlnPlot</span>(<span class="at">object =</span> seurat_integrated, </span>
<span id="cb25-354"><a href="#cb25-354" aria-hidden="true" tabindex="-1"></a>        <span class="at">features =</span> <span class="fu">c</span>(<span class="st">"HSPH1"</span>, <span class="st">"HSPE1"</span>, <span class="st">"DNAJB1"</span>))</span>
<span id="cb25-355"><a href="#cb25-355" aria-hidden="true" tabindex="-1"></a><span class="in">```</span>        </span>
<span id="cb25-356"><a href="#cb25-356" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-357"><a href="#cb25-357" aria-hidden="true" tabindex="-1"></a>These results and plots can help us determine the identity of these clusters or verify what we hypothesize the identity to be after exploring the canonical markers of expected cell types previously.</span>
<span id="cb25-358"><a href="#cb25-358" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-359"><a href="#cb25-359" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-360"><a href="#cb25-360" aria-hidden="true" tabindex="-1"></a><span class="fu">## Identifying gene markers for each cluster</span></span>
<span id="cb25-361"><a href="#cb25-361" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-362"><a href="#cb25-362" aria-hidden="true" tabindex="-1"></a>Sometimes the list of markers returned don't sufficiently separate some of the clusters. For instance, we had previously identified clusters 0, 4, 6 and 2 as CD4+ T cells, but when looking at marker gene lists we identfied markers to help us further subset cells. We were lucky and the signal observed from <span class="in">`FindAllMarkers()`</span> helped us differentiate between naive and activated cells. Another option to identify biologically meaningful differences would be to use the **`FindMarkers()` function to determine the genes that are differentially expressed between two specific clusters**. </span>
<span id="cb25-363"><a href="#cb25-363" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-364"><a href="#cb25-364" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;p</span> <span class="er">align</span><span class="ot">=</span><span class="st">"center"</span><span class="kw">&gt;</span></span>
<span id="cb25-365"><a href="#cb25-365" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;img</span> <span class="er">src</span><span class="ot">=</span><span class="st">"../img/marker_ident_function3.png"</span> <span class="er">width</span><span class="ot">=</span><span class="st">"350"</span><span class="kw">&gt;</span></span>
<span id="cb25-366"><a href="#cb25-366" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;/p&gt;</span></span>
<span id="cb25-367"><a href="#cb25-367" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-368"><a href="#cb25-368" aria-hidden="true" tabindex="-1"></a>We can try all combinations of comparisons, but we'll start with cluster 2 versus all other CD4+ T cell clusters:</span>
<span id="cb25-369"><a href="#cb25-369" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-372"><a href="#cb25-372" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-373"><a href="#cb25-373" aria-hidden="true" tabindex="-1"></a><span class="co"># Determine differentiating markers for CD4+ T cell</span></span>
<span id="cb25-374"><a href="#cb25-374" aria-hidden="true" tabindex="-1"></a>cd4_tcells <span class="ot">&lt;-</span> <span class="fu">FindMarkers</span>(seurat_integrated,</span>
<span id="cb25-375"><a href="#cb25-375" aria-hidden="true" tabindex="-1"></a>                          <span class="at">ident.1 =</span> <span class="dv">2</span>,</span>
<span id="cb25-376"><a href="#cb25-376" aria-hidden="true" tabindex="-1"></a>                          <span class="at">ident.2 =</span> <span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">4</span>,<span class="dv">6</span>))                  </span>
<span id="cb25-377"><a href="#cb25-377" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-378"><a href="#cb25-378" aria-hidden="true" tabindex="-1"></a><span class="co"># Add gene symbols to the DE table</span></span>
<span id="cb25-379"><a href="#cb25-379" aria-hidden="true" tabindex="-1"></a>cd4_tcells <span class="ot">&lt;-</span> cd4_tcells <span class="sc">%&gt;%</span></span>
<span id="cb25-380"><a href="#cb25-380" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames_to_column</span>(<span class="at">var =</span> <span class="st">"gene"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb25-381"><a href="#cb25-381" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(<span class="at">y =</span> <span class="fu">unique</span>(annotations[, <span class="fu">c</span>(<span class="st">"gene_name"</span>, <span class="st">"description"</span>)]),</span>
<span id="cb25-382"><a href="#cb25-382" aria-hidden="true" tabindex="-1"></a>             <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"gene"</span> <span class="ot">=</span> <span class="st">"gene_name"</span>))</span>
<span id="cb25-383"><a href="#cb25-383" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-384"><a href="#cb25-384" aria-hidden="true" tabindex="-1"></a><span class="co"># Reorder columns and sort by padj      </span></span>
<span id="cb25-385"><a href="#cb25-385" aria-hidden="true" tabindex="-1"></a>cd4_tcells <span class="ot">&lt;-</span> cd4_tcells[, <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">3</span><span class="sc">:</span><span class="dv">5</span>,<span class="dv">2</span>,<span class="dv">6</span><span class="sc">:</span><span class="dv">7</span>)]</span>
<span id="cb25-386"><a href="#cb25-386" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-387"><a href="#cb25-387" aria-hidden="true" tabindex="-1"></a>cd4_tcells <span class="ot">&lt;-</span> cd4_tcells <span class="sc">%&gt;%</span></span>
<span id="cb25-388"><a href="#cb25-388" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">arrange</span>(p_val_adj) </span>
<span id="cb25-389"><a href="#cb25-389" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-390"><a href="#cb25-390" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-393"><a href="#cb25-393" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-394"><a href="#cb25-394" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb25-395"><a href="#cb25-395" aria-hidden="true" tabindex="-1"></a><span class="co"># View data</span></span>
<span id="cb25-396"><a href="#cb25-396" aria-hidden="true" tabindex="-1"></a><span class="fu">View</span>(cd4_tcells)</span>
<span id="cb25-397"><a href="#cb25-397" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-398"><a href="#cb25-398" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-399"><a href="#cb25-399" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;p</span> <span class="er">align</span><span class="ot">=</span><span class="st">"center"</span><span class="kw">&gt;</span></span>
<span id="cb25-400"><a href="#cb25-400" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;img</span> <span class="er">src</span><span class="ot">=</span><span class="st">"../img/cd4t_markers_table_SCTv2.png"</span> <span class="er">width</span><span class="ot">=</span><span class="st">"800"</span><span class="kw">&gt;</span></span>
<span id="cb25-401"><a href="#cb25-401" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;/p&gt;</span></span>
<span id="cb25-402"><a href="#cb25-402" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-403"><a href="#cb25-403" aria-hidden="true" tabindex="-1"></a>Of these top genes the **CREM gene** stands out as a marker of activation with a positive fold change.  We also see markers of naive or memory cells include the SELL and CCR7 genes with negative fold changes, which is in line with previous results. </span>
<span id="cb25-404"><a href="#cb25-404" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-405"><a href="#cb25-405" aria-hidden="true" tabindex="-1"></a>As markers for the naive and activated states both showed up in the marker list, it is helpful to visualize expression. Based on these plots it seems as though clusters 0 and 2 are reliably the naive T cells. However, for the activated T cells it is hard to tell. We might say that clusters 4 and 18 are activated T cells, but the CD69 expression is not as apparent as CREM. We will label the naive cells and leave the remaining clusters labeled as CD4+ T cells.</span>
<span id="cb25-406"><a href="#cb25-406" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-407"><a href="#cb25-407" aria-hidden="true" tabindex="-1"></a>Now taking all of this information, we can surmise the cell types of the different clusters and plot the cells with cell type labels.</span>
<span id="cb25-408"><a href="#cb25-408" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-409"><a href="#cb25-409" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-410"><a href="#cb25-410" aria-hidden="true" tabindex="-1"></a>| Cluster ID    | Cell Type |</span>
<span id="cb25-411"><a href="#cb25-411" aria-hidden="true" tabindex="-1"></a>|:-----:|:-----:|</span>
<span id="cb25-412"><a href="#cb25-412" aria-hidden="true" tabindex="-1"></a>|0  | Naive or memory CD4+ T cells|</span>
<span id="cb25-413"><a href="#cb25-413" aria-hidden="true" tabindex="-1"></a>|1  | CD14+ monocytes |</span>
<span id="cb25-414"><a href="#cb25-414" aria-hidden="true" tabindex="-1"></a>|2  | Activated T cells|</span>
<span id="cb25-415"><a href="#cb25-415" aria-hidden="true" tabindex="-1"></a>|3  | CD14+ monocytes|</span>
<span id="cb25-416"><a href="#cb25-416" aria-hidden="true" tabindex="-1"></a>|4  | Stressed cells / Unknown|</span>
<span id="cb25-417"><a href="#cb25-417" aria-hidden="true" tabindex="-1"></a>|5  | CD8+ T cells |</span>
<span id="cb25-418"><a href="#cb25-418" aria-hidden="true" tabindex="-1"></a>|6  | Naive or memory CD4+ T cells |</span>
<span id="cb25-419"><a href="#cb25-419" aria-hidden="true" tabindex="-1"></a>|7  | B cells |</span>
<span id="cb25-420"><a href="#cb25-420" aria-hidden="true" tabindex="-1"></a>|8  | NK cells |</span>
<span id="cb25-421"><a href="#cb25-421" aria-hidden="true" tabindex="-1"></a>|9  | CD8+ T cells |</span>
<span id="cb25-422"><a href="#cb25-422" aria-hidden="true" tabindex="-1"></a>|10 | FCGR3A+ monocytes |</span>
<span id="cb25-423"><a href="#cb25-423" aria-hidden="true" tabindex="-1"></a>|11 | B cells |</span>
<span id="cb25-424"><a href="#cb25-424" aria-hidden="true" tabindex="-1"></a>|12 | NK cells |</span>
<span id="cb25-425"><a href="#cb25-425" aria-hidden="true" tabindex="-1"></a>|13 | B cells|</span>
<span id="cb25-426"><a href="#cb25-426" aria-hidden="true" tabindex="-1"></a>|14 | Conventional dendritic cells |</span>
<span id="cb25-427"><a href="#cb25-427" aria-hidden="true" tabindex="-1"></a>|15| Megakaryocytes |</span>
<span id="cb25-428"><a href="#cb25-428" aria-hidden="true" tabindex="-1"></a>|16| Plasmacytoid dendritic cells |</span>
<span id="cb25-429"><a href="#cb25-429" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-430"><a href="#cb25-430" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-431"><a href="#cb25-431" aria-hidden="true" tabindex="-1"></a>We can then reassign the identity of the clusters to these cell types:</span>
<span id="cb25-432"><a href="#cb25-432" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-435"><a href="#cb25-435" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-436"><a href="#cb25-436" aria-hidden="true" tabindex="-1"></a><span class="co"># Rename all identities</span></span>
<span id="cb25-437"><a href="#cb25-437" aria-hidden="true" tabindex="-1"></a>seurat_integrated <span class="ot">&lt;-</span> <span class="fu">RenameIdents</span>(<span class="at">object =</span> seurat_integrated, </span>
<span id="cb25-438"><a href="#cb25-438" aria-hidden="true" tabindex="-1"></a>                               <span class="st">"0"</span> <span class="ot">=</span> <span class="st">"Naive or memory CD4+ T cells"</span>,</span>
<span id="cb25-439"><a href="#cb25-439" aria-hidden="true" tabindex="-1"></a>                               <span class="st">"1"</span> <span class="ot">=</span> <span class="st">"CD14+ monocytes"</span>,</span>
<span id="cb25-440"><a href="#cb25-440" aria-hidden="true" tabindex="-1"></a>                               <span class="st">"2"</span> <span class="ot">=</span> <span class="st">"Activated T cells"</span>,</span>
<span id="cb25-441"><a href="#cb25-441" aria-hidden="true" tabindex="-1"></a>                               <span class="st">"3"</span> <span class="ot">=</span> <span class="st">"CD14+ monocytes"</span>,</span>
<span id="cb25-442"><a href="#cb25-442" aria-hidden="true" tabindex="-1"></a>                               <span class="st">"4"</span> <span class="ot">=</span> <span class="st">"Stressed cells / Unknown"</span>,</span>
<span id="cb25-443"><a href="#cb25-443" aria-hidden="true" tabindex="-1"></a>                               <span class="st">"5"</span> <span class="ot">=</span> <span class="st">"CD8+ T cells"</span>,</span>
<span id="cb25-444"><a href="#cb25-444" aria-hidden="true" tabindex="-1"></a>                               <span class="st">"6"</span> <span class="ot">=</span> <span class="st">"Naive or memory CD4+ T cells"</span>,</span>
<span id="cb25-445"><a href="#cb25-445" aria-hidden="true" tabindex="-1"></a>                               <span class="st">"7"</span> <span class="ot">=</span> <span class="st">"B cells"</span>,</span>
<span id="cb25-446"><a href="#cb25-446" aria-hidden="true" tabindex="-1"></a>                               <span class="st">"8"</span> <span class="ot">=</span> <span class="st">"NK cells"</span>,</span>
<span id="cb25-447"><a href="#cb25-447" aria-hidden="true" tabindex="-1"></a>                               <span class="st">"9"</span> <span class="ot">=</span> <span class="st">"CD8+ T cells"</span>,</span>
<span id="cb25-448"><a href="#cb25-448" aria-hidden="true" tabindex="-1"></a>                               <span class="st">"10"</span> <span class="ot">=</span> <span class="st">"FCGR3A+ monocytes"</span>,</span>
<span id="cb25-449"><a href="#cb25-449" aria-hidden="true" tabindex="-1"></a>                               <span class="st">"11"</span> <span class="ot">=</span> <span class="st">"B cells"</span>,</span>
<span id="cb25-450"><a href="#cb25-450" aria-hidden="true" tabindex="-1"></a>                               <span class="st">"12"</span> <span class="ot">=</span> <span class="st">"NK cells"</span>,</span>
<span id="cb25-451"><a href="#cb25-451" aria-hidden="true" tabindex="-1"></a>                               <span class="st">"13"</span> <span class="ot">=</span> <span class="st">"B cells"</span>,</span>
<span id="cb25-452"><a href="#cb25-452" aria-hidden="true" tabindex="-1"></a>                               <span class="st">"14"</span> <span class="ot">=</span> <span class="st">"Conventional dendritic cells"</span>,</span>
<span id="cb25-453"><a href="#cb25-453" aria-hidden="true" tabindex="-1"></a>                               <span class="st">"15"</span> <span class="ot">=</span> <span class="st">"Megakaryocytes"</span>,</span>
<span id="cb25-454"><a href="#cb25-454" aria-hidden="true" tabindex="-1"></a>                   <span class="st">"16"</span> <span class="ot">=</span> <span class="st">"Plasmacytoid dendritic cells"</span>)</span>
<span id="cb25-455"><a href="#cb25-455" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-456"><a href="#cb25-456" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-457"><a href="#cb25-457" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot the UMAP</span></span>
<span id="cb25-458"><a href="#cb25-458" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(<span class="at">object =</span> seurat_integrated, </span>
<span id="cb25-459"><a href="#cb25-459" aria-hidden="true" tabindex="-1"></a>        <span class="at">reduction =</span> <span class="st">"umap"</span>, </span>
<span id="cb25-460"><a href="#cb25-460" aria-hidden="true" tabindex="-1"></a>        <span class="at">label =</span> <span class="cn">TRUE</span>,</span>
<span id="cb25-461"><a href="#cb25-461" aria-hidden="true" tabindex="-1"></a>        <span class="at">label.size =</span> <span class="dv">3</span>,</span>
<span id="cb25-462"><a href="#cb25-462" aria-hidden="true" tabindex="-1"></a>        <span class="at">repel =</span> <span class="cn">TRUE</span>)</span>
<span id="cb25-463"><a href="#cb25-463" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-464"><a href="#cb25-464" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-465"><a href="#cb25-465" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-466"><a href="#cb25-466" aria-hidden="true" tabindex="-1"></a>If we wanted to remove the potentially stressed cells, we could use the <span class="in">`subset()`</span> function:</span>
<span id="cb25-467"><a href="#cb25-467" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-470"><a href="#cb25-470" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-471"><a href="#cb25-471" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove the stressed or dying cells</span></span>
<span id="cb25-472"><a href="#cb25-472" aria-hidden="true" tabindex="-1"></a>seurat_subset_labeled <span class="ot">&lt;-</span> <span class="fu">subset</span>(seurat_integrated,</span>
<span id="cb25-473"><a href="#cb25-473" aria-hidden="true" tabindex="-1"></a>                               <span class="at">idents =</span> <span class="st">"Stressed cells / Unknown"</span>, <span class="at">invert =</span> <span class="cn">TRUE</span>)</span>
<span id="cb25-474"><a href="#cb25-474" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-475"><a href="#cb25-475" aria-hidden="true" tabindex="-1"></a><span class="co"># Re-visualize the clusters</span></span>
<span id="cb25-476"><a href="#cb25-476" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(<span class="at">object =</span> seurat_subset_labeled, </span>
<span id="cb25-477"><a href="#cb25-477" aria-hidden="true" tabindex="-1"></a>        <span class="at">reduction =</span> <span class="st">"umap"</span>, </span>
<span id="cb25-478"><a href="#cb25-478" aria-hidden="true" tabindex="-1"></a>        <span class="at">label =</span> <span class="cn">TRUE</span>,</span>
<span id="cb25-479"><a href="#cb25-479" aria-hidden="true" tabindex="-1"></a>        <span class="at">label.size =</span> <span class="dv">3</span>,</span>
<span id="cb25-480"><a href="#cb25-480" aria-hidden="true" tabindex="-1"></a>    <span class="at">repel =</span> <span class="cn">TRUE</span>)</span>
<span id="cb25-481"><a href="#cb25-481" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-482"><a href="#cb25-482" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-483"><a href="#cb25-483" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-484"><a href="#cb25-484" aria-hidden="true" tabindex="-1"></a>Now we would want to save our final labelled Seurat object and the output of <span class="in">`sessionInfo()`</span>:</span>
<span id="cb25-485"><a href="#cb25-485" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-488"><a href="#cb25-488" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-489"><a href="#cb25-489" aria-hidden="true" tabindex="-1"></a><span class="co"># Save final R object</span></span>
<span id="cb25-490"><a href="#cb25-490" aria-hidden="true" tabindex="-1"></a><span class="fu">write_rds</span>(seurat_integrated,</span>
<span id="cb25-491"><a href="#cb25-491" aria-hidden="true" tabindex="-1"></a>          <span class="at">file =</span> <span class="st">"../results/seurat_labelled.rds"</span>)</span>
<span id="cb25-492"><a href="#cb25-492" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-493"><a href="#cb25-493" aria-hidden="true" tabindex="-1"></a><span class="co"># Create and save a text file with sessionInfo</span></span>
<span id="cb25-494"><a href="#cb25-494" aria-hidden="true" tabindex="-1"></a><span class="fu">sink</span>(<span class="st">"../results/sessionInfo_scrnaseq_Feb2023.txt"</span>)</span>
<span id="cb25-495"><a href="#cb25-495" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span>
<span id="cb25-496"><a href="#cb25-496" aria-hidden="true" tabindex="-1"></a><span class="fu">sink</span>()</span>
<span id="cb25-497"><a href="#cb25-497" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-498"><a href="#cb25-498" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-499"><a href="#cb25-499" aria-hidden="true" tabindex="-1"></a>::: callout-note</span>
<span id="cb25-500"><a href="#cb25-500" aria-hidden="true" tabindex="-1"></a>You can find out more about the <span class="in">`sink()`</span> function <span class="co">[</span><span class="ot">at this link</span><span class="co">](https://www.rdocumentation.org/packages/base/versions/3.6.2/topics/sink)</span>.</span>
<span id="cb25-501"><a href="#cb25-501" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb25-502"><a href="#cb25-502" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-503"><a href="#cb25-503" aria-hidden="true" tabindex="-1"></a>***</span>
<span id="cb25-504"><a href="#cb25-504" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-505"><a href="#cb25-505" aria-hidden="true" tabindex="-1"></a>Now that we have our clusters defined and the markers for each of our clusters, we have a few different questions we can answer:</span>
<span id="cb25-506"><a href="#cb25-506" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-507"><a href="#cb25-507" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Determine if there is a shift in cell populations between <span class="in">`ctrl`</span> and <span class="in">`stim`</span>. Ideally this would be done with replicates to determine if the changes are significant.</span>
<span id="cb25-508"><a href="#cb25-508" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-509"><a href="#cb25-509" aria-hidden="true" tabindex="-1"></a><span class="in">```{r fig.width=18}</span></span>
<span id="cb25-510"><a href="#cb25-510" aria-hidden="true" tabindex="-1"></a><span class="co"># Add celltype annotation as a column in meta.data </span></span>
<span id="cb25-511"><a href="#cb25-511" aria-hidden="true" tabindex="-1"></a>seurat_subset_labeled<span class="sc">$</span>celltype <span class="ot">&lt;-</span> <span class="fu">Idents</span>(seurat_subset_labeled)</span>
<span id="cb25-512"><a href="#cb25-512" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-513"><a href="#cb25-513" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute number of cells per celltype</span></span>
<span id="cb25-514"><a href="#cb25-514" aria-hidden="true" tabindex="-1"></a>n_cells <span class="ot">&lt;-</span> <span class="fu">FetchData</span>(seurat_subset_labeled, </span>
<span id="cb25-515"><a href="#cb25-515" aria-hidden="true" tabindex="-1"></a>                     <span class="at">vars =</span> <span class="fu">c</span>(<span class="st">"celltype"</span>, <span class="st">"sample"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb25-516"><a href="#cb25-516" aria-hidden="true" tabindex="-1"></a>        dplyr<span class="sc">::</span><span class="fu">count</span>(celltype, sample)</span>
<span id="cb25-517"><a href="#cb25-517" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-518"><a href="#cb25-518" aria-hidden="true" tabindex="-1"></a><span class="co"># Barplot of number of cells per celltype by sample</span></span>
<span id="cb25-519"><a href="#cb25-519" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(n_cells, <span class="fu">aes</span>(<span class="at">x=</span>celltype, <span class="at">y=</span>n, <span class="at">fill=</span>sample)) <span class="sc">+</span></span>
<span id="cb25-520"><a href="#cb25-520" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_bar</span>(<span class="at">position=</span><span class="fu">position_dodge</span>(), <span class="at">stat=</span><span class="st">"identity"</span>) <span class="sc">+</span></span>
<span id="cb25-521"><a href="#cb25-521" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb25-522"><a href="#cb25-522" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_text</span>(<span class="fu">aes</span>(<span class="at">label=</span>n), <span class="at">vjust =</span> <span class="sc">-</span>.<span class="dv">2</span>, <span class="at">position=</span><span class="fu">position_dodge</span>(<span class="dv">1</span>))</span>
<span id="cb25-523"><a href="#cb25-523" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-524"><a href="#cb25-524" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-525"><a href="#cb25-525" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Perform differential expression analysis between conditions <span class="in">`ctrl`</span> and <span class="in">`stim`</span></span>
<span id="cb25-526"><a href="#cb25-526" aria-hidden="true" tabindex="-1"></a><span class="ss">    - </span>Biological replicates are **necessary** to proceed with this analysis, and we have <span class="co">[</span><span class="ot">additional materials</span><span class="co">](https://hbctraining.github.io/Pseudobulk-for-scRNAseq/)</span> to help walk through this analysis.</span>
<span id="cb25-527"><a href="#cb25-527" aria-hidden="true" tabindex="-1"></a><span class="ss">  - </span>For a first pass look, we can use the <span class="in">`FindMarkers()`</span> function we have been using to do a simple wilcox test to see the difference in gene expression between conditions for the B cells</span>
<span id="cb25-528"><a href="#cb25-528" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-531"><a href="#cb25-531" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-532"><a href="#cb25-532" aria-hidden="true" tabindex="-1"></a><span class="co"># Subset seurat object to just B cells</span></span>
<span id="cb25-533"><a href="#cb25-533" aria-hidden="true" tabindex="-1"></a>seurat_b_cells <span class="ot">&lt;-</span> <span class="fu">subset</span>(seurat_subset_labeled, <span class="at">subset =</span> (celltype <span class="sc">==</span> <span class="st">"B cells"</span>))</span>
<span id="cb25-534"><a href="#cb25-534" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-535"><a href="#cb25-535" aria-hidden="true" tabindex="-1"></a><span class="co"># Run a wilcox test to compare ctrl vs stim</span></span>
<span id="cb25-536"><a href="#cb25-536" aria-hidden="true" tabindex="-1"></a><span class="fu">Idents</span>(seurat_b_cells) <span class="ot">&lt;-</span> <span class="st">"sample"</span></span>
<span id="cb25-537"><a href="#cb25-537" aria-hidden="true" tabindex="-1"></a>b_markers <span class="ot">&lt;-</span> <span class="fu">FindMarkers</span>(seurat_b_cells,</span>
<span id="cb25-538"><a href="#cb25-538" aria-hidden="true" tabindex="-1"></a>                          <span class="at">ident.1 =</span> <span class="st">"ctrl"</span>,</span>
<span id="cb25-539"><a href="#cb25-539" aria-hidden="true" tabindex="-1"></a>                          <span class="at">ident.2 =</span> <span class="st">"stim"</span>,</span>
<span id="cb25-540"><a href="#cb25-540" aria-hidden="true" tabindex="-1"></a>                          <span class="at">grouping.var =</span> <span class="st">"sample"</span>,</span>
<span id="cb25-541"><a href="#cb25-541" aria-hidden="true" tabindex="-1"></a>                          <span class="at">only.pos =</span> <span class="cn">FALSE</span>,</span>
<span id="cb25-542"><a href="#cb25-542" aria-hidden="true" tabindex="-1"></a>                          <span class="at">logfc.threshold =</span> <span class="fl">0.25</span>)</span>
<span id="cb25-543"><a href="#cb25-543" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(b_markers)</span>
<span id="cb25-544"><a href="#cb25-544" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-545"><a href="#cb25-545" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-546"><a href="#cb25-546" aria-hidden="true" tabindex="-1"></a><span class="ss">  - </span>For added visualization, we can used the <span class="in">`EnhancedVolcano()`</span> function to see how the genes fall on a volcano plot.</span>
<span id="cb25-547"><a href="#cb25-547" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-548"><a href="#cb25-548" aria-hidden="true" tabindex="-1"></a><span class="in">```{r fig.height=6, fig.width=10}</span></span>
<span id="cb25-549"><a href="#cb25-549" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(EnhancedVolcano)</span>
<span id="cb25-550"><a href="#cb25-550" aria-hidden="true" tabindex="-1"></a><span class="fu">EnhancedVolcano</span>(b_markers,</span>
<span id="cb25-551"><a href="#cb25-551" aria-hidden="true" tabindex="-1"></a>    <span class="fu">row.names</span>(b_markers),</span>
<span id="cb25-552"><a href="#cb25-552" aria-hidden="true" tabindex="-1"></a>    <span class="at">x=</span><span class="st">"avg_log2FC"</span>,</span>
<span id="cb25-553"><a href="#cb25-553" aria-hidden="true" tabindex="-1"></a>    <span class="at">y=</span><span class="st">"p_val_adj"</span>,</span>
<span id="cb25-554"><a href="#cb25-554" aria-hidden="true" tabindex="-1"></a>    <span class="at">title=</span><span class="st">"B Cells"</span>,</span>
<span id="cb25-555"><a href="#cb25-555" aria-hidden="true" tabindex="-1"></a>    <span class="at">subtitle=</span><span class="st">"Stim vs. Ctrl"</span></span>
<span id="cb25-556"><a href="#cb25-556" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb25-557"><a href="#cb25-557" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-558"><a href="#cb25-558" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-559"><a href="#cb25-559" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Experimentally validate intriguing markers for our identified cell types.</span>
<span id="cb25-560"><a href="#cb25-560" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Explore a subset of the cell types to discover subclusters of cells as described <span class="co">[</span><span class="ot">here</span><span class="co">](seurat_subclustering.qmd)</span></span>
<span id="cb25-561"><a href="#cb25-561" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Trajectory analysis, or lineage tracing, could be performed if trying to determine the progression between cell types or cell states. For example, we could explore any of the following using this type of analysis:</span>
<span id="cb25-562"><a href="#cb25-562" aria-hidden="true" tabindex="-1"></a><span class="ss">    - </span>Differentiation processes</span>
<span id="cb25-563"><a href="#cb25-563" aria-hidden="true" tabindex="-1"></a><span class="ss">    - </span>Expression changes over time</span>
<span id="cb25-564"><a href="#cb25-564" aria-hidden="true" tabindex="-1"></a><span class="ss">    - </span>Cell state changes in expression</span>
</code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center">This lesson has been developed by members of the teaching team at the <a href="http://bioinformatics.sph.harvard.edu/">Harvard Chan Bioinformatics Core (HBC)</a>. <br> These are open access materials distributed under the terms of the <a href="https://creativecommons.org/licenses/by/4.0/">Creative Commons Attribution license</a> (CC BY 4.0), which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are credited. <br> A portion of these materials and hands-on activities were adapted from the <a href="https://satijalab.org/">Satija Lab’s</a> <a href="https://satijalab.org/seurat/pbmc3k_tutorial.html">Seurat - Guided Clustering Tutorial</a></div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>



</body></html>