<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.353">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Introduction to single-cell RNA-seq – schedule</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>


<link rel="stylesheet" href="../css/styles.css">
</head>

<body class="nav-fixed fullcontent">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">Introduction to single-cell RNA-seq</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link active" href="../lessons/schedule.html" rel="" target="" aria-current="page">
 <span class="menu-text">Schedule</span></a>
  </li>  
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="https://bioinformatics.sph.harvard.edu/" rel="" target="">
 <span class="menu-text">HBC</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="https://github.com/hbctraining/intro_r_rmd" rel="" target="">
 <span class="menu-text">GitHub</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="mailto:hbctraining@hsph.harvard.edu" rel="" target="">
 <span class="menu-text">Contact us</span></a>
  </li>  
</ul>
            <div class="quarto-navbar-tools">
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    
<!-- main -->
<main class="content" id="quarto-document-content"><div class="quarto-title-block"><div class="quarto-title-tools-only"><h1></h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source"><i class="bi"></i> Code</button></div></div>



<section id="introduction-to-single-cell-rna-seq" class="level1">
<h1>Introduction to single-cell RNA-seq</h1>
<section id="pre-reading" class="level2">
<h2 class="anchored" data-anchor-id="pre-reading">Pre-reading</h2>
<ul>
<li><a href="../lessons/01_intro_to_scRNA-seq.html">Introduction to scRNA-seq</a></li>
<li><a href="../lessons/02_SC_generation_of_count_matrix.html">Raw data to count matrix</a></li>
<li><a href="https://www.dropbox.com/s/vop78wq76h02a2f/single_cell_rnaseq.zip?dl=1">Download this project</a></li>
</ul>
</section>
<section id="day-1" class="level2">
<h2 class="anchored" data-anchor-id="day-1">Day 1</h2>
<table class="table">
<thead>
<tr class="header">
<th style="text-align: center;">Time</th>
<th style="text-align: center;">Topic</th>
<th style="text-align: center;">Instructor</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">09:30 - 09:45</td>
<td style="text-align: center;"><a href="../slides/workshop_intro_slides.pdf">Workshop introduction</a></td>
<td style="text-align: center;">Will</td>
</tr>
<tr class="even">
<td style="text-align: center;">09:45 - 10:35</td>
<td style="text-align: center;"><a href="../slides/072624_Chan_scRNAseq_workshop_Arpita_Kulkarni.pdf">Introduction to Single Cell RNA-sequencing: a practical guide</a></td>
<td style="text-align: center;"><a href="https://singlecellcore.hms.harvard.edu/people/arpita-kulkarni-phd">Dr.&nbsp;Arpita Kulkarni</a></td>
</tr>
<tr class="odd">
<td style="text-align: center;">10:35 - 10:40</td>
<td style="text-align: center;">Break</td>
<td style="text-align: center;"></td>
</tr>
<tr class="even">
<td style="text-align: center;">10:40 - 11:00</td>
<td style="text-align: center;">scRNA-seq pre-reading discussion</td>
<td style="text-align: center;">All</td>
</tr>
<tr class="odd">
<td style="text-align: center;">11:00 - 11:45</td>
<td style="text-align: center;"><a href="../lessons/03_SC_quality_control-setup.html">Quality control set-up</a></td>
<td style="text-align: center;">Noor</td>
</tr>
<tr class="even">
<td style="text-align: center;">11:45 - 12:00</td>
<td style="text-align: center;">Overview of self-learning materials and homework submission</td>
<td style="text-align: center;">Will</td>
</tr>
</tbody>
</table>
<section id="before-the-next-class" class="level3">
<h3 class="anchored" data-anchor-id="before-the-next-class">Before the next class:</h3>
<p>I. Please <strong>study the contents</strong> and <strong>work through all the code</strong> within the following lessons:</p>
<ol type="1">
<li><a href="../lessons/04_cellranger_QC.html">Quality control of cellranger counts</a>
<details>
<summary>
<i>Click here for a preview of this lesson</i>
</summary>
<br>Before you start any analysis, it’s important to know whether or not you have good quality cells. At these early stages you can flag or remove samples that could produce erroneous results downstream. <br><br>In this lesson you will:<br> - Discuss the outputs of cellranger and how to run it <br> - Review web summary HTML report<br> - Create plots from metrics_summary.csv file <br><br>
</details></li>
<li><a href="../lessons/04_SC_quality_control.html">Quality control with additional metrics</a>
<details>
<summary>
<i>Click here for a preview of this lesson</i>
</summary>
<br>In addition to the QC generated by cellranger, we can also compute some of our own metrics based on the raw data we have loaded into our Seurat object. <br><br>In this lesson you will:<br> - Compute essential QC metrics for each sample<br> - Create plots to visualize metrics across cells per sample<br> - Critically evaluate each plot and learn what each QC metric means<br><br>
</details></li>
<li><a href="../lessons/05_theory_of_PCA.html">Theory of PCA</a>
<details>
<summary>
<i>Click here for a preview of this lesson</i>
</summary>
<br>Before we can begin the next steps of the workflow, we need to make sure you have a good understanding of Principal Components Analysis (PCA). This method will be utilized in the scRNA-seq analysis workflow, and this foundation will help you better navigate those steps and interpretation of results.<br><br>
</details></li>
</ol>
<ol start="2" type="I">
<li><strong>Submit your work</strong>:</li>
</ol>
<ul>
<li>Each lesson above contains exercises; please go through each of them.</li>
<li><strong>Submit your answers</strong> to the exercises using <a href="https://docs.google.com/forms/d/e/1FAIpQLSf4zw37eUcuNCmHxCLikg3JENoxGWIuv5c_5pqHdYNId81Arg/viewform">this Google form</a> on <strong>the day <em>before</em> the next class</strong>.</li>
</ul>
</section>
<section id="questions" class="level3">
<h3 class="anchored" data-anchor-id="questions">Questions?</h3>
<ul>
<li><strong><em>If you get stuck due to an error</em></strong> while runnning code in the lesson, <a href="mailto:hbctraining@hsph.harvard.edu">email us</a></li>
</ul>
<hr>
</section>
</section>
<section id="day-2" class="level2">
<h2 class="anchored" data-anchor-id="day-2">Day 2</h2>
<table class="table">
<thead>
<tr class="header">
<th style="text-align: center;">Time</th>
<th style="text-align: center;">Topic</th>
<th style="text-align: center;">Instructor</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">09:30 - 10:15</td>
<td style="text-align: center;">Self-learning lessons discussion</td>
<td style="text-align: center;">All</td>
</tr>
<tr class="even">
<td style="text-align: center;">10:15 - 11:15</td>
<td style="text-align: center;"><a href="../lessons/06_SC_SCT_normalization.html">Normalization and regressing out unwanted variation</a></td>
<td style="text-align: center;">Noor</td>
</tr>
<tr class="odd">
<td style="text-align: center;">11:15 - 11:25</td>
<td style="text-align: center;">Break</td>
<td style="text-align: center;"></td>
</tr>
<tr class="even">
<td style="text-align: center;">11:25 - 12:00</td>
<td style="text-align: center;"><a href="../lessons/06a_integration_cca_theory.md">A brief introduction to Integration</a></td>
<td style="text-align: center;">Will</td>
</tr>
</tbody>
</table>
<section id="before-the-next-class-1" class="level3">
<h3 class="anchored" data-anchor-id="before-the-next-class-1">Before the next class:</h3>
<p>I. Please <strong>study the contents</strong> and <strong>work through all the code</strong> within the following lessons:</p>
<ol type="1">
<li><a href="../lessons/06b_integration_code_harmony.md">Running CCA integration and complex integration tasks</a>
<details>
<summary>
<i>Click here for a preview of this lesson</i>
</summary>
<br>In class, we described the theory of integration and in what situations we would implement it. <br><br>In this lesson you will:<br> - Run the code to implement CCA integration <br> - Evaluate the effect of integration on the UMAP <br> - Learn about methods for complex integration tasks (Harmonizing samples) <br>
</details></li>
<li><a href="../lessons/07_SC_clustering_cells_SCT.md">Clustering</a>
<details>
<summary>
<i>Click here for a preview of this lesson</i>
</summary>
<br>From the UMAP visualization of our data we can see that the cells are positioned into groups. Our next task is to isolate clusters of cells that are most similar to one another based on gene expression. <br><br>In this lesson you will:<br> - Learn the theory behind clustering and how it is performed in Seurat<br> - Cluster cells and visualize them on the UMAP<br>
</details></li>
<li><a href="../lessons/08_SC_clustering_quality_control.md">Clustering quality control</a>
<details>
<summary>
<i>Click here for a preview of this lesson</i>
</summary>
<br>After separating cells into clusters, it is crtical to evaluate whether they are biologically meaningful or not. At this point we can also decide if we need to re-cluster and/or potentialy go back to a previous QC step. <br><br>In this lesson you will:<br> - Check to see that clusters are not influenced by uninteresting sources of variation<br> - Check to see whether the major principal components are driving the different clusters<br> - Explore the cell type identities by looking at the expression for known markers across the clusters.<br>
</details></li>
<li><a href="../lessons/seurat_cheatsheet.md">Seurat Cheatsheet</a>
<details>
<summary>
<i>Click here for a preview of this lesson</i>
</summary>
<br>At this point, we have populated our seurat object with many different pieces of information. Knowing how to access different values will allow you to interact more efficiently with your dataset. <br><br>In this lesson you will:<br> - Explore the different parts of a seurat object.<br> - Use the built-in functions from the Seurat package for visualizations and grabbing data.<br>
</details></li>
</ol>
<ol start="2" type="I">
<li><strong>Submit your work</strong>:</li>
</ol>
<ul>
<li>Each lesson above contains exercises; please go through each of them.</li>
<li><strong>Submit your answers</strong> to the exercises using <a href="https://docs.google.com/forms/d/e/1FAIpQLSd8-k5YwfQekQQdFuDVPhGCy-eK70CqPiUkDNYDeWSEQYBAZg/viewform?usp=sf_link">this Google form</a> on <strong>the day <em>before</em> the next class</strong>.</li>
</ul>
</section>
<section id="questions-1" class="level3">
<h3 class="anchored" data-anchor-id="questions-1">Questions?</h3>
<ul>
<li><strong><em>If you get stuck due to an error</em></strong> while runnning code in the lesson, <a href="mailto:hbctraining@hsph.harvard.edu">email us</a></li>
</ul>
<hr>
</section>
</section>
<section id="day-3" class="level2">
<h2 class="anchored" data-anchor-id="day-3">Day 3</h2>
<table class="table">
<thead>
<tr class="header">
<th style="text-align: center;">Time</th>
<th style="text-align: center;">Topic</th>
<th style="text-align: center;">Instructor</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">9:30 - 10:00</td>
<td style="text-align: center;">Self-learning lessons discussion</td>
<td style="text-align: center;">All</td>
</tr>
<tr class="even">
<td style="text-align: center;">10:00 - 11:00</td>
<td style="text-align: center;"><a href="../lessons/09_merged_SC_marker_identification.md">Marker identification</a></td>
<td style="text-align: center;">Noor</td>
</tr>
<tr class="odd">
<td style="text-align: center;">11:00 - 11:10</td>
<td style="text-align: center;">Break</td>
<td style="text-align: center;"></td>
</tr>
<tr class="even">
<td style="text-align: center;">11:10 - 11:30</td>
<td style="text-align: center;"><a href="../lessons/scRNAseq_workflow.md">Workflow summary</a></td>
<td style="text-align: center;">Will</td>
</tr>
<tr class="odd">
<td style="text-align: center;">11:30 - 11:45</td>
<td style="text-align: center;">Overview and Final Q &amp; A</td>
<td style="text-align: center;">All</td>
</tr>
<tr class="even">
<td style="text-align: center;">11:45- 12:00</td>
<td style="text-align: center;"><a href="../slides/Workshop_wrapup.pdf">Wrap up</a></td>
<td style="text-align: center;">Will</td>
</tr>
</tbody>
</table>
<hr>
</section>
<section id="answer-keys" class="level2">
<h2 class="anchored" data-anchor-id="answer-keys">Answer Keys</h2>
<ul>
<li><strong><a href="../homework/Day1_exercise_answer_key.md">Answer key - assignment #1</a></strong></li>
<li><strong><a href="../homework/Day2_exercise_answer_key.R">Answer key - assignment #2</a></strong></li>
</ul>
</section>
<section id="downstream-analyses" class="level2">
<h2 class="anchored" data-anchor-id="downstream-analyses">Downstream analyses</h2>
<ul>
<li><a href="../lessons/pseudobulk_DESeq2_scrnaseq.md">Differential expression between conditions</a></li>
</ul>
<hr>
</section>
<section id="resources" class="level2">
<h2 class="anchored" data-anchor-id="resources">Resources</h2>
<p>We have covered the analysis steps in quite a bit of detail for scRNA-seq exploration of cellular heterogeneity using the Seurat package. For more information on topics covered, we encourage you to take a look at the following resources:</p>
<section id="seurat-focused" class="level3">
<h3 class="anchored" data-anchor-id="seurat-focused">Seurat-focused</h3>
<ul>
<li><a href="https://satijalab.org/seurat/vignettes.html">Seurat vignettes</a></li>
<li><a href="https://satijalab.org/seurat/essential_commands.html">Seurat cheatsheet</a></li>
<li><a href="https://satijalab.org/scgd21/">Satija Lab: Single Cell Genomics Day</a></li>
<li><a href="../lessons/cell_cycle_scoring.md">Additional information about cell cycle scoring</a></li>
<li><a href="https://divingintogeneticsandgenomics.com/post/cca-alignment/">A nice explanation on CCA and cell label transfer in Seurat</a></li>
</ul>
</section>
<section id="scaling-up-scrna-seq-analysis-on-hpc" class="level3">
<h3 class="anchored" data-anchor-id="scaling-up-scrna-seq-analysis-on-hpc">Scaling up: scRNA-seq analysis on HPC</h3>
<ul>
<li>Using RStudio on O2
<ul>
<li><a href="https://harvardmed.atlassian.net/wiki/spaces/O2/pages/1623425967/RStudio+on+O2">HMSRC wiki page</a></li>
<li><a href="https://hbctraining.github.io/Intro-to-Unix-QMB/lessons/R_studio_on_02.html">HBC RStudio on O2 tutorial</a></li>
</ul></li>
</ul>
</section>
<section id="cell-type-annotation" class="level3">
<h3 class="anchored" data-anchor-id="cell-type-annotation">Cell type annotation</h3>
<ul>
<li>Databases with markers for manual annotation
<ul>
<li><a href="http://bio-bigdata.hrbmu.edu.cn/CellMarker/">CellMarker 2.0</a></li>
<li>Cell type signature gene sets from <a href="https://www.gsea-msigdb.org/gsea/msigdb/human/genesets.jsp?collection=C8">MSigDb</a></li>
<li><a href="https://cellxgene.cziscience.com/gene-expression">CELL x GENE from CZI</a></li>
</ul></li>
<li>Reference-based automated celltype annotation
<ul>
<li><a href="https://azimuth.hubmapconsortium.org/">Azimuth</a></li>
<li><a href="https://www.celltypist.org/">Celltypist</a></li>
</ul></li>
</ul>
</section>
<section id="highlighted-papers" class="level3">
<h3 class="anchored" data-anchor-id="highlighted-papers">Highlighted papers</h3>
<ul>
<li><a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-020-02032-0">“Sampling time-dependent artifacts in single-cell genomics studies.”</a> <em>Massoni-Badosa et al.</em> 2019</li>
<li><a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1830-0">“Dissociation of solid tumor tissues with cold active protease for single-cell RNA-seq minimizes conserved collagenase-associated stress responses.”</a> <em>O’Flanagan et al.</em> 2020</li>
<li><a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-020-02048-6">“Systematic assessment of tissue dissociation and storage biases in single-cell and single-nucleus RNA-seq workflows.”</a> <em>Denisenko et al.</em> 2020</li>
<li><a href="https://www.nature.com/articles/s41467-021-25960-2">“Confronting false discoveries in single-cell differential expression”, <em>Nature Communications</em> 2021</a></li>
<li><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6306246/">Single-nucleus and single-cell transcriptomes compared in matched cortical cell types</a></li>
<li><a href="https://www.nature.com/articles/s41591-020-0844-1">A single-cell and single-nucleus RNA-Seq toolbox for fresh and frozen human tumors</a></li>
<li><a href="https://www.nature.com/articles/s41576-020-00292-x">Ligand-receptor analysis with CellphoneDB</a></li>
<li><a href="https://www.nature.com/articles/s41576-023-00586-w">Best practices for single-cell analysis across modalities</a></li>
</ul>
</section>
<section id="other-online-scrna-seq-courses" class="level3">
<h3 class="anchored" data-anchor-id="other-online-scrna-seq-courses">Other online scRNA-seq courses:</h3>
<ul>
<li><a href="http://bioconductor.org/books/release/OSCA/">OSCA with Bioconductor</a></li>
<li><a href="https://liulab-dfci.github.io/bioinfo-combio/">DFCI/Shirley Liu</a></li>
<li><a href="https://www.singlecellcourse.org/">Wellcome Sanger Institute/Hemmberg Lab</a></li>
<li><a href="https://github.com/SingleCellTranscriptomics">ISCB Workshop</a></li>
<li><a href="https://broadinstitute.github.io/2020_scWorkshop/">Broad workshop</a></li>
<li><a href="https://nbisweden.github.io/workshop-scRNAseq/">SciLifeLab workshop</a></li>
</ul>


<!-- -->

</section>
</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb1" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="an">sidebar:</span><span class="co"> false</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="an">toc:</span><span class="co"> false</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="an">execute:</span><span class="co"> </span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="co">  eval: false</span></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="fu"># Introduction to single-cell RNA-seq</span></span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="fu">## Pre-reading</span></span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span><span class="co">[</span><span class="ot">Introduction to scRNA-seq</span><span class="co">](01_intro_to_scRNA-seq.qmd)</span></span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span><span class="co">[</span><span class="ot">Raw data to count matrix</span><span class="co">](02_SC_generation_of_count_matrix.qmd)</span></span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span><span class="co">[</span><span class="ot">Download this project</span><span class="co">](https://www.dropbox.com/s/vop78wq76h02a2f/single_cell_rnaseq.zip?dl=1)</span></span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a><span class="fu">## Day 1</span></span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a>| Time |  Topic  | Instructor |</span>
<span id="cb1-19"><a href="#cb1-19" aria-hidden="true" tabindex="-1"></a>|:-----------:|:----------:|:--------:|</span>
<span id="cb1-20"><a href="#cb1-20" aria-hidden="true" tabindex="-1"></a>| 09:30 - 09:45 | <span class="co">[</span><span class="ot">Workshop introduction</span><span class="co">](../slides/workshop_intro_slides.pdf)</span> | Will |</span>
<span id="cb1-21"><a href="#cb1-21" aria-hidden="true" tabindex="-1"></a>| 09:45 - 10:35| <span class="co">[</span><span class="ot">Introduction to Single Cell RNA-sequencing: a practical guide</span><span class="co">](../slides/072624_Chan_scRNAseq_workshop_Arpita_Kulkarni.pdf)</span> | <span class="co">[</span><span class="ot">Dr. Arpita Kulkarni</span><span class="co">](https://singlecellcore.hms.harvard.edu/people/arpita-kulkarni-phd)</span> |</span>
<span id="cb1-22"><a href="#cb1-22" aria-hidden="true" tabindex="-1"></a>| 10:35 - 10:40 | Break |</span>
<span id="cb1-23"><a href="#cb1-23" aria-hidden="true" tabindex="-1"></a>| 10:40 - 11:00 | scRNA-seq pre-reading discussion | All |</span>
<span id="cb1-24"><a href="#cb1-24" aria-hidden="true" tabindex="-1"></a>| 11:00 - 11:45 | <span class="co">[</span><span class="ot">Quality control set-up</span><span class="co">](03_SC_quality_control-setup.qmd)</span> | Noor |</span>
<span id="cb1-25"><a href="#cb1-25" aria-hidden="true" tabindex="-1"></a>| 11:45 - 12:00 | Overview of self-learning materials and homework submission | Will |</span>
<span id="cb1-26"><a href="#cb1-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-27"><a href="#cb1-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-28"><a href="#cb1-28" aria-hidden="true" tabindex="-1"></a><span class="fu">### Before the next class:</span></span>
<span id="cb1-29"><a href="#cb1-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-30"><a href="#cb1-30" aria-hidden="true" tabindex="-1"></a>I. Please **study the contents** and **work through all the code** within the following lessons:</span>
<span id="cb1-31"><a href="#cb1-31" aria-hidden="true" tabindex="-1"></a> </span>
<span id="cb1-32"><a href="#cb1-32" aria-hidden="true" tabindex="-1"></a><span class="ss">   1. </span><span class="co">[</span><span class="ot">Quality control of cellranger counts</span><span class="co">](04_cellranger_QC.qmd)</span></span>
<span id="cb1-33"><a href="#cb1-33" aria-hidden="true" tabindex="-1"></a>      <span class="kw">&lt;details&gt;</span></span>
<span id="cb1-34"><a href="#cb1-34" aria-hidden="true" tabindex="-1"></a>       <span class="kw">&lt;summary&gt;&lt;i&gt;</span>Click here for a preview of this lesson<span class="kw">&lt;/i&gt;&lt;/summary&gt;</span></span>
<span id="cb1-35"><a href="#cb1-35" aria-hidden="true" tabindex="-1"></a>       <span class="kw">&lt;br&gt;</span>Before you start any analysis, it’s important to know whether or not you have good quality cells. At these early stages you can flag or remove samples that could produce erroneous results downstream. <span class="kw">&lt;br&gt;&lt;br&gt;</span>In this lesson you will:<span class="kw">&lt;br&gt;</span></span>
<span id="cb1-36"><a href="#cb1-36" aria-hidden="true" tabindex="-1"></a><span class="ss">             - </span>Discuss the outputs of cellranger and how to run it <span class="kw">&lt;br&gt;</span></span>
<span id="cb1-37"><a href="#cb1-37" aria-hidden="true" tabindex="-1"></a><span class="ss">             - </span>Review web summary HTML report<span class="kw">&lt;br&gt;</span></span>
<span id="cb1-38"><a href="#cb1-38" aria-hidden="true" tabindex="-1"></a><span class="ss">             - </span>Create plots from metrics_summary.csv file <span class="kw">&lt;br&gt;&lt;br&gt;</span></span>
<span id="cb1-39"><a href="#cb1-39" aria-hidden="true" tabindex="-1"></a>        <span class="kw">&lt;/details&gt;</span></span>
<span id="cb1-40"><a href="#cb1-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-41"><a href="#cb1-41" aria-hidden="true" tabindex="-1"></a><span class="ss"> 2.  </span><span class="co">[</span><span class="ot">Quality control with additional metrics</span><span class="co">](04_SC_quality_control.qmd)</span></span>
<span id="cb1-42"><a href="#cb1-42" aria-hidden="true" tabindex="-1"></a>      <span class="kw">&lt;details&gt;</span></span>
<span id="cb1-43"><a href="#cb1-43" aria-hidden="true" tabindex="-1"></a>       <span class="kw">&lt;summary&gt;&lt;i&gt;</span>Click here for a preview of this lesson<span class="kw">&lt;/i&gt;&lt;/summary&gt;</span></span>
<span id="cb1-44"><a href="#cb1-44" aria-hidden="true" tabindex="-1"></a>         <span class="kw">&lt;br&gt;</span>In addition to the QC generated by cellranger, we can also compute some of our own metrics based on the raw data we have loaded into our Seurat object. <span class="kw">&lt;br&gt;&lt;br&gt;</span>In this lesson you will:<span class="kw">&lt;br&gt;</span></span>
<span id="cb1-45"><a href="#cb1-45" aria-hidden="true" tabindex="-1"></a><span class="ss">             - </span>Compute essential QC metrics for each sample<span class="kw">&lt;br&gt;</span></span>
<span id="cb1-46"><a href="#cb1-46" aria-hidden="true" tabindex="-1"></a><span class="ss">             - </span>Create plots to visualize metrics across cells per sample<span class="kw">&lt;br&gt;</span></span>
<span id="cb1-47"><a href="#cb1-47" aria-hidden="true" tabindex="-1"></a><span class="ss">             - </span>Critically evaluate each plot and learn what each QC metric means<span class="kw">&lt;br&gt;&lt;br&gt;</span></span>
<span id="cb1-48"><a href="#cb1-48" aria-hidden="true" tabindex="-1"></a>        <span class="kw">&lt;/details&gt;</span></span>
<span id="cb1-49"><a href="#cb1-49" aria-hidden="true" tabindex="-1"></a>     </span>
<span id="cb1-50"><a href="#cb1-50" aria-hidden="true" tabindex="-1"></a><span class="ss">   3. </span><span class="co">[</span><span class="ot">Theory of PCA</span><span class="co">](05_theory_of_PCA.qmd)</span></span>
<span id="cb1-51"><a href="#cb1-51" aria-hidden="true" tabindex="-1"></a>      <span class="kw">&lt;details&gt;</span></span>
<span id="cb1-52"><a href="#cb1-52" aria-hidden="true" tabindex="-1"></a>       <span class="kw">&lt;summary&gt;&lt;i&gt;</span>Click here for a preview of this lesson<span class="kw">&lt;/i&gt;&lt;/summary&gt;</span></span>
<span id="cb1-53"><a href="#cb1-53" aria-hidden="true" tabindex="-1"></a>         <span class="kw">&lt;br&gt;</span>Before we can begin the next steps of the workflow, we need to make sure you have a good understanding of Principal Components Analysis (PCA). This method will be utilized in the scRNA-seq analysis workflow, and this foundation will help you better navigate those steps and interpretation of results.<span class="kw">&lt;br&gt;&lt;br&gt;</span></span>
<span id="cb1-54"><a href="#cb1-54" aria-hidden="true" tabindex="-1"></a>        <span class="kw">&lt;/details&gt;</span></span>
<span id="cb1-55"><a href="#cb1-55" aria-hidden="true" tabindex="-1"></a>         </span>
<span id="cb1-56"><a href="#cb1-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-57"><a href="#cb1-57" aria-hidden="true" tabindex="-1"></a>II. **Submit your work**:</span>
<span id="cb1-58"><a href="#cb1-58" aria-hidden="true" tabindex="-1"></a><span class="ss">   * </span>Each lesson above contains exercises; please go through each of them.</span>
<span id="cb1-59"><a href="#cb1-59" aria-hidden="true" tabindex="-1"></a><span class="ss">   * </span>**Submit your answers** to the exercises using [this Google form](https://docs.google.com/forms/d/e/1FAIpQLSf4zw37eUcuNCmHxCLikg3JENoxGWIuv5c_5pqHdYNId81Arg/viewform) on **the day *before* the next class**.</span>
<span id="cb1-60"><a href="#cb1-60" aria-hidden="true" tabindex="-1"></a>   </span>
<span id="cb1-61"><a href="#cb1-61" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-62"><a href="#cb1-62" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-63"><a href="#cb1-63" aria-hidden="true" tabindex="-1"></a><span class="fu">### Questions?</span></span>
<span id="cb1-64"><a href="#cb1-64" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>***If you get stuck due to an error*** while runnning code in the lesson, <span class="co">[</span><span class="ot">email us</span><span class="co">](mailto:hbctraining@hsph.harvard.edu)</span> </span>
<span id="cb1-65"><a href="#cb1-65" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-66"><a href="#cb1-66" aria-hidden="true" tabindex="-1"></a>***</span>
<span id="cb1-67"><a href="#cb1-67" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-68"><a href="#cb1-68" aria-hidden="true" tabindex="-1"></a><span class="fu">## Day 2</span></span>
<span id="cb1-69"><a href="#cb1-69" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-70"><a href="#cb1-70" aria-hidden="true" tabindex="-1"></a>| Time |  Topic  | Instructor |</span>
<span id="cb1-71"><a href="#cb1-71" aria-hidden="true" tabindex="-1"></a>|:-----------:|:----------:|:--------:|</span>
<span id="cb1-72"><a href="#cb1-72" aria-hidden="true" tabindex="-1"></a>| 09:30 - 10:15 | Self-learning lessons discussion | All |</span>
<span id="cb1-73"><a href="#cb1-73" aria-hidden="true" tabindex="-1"></a>| 10:15 - 11:15|  <span class="co">[</span><span class="ot">Normalization and regressing out unwanted variation</span><span class="co">](06_SC_SCT_normalization.qmd)</span> | Noor |</span>
<span id="cb1-74"><a href="#cb1-74" aria-hidden="true" tabindex="-1"></a>| 11:15 - 11:25 | Break |</span>
<span id="cb1-75"><a href="#cb1-75" aria-hidden="true" tabindex="-1"></a>| 11:25 - 12:00| <span class="co">[</span><span class="ot">A brief introduction to Integration</span><span class="co">](../lessons/06a_integration_cca_theory.md)</span> | Will |</span>
<span id="cb1-76"><a href="#cb1-76" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-77"><a href="#cb1-77" aria-hidden="true" tabindex="-1"></a><span class="fu">### Before the next class:</span></span>
<span id="cb1-78"><a href="#cb1-78" aria-hidden="true" tabindex="-1"></a>I. Please **study the contents** and **work through all the code** within the following lessons:</span>
<span id="cb1-79"><a href="#cb1-79" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-80"><a href="#cb1-80" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span><span class="co">[</span><span class="ot">Running CCA integration and complex integration tasks</span><span class="co">](../lessons/06b_integration_code_harmony.md)</span></span>
<span id="cb1-81"><a href="#cb1-81" aria-hidden="true" tabindex="-1"></a>      <span class="kw">&lt;details&gt;</span></span>
<span id="cb1-82"><a href="#cb1-82" aria-hidden="true" tabindex="-1"></a>       <span class="kw">&lt;summary&gt;&lt;i&gt;</span>Click here for a preview of this lesson<span class="kw">&lt;/i&gt;&lt;/summary&gt;</span></span>
<span id="cb1-83"><a href="#cb1-83" aria-hidden="true" tabindex="-1"></a>        <span class="kw">&lt;br&gt;</span>In class, we described the theory of integration and in what situations we would implement it. <span class="kw">&lt;br&gt;&lt;br&gt;</span>In this lesson you will:<span class="kw">&lt;br&gt;</span></span>
<span id="cb1-84"><a href="#cb1-84" aria-hidden="true" tabindex="-1"></a><span class="ss">             - </span>Run the code to implement CCA integration <span class="kw">&lt;br&gt;</span></span>
<span id="cb1-85"><a href="#cb1-85" aria-hidden="true" tabindex="-1"></a><span class="ss">             - </span>Evaluate the effect of integration on the UMAP <span class="kw">&lt;br&gt;</span></span>
<span id="cb1-86"><a href="#cb1-86" aria-hidden="true" tabindex="-1"></a><span class="ss">             - </span>Learn about methods for complex integration tasks (Harmonizing samples) <span class="kw">&lt;br&gt;</span></span>
<span id="cb1-87"><a href="#cb1-87" aria-hidden="true" tabindex="-1"></a>        <span class="kw">&lt;/details&gt;</span></span>
<span id="cb1-88"><a href="#cb1-88" aria-hidden="true" tabindex="-1"></a>   </span>
<span id="cb1-89"><a href="#cb1-89" aria-hidden="true" tabindex="-1"></a><span class="ss">2. </span><span class="co">[</span><span class="ot">Clustering</span><span class="co">](../lessons/07_SC_clustering_cells_SCT.md)</span></span>
<span id="cb1-90"><a href="#cb1-90" aria-hidden="true" tabindex="-1"></a>      <span class="kw">&lt;details&gt;</span></span>
<span id="cb1-91"><a href="#cb1-91" aria-hidden="true" tabindex="-1"></a>       <span class="kw">&lt;summary&gt;&lt;i&gt;</span>Click here for a preview of this lesson<span class="kw">&lt;/i&gt;&lt;/summary&gt;</span></span>
<span id="cb1-92"><a href="#cb1-92" aria-hidden="true" tabindex="-1"></a>         <span class="kw">&lt;br&gt;</span>From the UMAP visualization of our data  we can see that the cells are positioned into groups. Our next task is to isolate clusters of cells that are most similar to one another based on gene expression. <span class="kw">&lt;br&gt;&lt;br&gt;</span>In this lesson you will:<span class="kw">&lt;br&gt;</span></span>
<span id="cb1-93"><a href="#cb1-93" aria-hidden="true" tabindex="-1"></a><span class="ss">             - </span>Learn the theory behind clustering and how it is performed in Seurat<span class="kw">&lt;br&gt;</span></span>
<span id="cb1-94"><a href="#cb1-94" aria-hidden="true" tabindex="-1"></a><span class="ss">             - </span>Cluster cells and visualize them on the UMAP<span class="kw">&lt;br&gt;</span></span>
<span id="cb1-95"><a href="#cb1-95" aria-hidden="true" tabindex="-1"></a>        <span class="kw">&lt;/details&gt;</span></span>
<span id="cb1-96"><a href="#cb1-96" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-97"><a href="#cb1-97" aria-hidden="true" tabindex="-1"></a><span class="ss">3. </span><span class="co">[</span><span class="ot">Clustering quality control</span><span class="co">](../lessons/08_SC_clustering_quality_control.md)</span></span>
<span id="cb1-98"><a href="#cb1-98" aria-hidden="true" tabindex="-1"></a>      <span class="kw">&lt;details&gt;</span></span>
<span id="cb1-99"><a href="#cb1-99" aria-hidden="true" tabindex="-1"></a>       <span class="kw">&lt;summary&gt;&lt;i&gt;</span>Click here for a preview of this lesson<span class="kw">&lt;/i&gt;&lt;/summary&gt;</span></span>
<span id="cb1-100"><a href="#cb1-100" aria-hidden="true" tabindex="-1"></a>         <span class="kw">&lt;br&gt;</span>After separating cells into clusters, it is crtical to evaluate whether they are biologically meaningful or not. At this point we can also decide if we need to re-cluster and/or potentialy go back to a previous QC step.</span>
<span id="cb1-101"><a href="#cb1-101" aria-hidden="true" tabindex="-1"></a>         <span class="kw">&lt;br&gt;&lt;br&gt;</span>In this lesson you will:<span class="kw">&lt;br&gt;</span></span>
<span id="cb1-102"><a href="#cb1-102" aria-hidden="true" tabindex="-1"></a><span class="ss">           - </span>Check to see that clusters are not influenced by uninteresting sources of variation<span class="kw">&lt;br&gt;</span></span>
<span id="cb1-103"><a href="#cb1-103" aria-hidden="true" tabindex="-1"></a><span class="ss">           - </span>Check to see whether the major principal components are driving the different clusters<span class="kw">&lt;br&gt;</span></span>
<span id="cb1-104"><a href="#cb1-104" aria-hidden="true" tabindex="-1"></a><span class="ss">           - </span>Explore the cell type identities by looking at the expression for known markers across the clusters.<span class="kw">&lt;br&gt;</span></span>
<span id="cb1-105"><a href="#cb1-105" aria-hidden="true" tabindex="-1"></a>        <span class="kw">&lt;/details&gt;</span></span>
<span id="cb1-106"><a href="#cb1-106" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-107"><a href="#cb1-107" aria-hidden="true" tabindex="-1"></a><span class="ss">4. </span><span class="co">[</span><span class="ot">Seurat Cheatsheet</span><span class="co">](../lessons/seurat_cheatsheet.md)</span> </span>
<span id="cb1-108"><a href="#cb1-108" aria-hidden="true" tabindex="-1"></a>      <span class="kw">&lt;details&gt;</span></span>
<span id="cb1-109"><a href="#cb1-109" aria-hidden="true" tabindex="-1"></a>       <span class="kw">&lt;summary&gt;&lt;i&gt;</span>Click here for a preview of this lesson<span class="kw">&lt;/i&gt;&lt;/summary&gt;</span></span>
<span id="cb1-110"><a href="#cb1-110" aria-hidden="true" tabindex="-1"></a>         <span class="kw">&lt;br&gt;</span>At this point, we have populated our seurat object with many different pieces of information. Knowing how to access different values will allow you to interact more efficiently with your dataset.</span>
<span id="cb1-111"><a href="#cb1-111" aria-hidden="true" tabindex="-1"></a>         <span class="kw">&lt;br&gt;&lt;br&gt;</span>In this lesson you will:<span class="kw">&lt;br&gt;</span></span>
<span id="cb1-112"><a href="#cb1-112" aria-hidden="true" tabindex="-1"></a><span class="ss">           - </span>Explore the different parts of a seurat object.<span class="kw">&lt;br&gt;</span></span>
<span id="cb1-113"><a href="#cb1-113" aria-hidden="true" tabindex="-1"></a><span class="ss">           - </span>Use the built-in functions from the Seurat package for visualizations and grabbing data.<span class="kw">&lt;br&gt;</span></span>
<span id="cb1-114"><a href="#cb1-114" aria-hidden="true" tabindex="-1"></a>        <span class="kw">&lt;/details&gt;</span></span>
<span id="cb1-115"><a href="#cb1-115" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-116"><a href="#cb1-116" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-117"><a href="#cb1-117" aria-hidden="true" tabindex="-1"></a>II. **Submit your work**:</span>
<span id="cb1-118"><a href="#cb1-118" aria-hidden="true" tabindex="-1"></a><span class="ss">   * </span>Each lesson above contains exercises; please go through each of them.</span>
<span id="cb1-119"><a href="#cb1-119" aria-hidden="true" tabindex="-1"></a><span class="ss">   * </span>**Submit your answers** to the exercises using [this Google form](https://docs.google.com/forms/d/e/1FAIpQLSd8-k5YwfQekQQdFuDVPhGCy-eK70CqPiUkDNYDeWSEQYBAZg/viewform?usp=sf_link) on **the day *before* the next class**.</span>
<span id="cb1-120"><a href="#cb1-120" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-121"><a href="#cb1-121" aria-hidden="true" tabindex="-1"></a><span class="fu">### Questions?</span></span>
<span id="cb1-122"><a href="#cb1-122" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>***If you get stuck due to an error*** while runnning code in the lesson, <span class="co">[</span><span class="ot">email us</span><span class="co">](mailto:hbctraining@hsph.harvard.edu)</span> </span>
<span id="cb1-123"><a href="#cb1-123" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-124"><a href="#cb1-124" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-125"><a href="#cb1-125" aria-hidden="true" tabindex="-1"></a>***</span>
<span id="cb1-126"><a href="#cb1-126" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-127"><a href="#cb1-127" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-128"><a href="#cb1-128" aria-hidden="true" tabindex="-1"></a><span class="fu">## Day 3</span></span>
<span id="cb1-129"><a href="#cb1-129" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-130"><a href="#cb1-130" aria-hidden="true" tabindex="-1"></a>| Time |  Topic  | Instructor |</span>
<span id="cb1-131"><a href="#cb1-131" aria-hidden="true" tabindex="-1"></a>|:-----------:|:----------:|:--------:|</span>
<span id="cb1-132"><a href="#cb1-132" aria-hidden="true" tabindex="-1"></a>| 9:30 - 10:00 | Self-learning lessons discussion | All |</span>
<span id="cb1-133"><a href="#cb1-133" aria-hidden="true" tabindex="-1"></a>| 10:00 - 11:00 |<span class="co">[</span><span class="ot">Marker identification</span><span class="co">](../lessons/09_merged_SC_marker_identification.md)</span> | Noor |</span>
<span id="cb1-134"><a href="#cb1-134" aria-hidden="true" tabindex="-1"></a>| 11:00 - 11:10 | Break |</span>
<span id="cb1-135"><a href="#cb1-135" aria-hidden="true" tabindex="-1"></a>| 11:10 - 11:30 | <span class="co">[</span><span class="ot">Workflow summary</span><span class="co">](../lessons/scRNAseq_workflow.md)</span> | Will |</span>
<span id="cb1-136"><a href="#cb1-136" aria-hidden="true" tabindex="-1"></a>| 11:30 - 11:45 | Overview and Final Q &amp; A | All |</span>
<span id="cb1-137"><a href="#cb1-137" aria-hidden="true" tabindex="-1"></a>| 11:45- 12:00 | <span class="co">[</span><span class="ot">Wrap up</span><span class="co">](../slides/Workshop_wrapup.pdf)</span> | Will |</span>
<span id="cb1-138"><a href="#cb1-138" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-139"><a href="#cb1-139" aria-hidden="true" tabindex="-1"></a>***</span>
<span id="cb1-140"><a href="#cb1-140" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-141"><a href="#cb1-141" aria-hidden="true" tabindex="-1"></a><span class="fu">## Answer Keys</span></span>
<span id="cb1-142"><a href="#cb1-142" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-143"><a href="#cb1-143" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>**[Answer key - assignment #1](../homework/Day1_exercise_answer_key.md)**</span>
<span id="cb1-144"><a href="#cb1-144" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>**[Answer key - assignment #2](../homework/Day2_exercise_answer_key.R)**</span>
<span id="cb1-145"><a href="#cb1-145" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-146"><a href="#cb1-146" aria-hidden="true" tabindex="-1"></a><span class="fu">## Downstream analyses</span></span>
<span id="cb1-147"><a href="#cb1-147" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-148"><a href="#cb1-148" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span><span class="co">[</span><span class="ot">Differential expression between conditions</span><span class="co">](../lessons/pseudobulk_DESeq2_scrnaseq.md)</span></span>
<span id="cb1-149"><a href="#cb1-149" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-150"><a href="#cb1-150" aria-hidden="true" tabindex="-1"></a>***</span>
<span id="cb1-151"><a href="#cb1-151" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-152"><a href="#cb1-152" aria-hidden="true" tabindex="-1"></a><span class="fu">## Resources</span></span>
<span id="cb1-153"><a href="#cb1-153" aria-hidden="true" tabindex="-1"></a>We have covered the analysis steps in quite a bit of detail for scRNA-seq exploration of cellular heterogeneity using the Seurat package. For more information on topics covered, we encourage you to take a look at the following resources:</span>
<span id="cb1-154"><a href="#cb1-154" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-155"><a href="#cb1-155" aria-hidden="true" tabindex="-1"></a><span class="fu">### Seurat-focused</span></span>
<span id="cb1-156"><a href="#cb1-156" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span><span class="co">[</span><span class="ot">Seurat vignettes</span><span class="co">](https://satijalab.org/seurat/vignettes.html)</span></span>
<span id="cb1-157"><a href="#cb1-157" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span><span class="co">[</span><span class="ot">Seurat cheatsheet</span><span class="co">](https://satijalab.org/seurat/essential_commands.html)</span></span>
<span id="cb1-158"><a href="#cb1-158" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span><span class="co">[</span><span class="ot">Satija Lab: Single Cell Genomics Day</span><span class="co">](https://satijalab.org/scgd21/)</span></span>
<span id="cb1-159"><a href="#cb1-159" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span><span class="co">[</span><span class="ot">Additional information about cell cycle scoring</span><span class="co">](../lessons/cell_cycle_scoring.md)</span></span>
<span id="cb1-160"><a href="#cb1-160" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span><span class="co">[</span><span class="ot">A nice explanation on CCA and cell label transfer in Seurat</span><span class="co">](https://divingintogeneticsandgenomics.com/post/cca-alignment/)</span></span>
<span id="cb1-161"><a href="#cb1-161" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-162"><a href="#cb1-162" aria-hidden="true" tabindex="-1"></a><span class="fu">### Scaling up: scRNA-seq analysis on HPC  </span></span>
<span id="cb1-163"><a href="#cb1-163" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>Using RStudio on O2</span>
<span id="cb1-164"><a href="#cb1-164" aria-hidden="true" tabindex="-1"></a><span class="ss">    * </span><span class="co">[</span><span class="ot">HMSRC wiki page</span><span class="co">](https://harvardmed.atlassian.net/wiki/spaces/O2/pages/1623425967/RStudio+on+O2)</span></span>
<span id="cb1-165"><a href="#cb1-165" aria-hidden="true" tabindex="-1"></a><span class="ss">    * </span><span class="co">[</span><span class="ot">HBC RStudio on O2 tutorial</span><span class="co">](https://hbctraining.github.io/Intro-to-Unix-QMB/lessons/R_studio_on_02.html)</span></span>
<span id="cb1-166"><a href="#cb1-166" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-167"><a href="#cb1-167" aria-hidden="true" tabindex="-1"></a><span class="fu">### Cell type annotation</span></span>
<span id="cb1-168"><a href="#cb1-168" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Databases with markers for manual annotation</span>
<span id="cb1-169"><a href="#cb1-169" aria-hidden="true" tabindex="-1"></a><span class="ss">  - </span><span class="co">[</span><span class="ot">CellMarker 2.0</span><span class="co">](http://bio-bigdata.hrbmu.edu.cn/CellMarker/)</span></span>
<span id="cb1-170"><a href="#cb1-170" aria-hidden="true" tabindex="-1"></a><span class="ss">  - </span>Cell type signature gene sets from <span class="co">[</span><span class="ot">MSigDb</span><span class="co">](https://www.gsea-msigdb.org/gsea/msigdb/human/genesets.jsp?collection=C8)</span></span>
<span id="cb1-171"><a href="#cb1-171" aria-hidden="true" tabindex="-1"></a><span class="ss">  - </span><span class="co">[</span><span class="ot">CELL x GENE from CZI</span><span class="co">](https://cellxgene.cziscience.com/gene-expression)</span></span>
<span id="cb1-172"><a href="#cb1-172" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Reference-based automated celltype annotation</span>
<span id="cb1-173"><a href="#cb1-173" aria-hidden="true" tabindex="-1"></a><span class="ss">  - </span><span class="co">[</span><span class="ot">Azimuth</span><span class="co">](https://azimuth.hubmapconsortium.org/)</span></span>
<span id="cb1-174"><a href="#cb1-174" aria-hidden="true" tabindex="-1"></a><span class="ss">  - </span><span class="co">[</span><span class="ot">Celltypist</span><span class="co">](https://www.celltypist.org/)</span></span>
<span id="cb1-175"><a href="#cb1-175" aria-hidden="true" tabindex="-1"></a> </span>
<span id="cb1-176"><a href="#cb1-176" aria-hidden="true" tabindex="-1"></a>   </span>
<span id="cb1-177"><a href="#cb1-177" aria-hidden="true" tabindex="-1"></a><span class="fu">### Highlighted papers</span></span>
<span id="cb1-178"><a href="#cb1-178" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-179"><a href="#cb1-179" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span><span class="co">[</span><span class="ot">"Sampling time-dependent artifacts in single-cell genomics studies."</span><span class="co">](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-020-02032-0)</span> *Massoni-Badosa et al.* 2019</span>
<span id="cb1-180"><a href="#cb1-180" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span><span class="co">[</span><span class="ot">"Dissociation of solid tumor tissues with cold active protease for single-cell RNA-seq minimizes conserved collagenase-associated stress responses."</span><span class="co">](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1830-0)</span> *O'Flanagan et al.* 2020</span>
<span id="cb1-181"><a href="#cb1-181" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span><span class="co">[</span><span class="ot">"Systematic assessment of tissue dissociation and storage biases in single-cell and single-nucleus RNA-seq workflows."</span><span class="co">](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-020-02048-6)</span> *Denisenko et al.* 2020</span>
<span id="cb1-182"><a href="#cb1-182" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span><span class="co">[</span><span class="ot">"Confronting false discoveries in single-cell differential expression", _Nature Communications_ 2021</span><span class="co">](https://www.nature.com/articles/s41467-021-25960-2)</span></span>
<span id="cb1-183"><a href="#cb1-183" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span><span class="co">[</span><span class="ot">Single-nucleus and single-cell transcriptomes compared in matched cortical cell types</span><span class="co">](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6306246/)</span></span>
<span id="cb1-184"><a href="#cb1-184" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span><span class="co">[</span><span class="ot">A single-cell and single-nucleus RNA-Seq toolbox for fresh and frozen human tumors</span><span class="co">](https://www.nature.com/articles/s41591-020-0844-1)</span></span>
<span id="cb1-185"><a href="#cb1-185" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span><span class="co">[</span><span class="ot">Ligand-receptor analysis with CellphoneDB</span><span class="co">](https://www.nature.com/articles/s41576-020-00292-x)</span></span>
<span id="cb1-186"><a href="#cb1-186" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span><span class="co">[</span><span class="ot">Best practices for single-cell analysis across modalities</span><span class="co">](https://www.nature.com/articles/s41576-023-00586-w)</span></span>
<span id="cb1-187"><a href="#cb1-187" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-188"><a href="#cb1-188" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-189"><a href="#cb1-189" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-190"><a href="#cb1-190" aria-hidden="true" tabindex="-1"></a><span class="fu">### Other online scRNA-seq courses:</span></span>
<span id="cb1-191"><a href="#cb1-191" aria-hidden="true" tabindex="-1"></a><span class="ss">  - </span><span class="co">[</span><span class="ot">OSCA with Bioconductor</span><span class="co">](http://bioconductor.org/books/release/OSCA/)</span></span>
<span id="cb1-192"><a href="#cb1-192" aria-hidden="true" tabindex="-1"></a><span class="ss">  - </span><span class="co">[</span><span class="ot">DFCI/Shirley Liu</span><span class="co">](https://liulab-dfci.github.io/bioinfo-combio/)</span></span>
<span id="cb1-193"><a href="#cb1-193" aria-hidden="true" tabindex="-1"></a><span class="ss">  - </span><span class="co">[</span><span class="ot">Wellcome Sanger Institute/Hemmberg Lab</span><span class="co">](https://www.singlecellcourse.org/)</span></span>
<span id="cb1-194"><a href="#cb1-194" aria-hidden="true" tabindex="-1"></a><span class="ss">  - </span><span class="co">[</span><span class="ot">ISCB Workshop</span><span class="co">](https://github.com/SingleCellTranscriptomics)</span></span>
<span id="cb1-195"><a href="#cb1-195" aria-hidden="true" tabindex="-1"></a><span class="ss">  - </span><span class="co">[</span><span class="ot">Broad workshop</span><span class="co">](https://broadinstitute.github.io/2020_scWorkshop/)</span></span>
<span id="cb1-196"><a href="#cb1-196" aria-hidden="true" tabindex="-1"></a><span class="ss">  - </span><span class="co">[</span><span class="ot">SciLifeLab workshop</span><span class="co">](https://nbisweden.github.io/workshop-scRNAseq/)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center">This lesson has been developed by members of the teaching team at the <a href="http://bioinformatics.sph.harvard.edu/">Harvard Chan Bioinformatics Core (HBC)</a>. <br> These are open access materials distributed under the terms of the <a href="https://creativecommons.org/licenses/by/4.0/">Creative Commons Attribution license</a> (CC BY 4.0), which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are credited. <br></div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>



</body></html>